{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyN0ExN5/vKjojz4VwY7zay0"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# PDB rebuidl missing residue but with exact aa len as given pdb"],"metadata":{"id":"Npye-tbcShif"}},{"cell_type":"code","source":["#Installing biopython using pip\n","!pip install biopython\n","\n","# Install pdb-tools if not already installed:\n","!pip install pdb-tools\n","\n","#Installing py3Dmol using pip\n","!pip install py3Dmol\n","\n","#And importing the py3Dmol module\n","import py3Dmol"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"R-jBTTHZTFlf","executionInfo":{"status":"ok","timestamp":1744068745313,"user_tz":240,"elapsed":21789,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"e0176c3d-1217-4fa4-8d87-cf649f3d2a06"},"execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting biopython\n","  Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (2.0.2)\n","Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m23.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hInstalling collected packages: biopython\n","Successfully installed biopython-1.85\n","Collecting pdb-tools\n","  Downloading pdb_tools-2.5.0-py3-none-any.whl.metadata (6.6 kB)\n","Downloading pdb_tools-2.5.0-py3-none-any.whl (207 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m207.3/207.3 kB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hInstalling collected packages: pdb-tools\n","Successfully installed pdb-tools-2.5.0\n","Collecting py3Dmol\n","  Downloading py3Dmol-2.4.2-py2.py3-none-any.whl.metadata (1.9 kB)\n","Downloading py3Dmol-2.4.2-py2.py3-none-any.whl (7.0 kB)\n","Installing collected packages: py3Dmol\n","Successfully installed py3Dmol-2.4.2\n"]}]},{"cell_type":"code","execution_count":2,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"uMhK9Ub2SejP","executionInfo":{"status":"ok","timestamp":1744068751804,"user_tz":240,"elapsed":6489,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"d298e191-72bd-4df7-975b-ef7e7bc4c0d7"},"outputs":[{"output_type":"stream","name":"stdout","text":["--2025-04-07 23:32:25--  https://salilab.org/modeller/10.4/modeller-10.4.tar.gz\n","Resolving salilab.org (salilab.org)... 169.230.79.19\n","Connecting to salilab.org (salilab.org)|169.230.79.19|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 38244158 (36M) [application/x-gzip]\n","Saving to: ‘modeller-10.4.tar.gz’\n","\n","modeller-10.4.tar.g 100%[===================>]  36.47M   100MB/s    in 0.4s    \n","\n","2025-04-07 23:32:25 (100 MB/s) - ‘modeller-10.4.tar.gz’ saved [38244158/38244158]\n","\n","/content/modeller-10.4\n","\u001b[H\u001b[2JInstallation of MODELLER 10.4\n","\n","This script will install MODELLER 10.4 into a specified directory\n","for which you have read/write permissions.\n","\n","To accept the default answers indicated in [...], press <Enter> only.\n","\n","------------------------------------------------------------------------\n","\n","The currently supported architectures are as follows:\n","\n","   1) Linux x86 PC (e.g. RedHat, SuSe).\n","   2) x86_64 (Opteron/EM64T) box (Linux).\n","   3) Alternative x86 Linux binary (e.g. for FreeBSD).\n","   4) Linux on 32-bit ARM (e.g. for Raspberry Pi).\n","   5) Linux on 64-bit ARM (e.g. for Raspberry Pi 3 or later).\n","\n","Select the type of your computer from the list above [2]: \n","\n","Full directory name in which to install MODELLER 10.4\n","[/root/bin/modeller10.4]: \n","\n","License key, obtained from our academic license server at\n","https://salilab.org/modeller/registration.html: \n","------------------------------------------------------------------------\n","\n","The following settings will be used: \n","\n","The type of this machine          : x86_64-intel8\n","Destination installation directory: /content/compiled/MODELLER\n","License key                       : MODELIRANJE\n","\n","If they are incorrect, abort installation (ctrl-c) and run Install again.\n","\n","Press <Enter> to begin the installation: \n","------------------------------------------------------------------------\n","\n","Installing Modeller files ...\n","\n","------------------------------------------------------------------------\n","\n","Creating Modeller startup scripts ...\n","Configuring install ...\n","Installation complete.\n","Press <Enter> to continue: \n","------------------------------------------------------------------------\n","The best way to use Modeller is as a Python module, if you have Python\n","installed (any version between 2.3 and 3.11). Do this using the script:\n","/content/compiled/MODELLER/bin/modpy.sh\n","Simply add your regular Python command line after this script, e.g.\n","/content/compiled/MODELLER/bin/modpy.sh python3 myscript.py\n","\n","Alternatively, you can skip the 'modpy.sh' prefix if you ensure that the\n","following directories containing Python modules are in your Python path:\n","/content/compiled/MODELLER/modlib\n","/content/compiled/MODELLER/lib/x86_64-intel8 (for Python 2.3 or Python 2.4)\n","/content/compiled/MODELLER/lib/x86_64-intel8/python2.5 (for Python 2.5 to 2.7)\n","/content/compiled/MODELLER/lib/x86_64-intel8/python3.0 (for Python 3.0, 3.1)\n","/content/compiled/MODELLER/lib/x86_64-intel8/python3.2 (for Python 3.2)\n","/content/compiled/MODELLER/lib/x86_64-intel8/python3.3 (for Python 3.3 to 3.11)\n","(e.g. by setting the PYTHONPATH environment variable) and that the\n","following directory containing dynamic libraries is in your dynamic\n","library search path:\n","/content/compiled/MODELLER/lib/x86_64-intel8\n","(e.g. by setting the LD_LIBRARY_PATH variable - LIBPATH on AIX systems).\n","Ask your systems administrator for help if you wish to do this.\n","------------------------------------------------------------------------\n","\n","Press <Enter> to continue: \n","------------------------------------------------------------------------\n","\n","If you don't have Python installed, you can also use the\n","/content/compiled/MODELLER/bin/mod10.4 script to run Modeller,\n","although note that some advanced functionality, such as parallel job\n","support, requires Python.\n","\n","------------------------------------------------------------------------\n","\n","If you have problems with installation, please see the INSTALLATION file.\n","\n","Documentation in HTML format can be found in the\n","/content/compiled/MODELLER/doc directory.\n","\n","A simple homology modeling test case is in examples/automodel/.\n","\n","------------------------------------------------------------------------\n","\n","Happy modeling!\n","MODELLER Team, 2022/10/28, r12463\n","\n","✅ Modeller successfully installed\n","/content\n"]}],"source":["# Step 1: Download and extract Modeller tarball\n","!wget https://salilab.org/modeller/10.4/modeller-10.4.tar.gz\n","!tar -zxf modeller-10.4.tar.gz\n","!mkdir -p /content/compiled/MODELLER\n","\n","# Step 2: Install Modeller (replace MODELIRANJE with your license key if needed)\n","%cd modeller-10.4\n","with open('modeller_config', 'w') as f:\n","    f.write(\"2\\n\")\n","    f.write(\"/content/compiled/MODELLER\\n\")\n","    f.write(\"MODELIRANJE\\n\")  # <-- Replace with your Modeller license key\n","!./Install < modeller_config\n","\n","# Step 3: Create symbolic link\n","!ln -sf /content/compiled/MODELLER/bin/mod10.4 /usr/bin/\n","\n","# Verify install\n","!mod10.4 | awk 'NR==1{if($1==\"usage:\") print \"✅ Modeller successfully installed\"; else print \"❌ Installation failed\"}'\n","\n","%cd /content\n"]},{"cell_type":"code","source":["# Example: 4bgq and UniProt ID O76039\n","pdb_id = \"4bgq\"\n","uniprot_id = \"O76039\"\n","\n","# Download PDB\n","!wget -q https://files.rcsb.org/download/{pdb_id}.pdb\n","\n","# Download UniProt FASTA\n","!wget -q https://www.uniprot.org/uniprot/{uniprot_id}.fasta -O {uniprot_id}.fasta\n","\n","# Create working directory\n","!mkdir -p /content/4bgq_fix\n","!mv {pdb_id}.pdb {uniprot_id}.fasta /content/4bgq_fix/\n","%cd /content/4bgq_fix\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"a-ONMUsRSwsF","executionInfo":{"status":"ok","timestamp":1744068754847,"user_tz":240,"elapsed":3041,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"d52e7e73-2ab6-4210-ebfb-5490fdb4582b"},"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["/content/4bgq_fix\n"]}]},{"cell_type":"code","source":["from Bio.PDB import PDBParser, PDBIO, Select\n","\n","class StandardResidueSelect(Select):\n","    def accept_residue(self, residue):\n","        return residue.id[0] == ' '\n","\n","parser = PDBParser(QUIET=True)\n","structure = parser.get_structure(\"4bgq\", \"4bgq.pdb\")\n","\n","io = PDBIO()\n","io.set_structure(structure)\n","io.save(\"4bgq_clean.pdb\", select=StandardResidueSelect())\n","\n","print(\"✅ Cleaned PDB saved as 4bgq_clean.pdb\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"GXvrHsRTSzJ7","executionInfo":{"status":"ok","timestamp":1744068755611,"user_tz":240,"elapsed":763,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"7868ecad-945b-4c7d-ef32-347c846acfd1"},"execution_count":4,"outputs":[{"output_type":"stream","name":"stdout","text":["✅ Cleaned PDB saved as 4bgq_clean.pdb\n"]}]},{"cell_type":"code","source":["from Bio import SeqIO\n","from Bio.PDB import PDBParser\n","from Bio.Seq import Seq\n","from Bio.SeqRecord import SeqRecord\n","\n","# Parse PDB for residue range\n","structure = PDBParser(QUIET=True).get_structure(\"4bgq\", \"4bgq_clean.pdb\")\n","chain = next(structure[0].get_chains())\n","residues = [r for r in chain.get_residues() if r.id[0] == ' ']\n","start_res, end_res = residues[0].id[1], residues[-1].id[1]\n","\n","# Extract UniProt sequence\n","uniprot_seq = str(next(SeqIO.parse(\"O76039.fasta\", \"fasta\")).seq)\n","\n","# Trim UniProt to match PDB span\n","pdb_seq = \"\".join([r.resname for r in residues])\n","trimmed_seq = uniprot_seq[:end_res]\n","\n","# Write PIR alignment\n","with open(\"alignment.ali\", \"w\") as f:\n","    f.write(f\"\"\">P1;4bgq\n","structureX:4bgq:{start_res}:A:{end_res}:A::::\n","{trimmed_seq}*\n",">P1;target\n","sequence:target:{start_res}:A:{end_res}:A::::\n","{trimmed_seq}*\n","\"\"\")\n","\n","print(\"✅ PIR alignment file created.\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"xmlQyKL-S13Z","executionInfo":{"status":"ok","timestamp":1744068755751,"user_tz":240,"elapsed":143,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"e2a516c5-74a2-4cdd-a070-b84854d9fd07"},"execution_count":5,"outputs":[{"output_type":"stream","name":"stdout","text":["✅ PIR alignment file created.\n"]}]},{"cell_type":"code","source":["from Bio.PDB import PDBParser\n","\n","def fix_alignment_ali(pdb_path, ali_path):\n","    parser = PDBParser(QUIET=True)\n","    structure = parser.get_structure(\"4bgq\", pdb_path)\n","    model = structure[0]\n","    chain = next(model.get_chains())\n","    chain_id = chain.id\n","    residues = [res for res in chain.get_residues() if res.id[0] == ' ']\n","    start_res = residues[0].id[1]\n","    end_res = residues[-1].id[1]\n","\n","    with open(ali_path) as f:\n","        lines = f.readlines()\n","\n","    for i, line in enumerate(lines):\n","        if line.startswith(\"structureX:\"):\n","            lines[i] = f\"structureX:4bgq:{start_res}:{chain_id}:{end_res}:{chain_id}::::\\n\"\n","        elif line.startswith(\"sequence:\"):\n","            lines[i] = f\"sequence:target:{start_res}:{chain_id}:{end_res}:{chain_id}::::\\n\"\n","\n","    with open(ali_path, \"w\") as f:\n","        f.writelines(lines)\n","\n","    print(f\"✅ Fixed alignment headers with start={start_res}, end={end_res}, chain={chain_id}\")\n","    return start_res, end_res, chain_id\n","\n","fix_alignment_ali(\"4bgq_clean.pdb\", \"alignment.ali\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"YG7oNh7lS3_J","executionInfo":{"status":"ok","timestamp":1744068755895,"user_tz":240,"elapsed":134,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"a51cc415-2f53-453e-a5f2-000823788450"},"execution_count":6,"outputs":[{"output_type":"stream","name":"stdout","text":["✅ Fixed alignment headers with start=9, end=302, chain=A\n"]},{"output_type":"execute_result","data":{"text/plain":["(9, 302, 'A')"]},"metadata":{},"execution_count":6}]},{"cell_type":"code","source":["from Bio.PDB import PDBParser\n","from Bio.Data.IUPACData import protein_letters_3to1\n","import os\n","\n","def three_to_one(resname):\n","    return protein_letters_3to1.get(resname.capitalize(), 'X')\n","\n","# Extract sequence from cleaned PDB\n","pdb_parser = PDBParser(QUIET=True)\n","structure = pdb_parser.get_structure(\"4bgq\", \"4bgq_clean.pdb\")\n","chain = next(structure[0].get_chains())\n","\n","residues = [r for r in chain.get_residues() if r.id[0] == ' ']\n","pdb_seq = \"\".join([three_to_one(r.get_resname()) for r in residues])\n","\n","start_res = residues[0].id[1]\n","end_res = residues[-1].id[1]\n","chain_id = chain.id\n","\n","# Write PIR alignment\n","with open(\"alignment.ali\", \"w\") as f:\n","    f.write(f\"\"\">P1;4bgq\n","structureX:4bgq:{start_res}:{chain_id}:{end_res}:{chain_id}::::\n","{pdb_seq}*\n",">P1;target\n","sequence:target:{start_res}:{chain_id}:{end_res}:{chain_id}::::\n","{pdb_seq}*\n","\"\"\")\n","\n","print(f\"✅ Alignment written for {len(pdb_seq)} residues from {start_res} to {end_res} on chain {chain_id}\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"yBxQDHvKTc3r","executionInfo":{"status":"ok","timestamp":1744068756097,"user_tz":240,"elapsed":199,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"de051f59-f220-4842-83e5-72d97c84bfdb"},"execution_count":7,"outputs":[{"output_type":"stream","name":"stdout","text":["✅ Alignment written for 276 residues from 9 to 302 on chain A\n"]}]},{"cell_type":"code","source":["modeller_script = \"\"\"\n","from modeller import *\n","from modeller.automodel import *\n","\n","log.verbose()\n","env = environ()\n","env.io.hetatm = True\n","env.io.atom_files_directory = ['.']\n","\n","a = automodel(env,\n","              alnfile='alignment.ali',\n","              knowns='4bgq',\n","              sequence='target',\n","              assess_methods=(assess.DOPE, assess.GA341))\n","a.starting_model = 1\n","a.ending_model = 1\n","a.make()\n","\"\"\"\n","\n","with open(\"rebuild_missing_residues.py\", \"w\") as f:\n","    f.write(modeller_script)\n","\n","print(\"✅ Modeller script written. Running it...\")\n","!mod10.4 rebuild_missing_residues.py\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"w0TeCMy_TvO0","executionInfo":{"status":"ok","timestamp":1744068786761,"user_tz":240,"elapsed":30662,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"447eb6c4-3033-4745-954e-d8a64972394a"},"execution_count":8,"outputs":[{"output_type":"stream","name":"stdout","text":["✅ Modeller script written. Running it...\n","'import site' failed; use -v for traceback\n"]}]},{"cell_type":"code","source":["!pdb_reres -1 4bgq.pdb > 4bgq_renumbered.pdb"],"metadata":{"id":"eHJE12s-mpp6","executionInfo":{"status":"ok","timestamp":1744068925408,"user_tz":240,"elapsed":199,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}}},"execution_count":11,"outputs":[]}]}