{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyPtG5seVznM9yWaBkcpZtYJ"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# 00 Installation"],"metadata":{"id":"0vQTd_KqDCzi"}},{"cell_type":"markdown","source":["## 0.1 Install biopython, pdb-tools, py3dmol"],"metadata":{"id":"xZhjy7PwNlSG"}},{"cell_type":"code","source":["#Installing biopython using pip\n","!pip install biopython\n","\n","# Install pdb-tools if not already installed:\n","!pip install pdb-tools\n","\n","#Installing py3Dmol using pip\n","!pip install py3Dmol\n","\n","#And importing the py3Dmol module\n","import py3Dmol"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"POn8MlJfDJZx","executionInfo":{"status":"ok","timestamp":1744062823706,"user_tz":240,"elapsed":23195,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"55d12ec6-4955-4ea0-c504-43f2377d878b"},"execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting biopython\n","  Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (2.0.2)\n","Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m13.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hInstalling collected packages: biopython\n","Successfully installed biopython-1.85\n","Collecting pdb-tools\n","  Downloading pdb_tools-2.5.0-py3-none-any.whl.metadata (6.6 kB)\n","Downloading pdb_tools-2.5.0-py3-none-any.whl (207 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m207.3/207.3 kB\u001b[0m \u001b[31m3.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hInstalling collected packages: pdb-tools\n","Successfully installed pdb-tools-2.5.0\n","Collecting py3Dmol\n","  Downloading py3Dmol-2.4.2-py2.py3-none-any.whl.metadata (1.9 kB)\n","Downloading py3Dmol-2.4.2-py2.py3-none-any.whl (7.0 kB)\n","Installing collected packages: py3Dmol\n","Successfully installed py3Dmol-2.4.2\n"]}]},{"cell_type":"markdown","source":["## 0.2 Install Modeller 10.4"],"metadata":{"id":"w1aazI1QDO07"}},{"cell_type":"code","source":["# Step 1: Download and extract Modeller tarball\n","!wget https://salilab.org/modeller/10.4/modeller-10.4.tar.gz\n","!tar -zxf modeller-10.4.tar.gz\n","!echo \"Modeller extraction completed\"\n","\n","# Step 2: Install Modeller with your license key\n","# Change directory into the modeller folder\n","%cd modeller-10.4\n","\n","# Create the configuration file required for installation.\n","# This file contains:\n","#   1. The number 2 (for minimal setup).\n","#   2. The installation directory where Modeller will be installed.\n","#   3. Your Modeller license key.\n","with open('modeller_config', 'w') as f:\n","    f.write(\"2\\n\")\n","    f.write(\"/content/compiled/MODELLER\\n\")\n","    # Replace the following line with your actual license key if needed.\n","    f.write(\"MODELIRANJE\\n\")\n","\n","# Run the installer using the configuration file.\n","!./Install < modeller_config\n","!echo \"Modeller set up completed\"\n","\n","# Change back to the base directory.\n","%cd /content/\n","\n","# Step 3: Create a symbolic link to the mod10.4 executable for easy access\n","%cd modeller-10.4\n","!ln -sf /content/compiled/MODELLER/bin/mod10.4 /usr/bin/\n","%cd /content/\n","\n","# Step 4: Verify that the Modeller command-line tool works.\n","!mod10.4 | awk 'NR==1{if($1==\"usage:\") print \"Modeller successfully installed\"; else print \"Something went wrong\"}'\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Z_u_O38kDMpn","executionInfo":{"status":"ok","timestamp":1744062830259,"user_tz":240,"elapsed":6545,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"12a8c85e-ccb7-436b-d879-abee9d3d4ac7"},"execution_count":2,"outputs":[{"output_type":"stream","name":"stdout","text":["--2025-04-07 21:53:43--  https://salilab.org/modeller/10.4/modeller-10.4.tar.gz\n","Resolving salilab.org (salilab.org)... 169.230.79.19\n","Connecting to salilab.org (salilab.org)|169.230.79.19|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 38244158 (36M) [application/x-gzip]\n","Saving to: ‘modeller-10.4.tar.gz’\n","\n","modeller-10.4.tar.g 100%[===================>]  36.47M  23.1MB/s    in 1.6s    \n","\n","2025-04-07 21:53:45 (23.1 MB/s) - ‘modeller-10.4.tar.gz’ saved [38244158/38244158]\n","\n","Modeller extraction completed\n","/content/modeller-10.4\n","\u001b[H\u001b[2JInstallation of MODELLER 10.4\n","\n","This script will install MODELLER 10.4 into a specified directory\n","for which you have read/write permissions.\n","\n","To accept the default answers indicated in [...], press <Enter> only.\n","\n","------------------------------------------------------------------------\n","\n","The currently supported architectures are as follows:\n","\n","   1) Linux x86 PC (e.g. RedHat, SuSe).\n","   2) x86_64 (Opteron/EM64T) box (Linux).\n","   3) Alternative x86 Linux binary (e.g. for FreeBSD).\n","   4) Linux on 32-bit ARM (e.g. for Raspberry Pi).\n","   5) Linux on 64-bit ARM (e.g. for Raspberry Pi 3 or later).\n","\n","Select the type of your computer from the list above [2]: \n","\n","Full directory name in which to install MODELLER 10.4\n","[/root/bin/modeller10.4]: There is no such directory; it will be created.\n","\n","\n","License key, obtained from our academic license server at\n","https://salilab.org/modeller/registration.html: \n","------------------------------------------------------------------------\n","\n","The following settings will be used: \n","\n","The type of this machine          : x86_64-intel8\n","Destination installation directory: /content/compiled/MODELLER\n","License key                       : MODELIRANJE\n","\n","If they are incorrect, abort installation (ctrl-c) and run Install again.\n","\n","Press <Enter> to begin the installation: \n","------------------------------------------------------------------------\n","\n","Installing Modeller files ...\n","\n","------------------------------------------------------------------------\n","\n","Creating Modeller startup scripts ...\n","Configuring install ...\n","Installation complete.\n","Press <Enter> to continue: \n","------------------------------------------------------------------------\n","The best way to use Modeller is as a Python module, if you have Python\n","installed (any version between 2.3 and 3.11). Do this using the script:\n","/content/compiled/MODELLER/bin/modpy.sh\n","Simply add your regular Python command line after this script, e.g.\n","/content/compiled/MODELLER/bin/modpy.sh python3 myscript.py\n","\n","Alternatively, you can skip the 'modpy.sh' prefix if you ensure that the\n","following directories containing Python modules are in your Python path:\n","/content/compiled/MODELLER/modlib\n","/content/compiled/MODELLER/lib/x86_64-intel8 (for Python 2.3 or Python 2.4)\n","/content/compiled/MODELLER/lib/x86_64-intel8/python2.5 (for Python 2.5 to 2.7)\n","/content/compiled/MODELLER/lib/x86_64-intel8/python3.0 (for Python 3.0, 3.1)\n","/content/compiled/MODELLER/lib/x86_64-intel8/python3.2 (for Python 3.2)\n","/content/compiled/MODELLER/lib/x86_64-intel8/python3.3 (for Python 3.3 to 3.11)\n","(e.g. by setting the PYTHONPATH environment variable) and that the\n","following directory containing dynamic libraries is in your dynamic\n","library search path:\n","/content/compiled/MODELLER/lib/x86_64-intel8\n","(e.g. by setting the LD_LIBRARY_PATH variable - LIBPATH on AIX systems).\n","Ask your systems administrator for help if you wish to do this.\n","------------------------------------------------------------------------\n","\n","Press <Enter> to continue: \n","------------------------------------------------------------------------\n","\n","If you don't have Python installed, you can also use the\n","/content/compiled/MODELLER/bin/mod10.4 script to run Modeller,\n","although note that some advanced functionality, such as parallel job\n","support, requires Python.\n","\n","------------------------------------------------------------------------\n","\n","If you have problems with installation, please see the INSTALLATION file.\n","\n","Documentation in HTML format can be found in the\n","/content/compiled/MODELLER/doc directory.\n","\n","A simple homology modeling test case is in examples/automodel/.\n","\n","------------------------------------------------------------------------\n","\n","Happy modeling!\n","MODELLER Team, 2022/10/28, r12463\n","\n","Modeller set up completed\n","/content\n","/content/modeller-10.4\n","/content\n","Modeller successfully installed\n"]}]},{"cell_type":"markdown","source":["## 0.3 Configure modeller"],"metadata":{"id":"_kWvavA-De_T"}},{"cell_type":"code","source":["# Install pdb-tools if not already installed:\n","# !pip install pdb-tools [alrady installed]\n","\n","# Configure environment for Modeller:\n","import os, sys\n","\n","# Set the root where Modeller was installed.\n","# (In our installation above, Modeller was installed into /content/compiled/MODELLER)\n","modeller_root = '/content/compiled/MODELLER'\n","\n","# Append the \"modlib\" directory to the Python search path:\n","modlib_path = os.path.join(modeller_root, 'modlib')\n","if modlib_path not in sys.path:\n","    sys.path.append(modlib_path)\n","    print(\"Added modlib to sys.path:\", modlib_path)\n","\n","# Find the python module folder in modeller's lib folder and append it.\n","lib_base = os.path.join(modeller_root, 'lib', 'x86_64-intel8')\n","python_mod_dir = None\n","for folder in os.listdir(lib_base):\n","    if folder.startswith(\"python\"):\n","        python_mod_dir = os.path.join(lib_base, folder)\n","        if python_mod_dir not in sys.path:\n","            sys.path.append(python_mod_dir)\n","            print(\"Added python module directory to sys.path:\", python_mod_dir)\n","        break\n","if not python_mod_dir:\n","    print(\"Error: Could not find a python module folder in\", lib_base)\n","\n","# Set LD_LIBRARY_PATH so that shared libraries can be found.\n","ld_library_path = lib_base\n","current_ld = os.environ.get('LD_LIBRARY_PATH', '')\n","os.environ['LD_LIBRARY_PATH'] = ld_library_path + (':' + current_ld if current_ld else '')\n","print(\"LD_LIBRARY_PATH set to:\", os.environ['LD_LIBRARY_PATH'])\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"tyCwDISuDfRZ","executionInfo":{"status":"ok","timestamp":1744062830295,"user_tz":240,"elapsed":35,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"11a8dcac-4b0c-49f0-8ac8-a23d8fe83f1a"},"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["Added modlib to sys.path: /content/compiled/MODELLER/modlib\n","Added python module directory to sys.path: /content/compiled/MODELLER/lib/x86_64-intel8/python3.3\n","LD_LIBRARY_PATH set to: /content/compiled/MODELLER/lib/x86_64-intel8:/usr/local/nvidia/lib:/usr/local/nvidia/lib64\n"]}]},{"cell_type":"markdown","source":["## 0.4 Add modeller to python path"],"metadata":{"id":"vXc9IpVWDzBY"}},{"cell_type":"code","source":["import sys\n","import os\n","\n","# Add Modeller Python paths\n","sys.path.append('/content/compiled/MODELLER/modlib')\n","sys.path.append('/content/compiled/MODELLER/lib/x86_64-intel8/python3.3')\n","\n","# Set environment variable to help Python find shared libraries\n","os.environ['LD_LIBRARY_PATH'] = '/content/compiled/MODELLER/lib/x86_64-intel8'\n","\n","# Reload dynamic libraries (for some Colab environments)\n","!ldconfig /content/compiled/MODELLER/lib/x86_64-intel8\n","\n","# Try importing Modeller again\n","try:\n","    from modeller import *\n","    from modeller.scripts import complete_pdb\n","    print(\"✅ Modeller Python module imported successfully!\")\n","except Exception as e:\n","    print(\"❌ Still failed to import Modeller:\", e)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"PIu11oJNDyt9","executionInfo":{"status":"ok","timestamp":1744062834730,"user_tz":240,"elapsed":4446,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"00fed7dd-59eb-41a0-8f3b-ef6fd90d5045"},"execution_count":4,"outputs":[{"output_type":"stream","name":"stdout","text":["/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n","\n","/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n","\n","✅ Modeller Python module imported successfully!\n"]}]},{"cell_type":"markdown","source":["# 01 Make working directory"],"metadata":{"id":"VaLcR3AiHLpE"}},{"cell_type":"code","source":["!pwd\n","!mkdir /content/4bgq_fix\n","%cd /content/4bgq_fix\n","!pwd"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"fFMGsDE9HXUy","executionInfo":{"status":"ok","timestamp":1744062835089,"user_tz":240,"elapsed":310,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"92f10795-ab9e-40d5-a658-676190c14387"},"execution_count":5,"outputs":[{"output_type":"stream","name":"stdout","text":["/content\n","/content/4bgq_fix\n","/content/4bgq_fix\n"]}]},{"cell_type":"markdown","source":["# 02 Download PDB and UniProt"],"metadata":{"id":"RV6DuHp6D-Lc"}},{"cell_type":"code","source":["from Bio.PDB import PDBParser\n","from Bio import SeqIO, Entrez\n","import requests\n","\n","# Download 4bgq PDB file\n","pdb_id = \"4bgq\"\n","url = f\"https://files.rcsb.org/download/{pdb_id}.pdb\"\n","pdb_filename = f\"{pdb_id}.pdb\"\n","with open(pdb_filename, \"w\") as f:\n","    f.write(requests.get(url).text)\n","\n","# Parse structure\n","parser = PDBParser(QUIET=True)\n","structure = parser.get_structure(pdb_id, pdb_filename)\n","\n","# List HETATM and water\n","hetatm_residues = []\n","water_residues = []\n","for model in structure:\n","    for chain in model:\n","        for res in chain:\n","            if res.id[0].startswith(\"H\") and res.resname != \"HOH\":\n","                hetatm_residues.append(res)\n","            if res.resname == \"HOH\":\n","                water_residues.append(res)\n","\n","print(f\"Total HETATM residues: {len(hetatm_residues)}\")\n","print(f\"Total water molecules: {len(water_residues)}\")\n","\n","# Check for missing residues\n","print(\"\\n🧩 Residues in chain A:\")\n","for res in structure[0]['A']:\n","    print(f\"{res.resname} {res.id[1]}\")\n","\n","# Get Uniprot sequence for O76039\n","Entrez.email = \"your_email@example.com\"\n","handle = Entrez.efetch(db=\"protein\", id=\"O76039\", rettype=\"fasta\", retmode=\"text\")\n","seq_record = SeqIO.read(handle, \"fasta\")\n","handle.close()\n","print(f\"\\nUniprot sequence length: {len(seq_record.seq)}\")\n","print(f\"Uniprot sequence:\\n{seq_record.seq}\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"RN3KikEQD_cC","executionInfo":{"status":"ok","timestamp":1744062872717,"user_tz":240,"elapsed":2256,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"821bddfd-5cc4-4fa5-e4a1-92647a4f5fbc"},"execution_count":6,"outputs":[{"output_type":"stream","name":"stdout","text":["Total HETATM residues: 6\n","Total water molecules: 152\n","\n","🧩 Residues in chain A:\n","VAL 9\n","MET 10\n","ASN 11\n","LYS 12\n","PHE 13\n","GLU 14\n","ILE 15\n","LEU 16\n","GLY 17\n","VAL 18\n","VAL 19\n","GLY 20\n","GLU 21\n","GLY 22\n","ALA 23\n","TYR 24\n","GLY 25\n","VAL 26\n","VAL 27\n","LEU 28\n","LYS 29\n","CYS 30\n","ARG 31\n","HIS 32\n","LYS 33\n","GLU 34\n","THR 35\n","HIS 36\n","GLU 37\n","ILE 38\n","VAL 39\n","ALA 40\n","ILE 41\n","LYS 42\n","LYS 43\n","PHE 44\n","LYS 45\n","VAL 53\n","LYS 54\n","GLU 55\n","THR 56\n","THR 57\n","LEU 58\n","ARG 59\n","GLU 60\n","LEU 61\n","LYS 62\n","MET 63\n","LEU 64\n","ARG 65\n","THR 66\n","LEU 67\n","LYS 68\n","GLN 69\n","GLU 70\n","ASN 71\n","ILE 72\n","VAL 73\n","GLU 74\n","LEU 75\n","LYS 76\n","GLU 77\n","ALA 78\n","PHE 79\n","ARG 80\n","ARG 81\n","ARG 82\n","GLY 83\n","LYS 84\n","LEU 85\n","TYR 86\n","LEU 87\n","VAL 88\n","PHE 89\n","GLU 90\n","TYR 91\n","VAL 92\n","GLU 93\n","LYS 94\n","ASN 95\n","MET 96\n","LEU 97\n","GLU 98\n","LEU 99\n","LEU 100\n","GLU 101\n","GLU 102\n","MET 103\n","PRO 104\n","ASN 105\n","GLY 106\n","VAL 107\n","PRO 108\n","PRO 109\n","GLU 110\n","LYS 111\n","VAL 112\n","LYS 113\n","SER 114\n","TYR 115\n","ILE 116\n","TYR 117\n","GLN 118\n","LEU 119\n","ILE 120\n","LYS 121\n","ALA 122\n","ILE 123\n","HIS 124\n","TRP 125\n","CYS 126\n","HIS 127\n","LYS 128\n","ASN 129\n","ASP 130\n","ILE 131\n","VAL 132\n","HIS 133\n","ARG 134\n","ASP 135\n","ILE 136\n","LYS 137\n","PRO 138\n","GLU 139\n","ASN 140\n","LEU 141\n","LEU 142\n","ILE 143\n","SER 144\n","HIS 145\n","ASN 146\n","ASP 147\n","VAL 148\n","LEU 149\n","LYS 150\n","LEU 151\n","CYS 152\n","ASP 153\n","PHE 154\n","GLY 155\n","PHE 156\n","ALA 157\n","ARG 158\n","ASN 159\n","LEU 160\n","SER 161\n","GLU 162\n","THR 174\n","ARG 175\n","TRP 176\n","TYR 177\n","ARG 178\n","SER 179\n","PRO 180\n","GLU 181\n","LEU 182\n","LEU 183\n","LEU 184\n","GLY 185\n","ALA 186\n","PRO 187\n","TYR 188\n","GLY 189\n","LYS 190\n","SER 191\n","VAL 192\n","ASP 193\n","MET 194\n","TRP 195\n","SER 196\n","VAL 197\n","GLY 198\n","CYS 199\n","ILE 200\n","LEU 201\n","GLY 202\n","GLU 203\n","LEU 204\n","SER 205\n","ASP 206\n","GLY 207\n","GLN 208\n","PRO 209\n","LEU 210\n","PHE 211\n","PRO 212\n","GLY 213\n","GLU 214\n","SER 215\n","GLU 216\n","ILE 217\n","ASP 218\n","GLN 219\n","LEU 220\n","PHE 221\n","THR 222\n","ILE 223\n","GLN 224\n","LYS 225\n","VAL 226\n","LEU 227\n","GLY 228\n","PRO 229\n","LEU 230\n","PRO 231\n","SER 232\n","GLU 233\n","GLN 234\n","MET 235\n","LYS 236\n","LEU 237\n","PHE 238\n","TYR 239\n","SER 240\n","ASN 241\n","PRO 242\n","ARG 243\n","PHE 244\n","HIS 245\n","GLY 246\n","LEU 247\n","ARG 248\n","PHE 249\n","PRO 250\n","ALA 251\n","VAL 252\n","ASN 253\n","HIS 254\n","PRO 255\n","GLN 256\n","SER 257\n","LEU 258\n","GLU 259\n","ARG 260\n","ARG 261\n","TYR 262\n","LEU 263\n","GLY 264\n","ILE 265\n","LEU 266\n","ASN 267\n","SER 268\n","VAL 269\n","LEU 270\n","LEU 271\n","ASP 272\n","LEU 273\n","MET 274\n","LYS 275\n","ASN 276\n","LEU 277\n","LEU 278\n","LYS 279\n","LEU 280\n","ASP 281\n","PRO 282\n","ALA 283\n","ASP 284\n","ARG 285\n","TYR 286\n","LEU 287\n","THR 288\n","GLU 289\n","GLN 290\n","CYS 291\n","LEU 292\n","ASN 293\n","HIS 294\n","PRO 295\n","THR 296\n","PHE 297\n","GLN 298\n","THR 299\n","GLN 300\n","ARG 301\n","LEU 302\n","NA 1003\n","38R 1000\n","EDO 1001\n","ACT 1002\n","ACT 1004\n","EDO 1005\n","HOH 2001\n","HOH 2002\n","HOH 2003\n","HOH 2004\n","HOH 2005\n","HOH 2006\n","HOH 2007\n","HOH 2008\n","HOH 2009\n","HOH 2010\n","HOH 2011\n","HOH 2012\n","HOH 2013\n","HOH 2014\n","HOH 2015\n","HOH 2016\n","HOH 2017\n","HOH 2018\n","HOH 2019\n","HOH 2020\n","HOH 2021\n","HOH 2022\n","HOH 2023\n","HOH 2024\n","HOH 2025\n","HOH 2026\n","HOH 2027\n","HOH 2028\n","HOH 2029\n","HOH 2030\n","HOH 2031\n","HOH 2032\n","HOH 2033\n","HOH 2034\n","HOH 2035\n","HOH 2036\n","HOH 2037\n","HOH 2038\n","HOH 2039\n","HOH 2040\n","HOH 2041\n","HOH 2042\n","HOH 2043\n","HOH 2044\n","HOH 2045\n","HOH 2046\n","HOH 2047\n","HOH 2048\n","HOH 2049\n","HOH 2050\n","HOH 2051\n","HOH 2052\n","HOH 2053\n","HOH 2054\n","HOH 2055\n","HOH 2056\n","HOH 2057\n","HOH 2058\n","HOH 2059\n","HOH 2060\n","HOH 2061\n","HOH 2062\n","HOH 2063\n","HOH 2064\n","HOH 2065\n","HOH 2066\n","HOH 2067\n","HOH 2068\n","HOH 2069\n","HOH 2070\n","HOH 2071\n","HOH 2072\n","HOH 2073\n","HOH 2074\n","HOH 2075\n","HOH 2076\n","HOH 2077\n","HOH 2078\n","HOH 2079\n","HOH 2080\n","HOH 2081\n","HOH 2082\n","HOH 2083\n","HOH 2084\n","HOH 2085\n","HOH 2086\n","HOH 2087\n","HOH 2088\n","HOH 2089\n","HOH 2090\n","HOH 2091\n","HOH 2092\n","HOH 2093\n","HOH 2094\n","HOH 2095\n","HOH 2096\n","HOH 2097\n","HOH 2098\n","HOH 2099\n","HOH 2100\n","HOH 2101\n","HOH 2102\n","HOH 2103\n","HOH 2104\n","HOH 2105\n","HOH 2106\n","HOH 2107\n","HOH 2108\n","HOH 2109\n","HOH 2110\n","HOH 2111\n","HOH 2112\n","HOH 2113\n","HOH 2114\n","HOH 2115\n","HOH 2116\n","HOH 2117\n","HOH 2118\n","HOH 2119\n","HOH 2120\n","HOH 2121\n","HOH 2122\n","HOH 2123\n","HOH 2124\n","HOH 2125\n","HOH 2126\n","HOH 2127\n","HOH 2128\n","HOH 2129\n","HOH 2130\n","HOH 2131\n","HOH 2132\n","HOH 2133\n","HOH 2134\n","HOH 2135\n","HOH 2136\n","HOH 2137\n","HOH 2138\n","HOH 2139\n","HOH 2140\n","HOH 2141\n","HOH 2142\n","HOH 2143\n","HOH 2144\n","HOH 2145\n","HOH 2146\n","HOH 2147\n","HOH 2148\n","HOH 2149\n","HOH 2150\n","HOH 2151\n","HOH 2152\n","\n","Uniprot sequence length: 960\n","Uniprot sequence:\n","MKIPNIGNVMNKFEILGVVGEGAYGVVLKCRHKETHEIVAIKKFKDSEENEEVKETTLRELKMLRTLKQENIVELKEAFRRRGKLYLVFEYVEKNMLELLEEMPNGVPPEKVKSYIYQLIKAIHWCHKNDIVHRDIKPENLLISHNDVLKLCDFGFARNLSEGNNANYTEYVATRWYRSPELLLGAPYGKSVDMWSVGCILGELSDGQPLFPGESEIDQLFTIQKVLGPLPSEQMKLFYSNPRFHGLRFPAVNHPQSLERRYLGILNSVLLDLMKNLLKLDPADRYLTEQCLNHPTFQTQRLLDRSPSRSAKRKPYHVESSTLSNRNQAGKSTALQSHHRSNSKDIQNLSVGLPRADEGLPANESFLNGNLAGASLSPLHTKTYQASSQPGSTSKDLTNNNIPHLLSPKEAKSKTEFDFNIDPKPSEGPGTKYLKSNSRSQQNRHSFMESSQSKAGTLQPNEKQSRHSYIDTIPQSSRSPSYRTKAKSHGALSDSKSVSNLSEARAQIAEPSTSRYFPSSCLDLNSPTSPTPTRHSDTRTLLSPSGRNNRNEGTLDSRRTTTRHSKTMEELKLPEHMDSSHSHSLSAPHESFSYGLGYTSPFSSQQRPHRHSMYVTRDKVRAKGLDGSLSIGQGMAARANSLQLLSPQPGEQLPPEMTVARSSVKETSREGTSSFHTRQKSEGGVYHDPHSDDGTAPKENRHLYNDPVPRRVGSFYRVPSPRPDNSFHENNVSTRVSSLPSESSSGTNHSKRQPAFDPWKSPENISHSEQLKEKEKQGFFRSMKKKKKKSQTVPNSDSPDLLTLQKSIHSASTPSSRPKEWRPEKISDLQTQSQPLKSLRKLLHLSSASNHPASSDPRFQPLTAQQTKNSFSEIRIHPLSQASGGSSNIRQEPAPKGRPALQLPGQMDPGWHVSSVTRSATEGPSYSEQLGAKSGPNGHPYNRTNRSRMPNLNDLKETAL\n"]}]},{"cell_type":"markdown","source":["# 03 CLEAN PDB"],"metadata":{"id":"97-sqhWfETxZ"}},{"cell_type":"code","source":["!pdb_selchain -A 4bgq.pdb | pdb_delhetatm > 4bgq_clean.pdb"],"metadata":{"id":"QbqDbnr1ETjZ","executionInfo":{"status":"ok","timestamp":1744062894828,"user_tz":240,"elapsed":323,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}}},"execution_count":9,"outputs":[]},{"cell_type":"markdown","source":["# 04 FIX MISSING RESIDUE"],"metadata":{"id":"GWfx5ilJFH3-"}},{"cell_type":"code","source":["from Bio import SeqIO, pairwise2\n","from modeller import environ, automodel, log\n","import os\n","import requests\n","\n","# === Setup ===\n","uniprot_id = \"O76039\"\n","pdb_file = \"4bgq_clean.pdb\"\n","alignment_file = \"alignment.ali\"\n","modeller_script = \"model_build.py\"\n","code = \"4bgq\"\n","chain = \"A\"\n","uniprot_fasta = f\"{uniprot_id}.fasta\"\n","\n","# === Step 1: Download UniProt sequence ===\n","def download_uniprot_fasta(uniprot_id):\n","    url = f\"https://www.uniprot.org/uniprot/{uniprot_id}.fasta\"\n","    response = requests.get(url)\n","    if response.status_code == 200:\n","        with open(uniprot_fasta, \"w\") as f:\n","            f.write(response.text)\n","        print(f\"✅ Downloaded UniProt FASTA: {uniprot_fasta}\")\n","    else:\n","        raise Exception(f\"❌ Failed to fetch UniProt sequence for {uniprot_id}\")\n","\n","# === Step 2: Extract sequences ===\n","def get_uniprot_seq(filename):\n","    return str(next(SeqIO.parse(filename, \"fasta\")).seq)\n","\n","def get_pdb_seq(pdb_path):\n","    from Bio.PDB import PDBParser, Polypeptide\n","    parser = PDBParser(QUIET=True)\n","    structure = parser.get_structure(\"pdb\", pdb_path)\n","    for model in structure:\n","        for chain_ in model:\n","            if chain_.id == chain:\n","                ppb = Polypeptide.PPBuilder()\n","                seq = \"\"\n","                for pp in ppb.build_peptides(chain_):\n","                    seq += str(pp.get_sequence())\n","                return seq\n","    return \"\"\n","\n","# === Step 3: Align UniProt with PDB and make .ali file ===\n","def align_and_make_ali(uniprot_seq, pdb_seq):\n","    alignment = pairwise2.align.globalxs(uniprot_seq, pdb_seq, -10, -0.5, one_alignment_only=True)[0]\n","    aligned_uni = alignment.seqA.replace(\"-\", \".\")\n","    aligned_pdb = alignment.seqB.replace(\"-\", \"-\")\n","\n","    with open(alignment_file, \"w\") as f:\n","        f.write(f\"\"\">P1;{code}\n","structureX:{code}:1:{chain}::960:{chain}::::\n","{aligned_pdb}*\n","\n",">P1;target\n","sequence:target:1:{chain}::960:{chain}::::\n","{aligned_uni}*\n","\"\"\")\n","\n","# === Step 4: Write Modeller script ===\n","def write_modeller_script():\n","    with open(modeller_script, \"w\") as f:\n","        f.write(f\"\"\"\n","from modeller import *\n","from modeller.automodel import *\n","\n","log.verbose()\n","env = environ()\n","env.io.atom_files_directory = ['.', './']\n","a = automodel(env, alnfile='{alignment_file}',\n","              knowns='{code}', sequence='target')\n","a.starting_model = 1\n","a.ending_model = 1\n","a.make()\n","\"\"\")\n","\n","# === Step 5: Run Modeller ===\n","def run_modeller_script():\n","    os.system(f\"mod9.25 {modeller_script}\")\n","\n","# === Full Pipeline ===\n","download_uniprot_fasta(uniprot_id)\n","uniprot_seq = get_uniprot_seq(uniprot_fasta)\n","pdb_seq = get_pdb_seq(pdb_file)\n","align_and_make_ali(uniprot_seq, pdb_seq)\n","write_modeller_script()\n","run_modeller_script()\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"lNnUxVmDFI7j","executionInfo":{"status":"ok","timestamp":1744062899613,"user_tz":240,"elapsed":1046,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"021365b7-8d31-4e9d-d872-e248d18c3263"},"execution_count":10,"outputs":[{"output_type":"stream","name":"stdout","text":["✅ Downloaded UniProt FASTA: O76039.fasta\n"]}]},{"cell_type":"markdown","source":["# 05 Create the modeller script to rebuild missing residues"],"metadata":{"id":"6PvpoFnbP_0k"}},{"cell_type":"code","source":["# Step 1: Create the modeller script to rebuild missing residues\n","modeller_script = \"\"\"\n","from modeller import *\n","from modeller.automodel import *\n","\n","log.verbose()\n","env = environ()\n","\n","env.io.atom_files_directory = ['./', '/content/4bgq_fix']\n","\n","class MyModel(automodel):\n","    def special_patches(self, aln):\n","        self.rename_segments(segment_ids=['A'], chain_ids=['A'])\n","\n","a = MyModel(env,\n","            alnfile  = '/content/4bgq_fix/alignment.ali',\n","            knowns   = '4bgq',\n","            sequence = 'target')\n","a.starting_model = 1\n","a.ending_model   = 1\n","a.make()\n","\"\"\"\n","\n","# Step 2: Write the script to a file\n","with open(\"/content/4bgq_fix/rebuild_missing_residues.py\", \"w\") as f:\n","    f.write(modeller_script)\n","\n","# Step 3: Run the modeller script\n","!mod10.4 /content/4bgq_fix/rebuild_missing_residues.py\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"9SiuEcU8FeJv","executionInfo":{"status":"ok","timestamp":1744062919902,"user_tz":240,"elapsed":305,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"eb10deda-a6cf-48d5-ce81-47593434094d"},"execution_count":11,"outputs":[{"output_type":"stream","name":"stdout","text":["'import site' failed; use -v for traceback\n","Traceback (most recent call last):\n","  File \"/content/4bgq_fix/rebuild_missing_residues.py\", line 20, in ?\n","    a.make()\n","  File \"/content/compiled/MODELLER/modlib/modeller/automodel/automodel.py\", line 141, in make\n","    self.homcsr(exit_stage)\n","  File \"/content/compiled/MODELLER/modlib/modeller/automodel/automodel.py\", line 612, in homcsr\n","    aln = self.read_alignment()\n","  File \"/content/compiled/MODELLER/modlib/modeller/automodel/automodel.py\", line 573, in read_alignment\n","    aln.append(file=self.alnfile, align_codes=codes)\n","  File \"/content/compiled/MODELLER/modlib/modeller/alignment.py\", line 82, in append\n","    allow_alternates)\n","_modeller.FileFormatError: parse_pir__E> Invalid PIR file header line: structureX:4bgq:1:A::960:A::::\n","There should be 10 fields separated by colons, :\n","This line actually contains 11 fields.\n","\n"]}]},{"cell_type":"markdown","source":["# 06 Fix the alignment.ali file in-place\n"],"metadata":{"id":"Tq72vVoAQxug"}},{"cell_type":"code","source":["# Fix the alignment.ali file in-place\n","with open(\"/content/4bgq_fix/alignment.ali\", \"r\") as f:\n","    lines = f.readlines()\n","\n","# Auto-correct the structure line if it's malformed\n","for i, line in enumerate(lines):\n","    if line.startswith(\"structureX:\"):\n","        parts = line.strip().split(\":\")\n","        if len(parts) > 10:\n","            # Fix to exactly 10 fields\n","            lines[i] = \":\".join(parts[:10]) + '\\n'\n","\n","with open(\"/content/4bgq_fix/alignment.ali\", \"w\") as f:\n","    f.writelines(lines)\n"],"metadata":{"id":"SFwgNr1DFg-S","executionInfo":{"status":"ok","timestamp":1744062929599,"user_tz":240,"elapsed":2,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}}},"execution_count":12,"outputs":[]},{"cell_type":"markdown","source":["# 07"],"metadata":{"id":"MfMguLQAQ25S"}},{"cell_type":"code","source":["from Bio.PDB import PDBParser\n","\n","def fix_alignment_header(pdb_path, ali_path):\n","    parser = PDBParser(QUIET=True)\n","    structure = parser.get_structure(\"4bgq\", pdb_path)\n","\n","    # Extract first model, first chain\n","    model = structure[0]\n","    chain = next(model.get_chains())\n","    chain_id = chain.id\n","\n","    residues = [res for res in chain.get_residues() if res.id[0] == ' ']\n","    start_res = residues[0].id[1]\n","    end_res = residues[-1].id[1]\n","\n","    # Fix the alignment.ali file\n","    with open(ali_path) as f:\n","        lines = f.readlines()\n","\n","    for i, line in enumerate(lines):\n","        if line.startswith(\"structureX:4bgq:\"):\n","            # Replace with corrected header\n","            lines[i] = f\"structureX:4bgq:{start_res}:{chain_id}:{start_res}:{chain_id}::::\\n\"\n","\n","    with open(ali_path, \"w\") as f:\n","        f.writelines(lines)\n","\n","    return start_res, end_res, chain_id\n","\n","# Apply fix\n","pdb_file = \"/content/4bgq_fix/4bgq_clean.pdb\"\n","ali_file = \"/content/4bgq_fix/alignment.ali\"\n","start_res, end_res, chain_id = fix_alignment_header(pdb_file, ali_file)\n","\n","print(f\"✅ alignment.ali header fixed with: start={start_res}, end={end_res}, chain={chain_id}\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"BeGlQBrWFi7F","executionInfo":{"status":"ok","timestamp":1744062934204,"user_tz":240,"elapsed":42,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"f2fc567b-b9f4-4092-c00b-a6e0c3dd4f6b"},"execution_count":13,"outputs":[{"output_type":"stream","name":"stdout","text":["✅ alignment.ali header fixed with: start=9, end=302, chain=A\n"]}]},{"cell_type":"markdown","source":["# ✅ 08 Full Automation Script:\n","Parse the cleaned PDB to extract correct start_res, end_res, and chain.\n","\n","Fix both structureX: and sequence: lines in alignment.ali.\n","\n","Run Modeller automatically after that."],"metadata":{"id":"Zt_3t-fGFn0L"}},{"cell_type":"code","source":["from Bio.PDB import PDBParser\n","import os\n","\n","# === Step 1: Fix alignment.ali header based on PDB ===\n","def fix_alignment_ali(pdb_path, ali_path):\n","    parser = PDBParser(QUIET=True)\n","    structure = parser.get_structure(\"4bgq\", pdb_path)\n","    model = structure[0]\n","    chain = next(model.get_chains())\n","    chain_id = chain.id\n","\n","    residues = [res for res in chain.get_residues() if res.id[0] == ' ']\n","    start_res = residues[0].id[1]\n","    end_res = residues[-1].id[1]\n","\n","    # Fix the alignment file\n","    with open(ali_path) as f:\n","        lines = f.readlines()\n","\n","    for i, line in enumerate(lines):\n","        if line.startswith(\"structureX:\"):\n","            lines[i] = f\"structureX:4bgq:{start_res}:{chain_id}:{end_res}:{chain_id}::::\\n\"\n","        elif line.startswith(\"sequence:\"):\n","            lines[i] = f\"sequence:target:{start_res}:{chain_id}:{end_res}:{chain_id}::::\\n\"\n","\n","    with open(ali_path, \"w\") as f:\n","        f.writelines(lines)\n","\n","    print(f\"✅ Fixed alignment headers with start={start_res}, end={end_res}, chain={chain_id}\")\n","    return start_res, end_res, chain_id\n","\n","# === Step 2: Rebuild missing residues using Modeller ===\n","def run_modeller_script():\n","    modeller_script = \"\"\"\n","from modeller import *\n","from modeller.automodel import *\n","\n","log.verbose()\n","env = environ()\n","env.io.hetatm = True\n","env.io.atom_files_directory = ['.']\n","\n","a = automodel(env,\n","              alnfile='alignment.ali',\n","              knowns='4bgq',\n","              sequence='target',\n","              assess_methods=(assess.DOPE, assess.GA341))\n","a.starting_model = 1\n","a.ending_model = 1\n","a.make()\n","\"\"\"\n","\n","    # Write script to file\n","    script_path = \"/content/4bgq_fix/rebuild_missing_residues.py\"\n","    with open(script_path, \"w\") as f:\n","        f.write(modeller_script)\n","\n","    print(\"✅ Modeller script created. Running it now...\")\n","    os.system(f\"mod10.4 {script_path}\")\n","\n","# === Run everything ===\n","pdb_file = \"/content/4bgq_fix/4bgq_clean.pdb\"\n","ali_file = \"/content/4bgq_fix/alignment.ali\"\n","fix_alignment_ali(pdb_file, ali_file)\n","run_modeller_script()\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"BJcsRHvEFqNE","executionInfo":{"status":"ok","timestamp":1744063074658,"user_tz":240,"elapsed":136488,"user":{"displayName":"Shamrat Kumar Paul","userId":"00882863796766801425"}},"outputId":"02907130-8f28-4b6b-a153-7f1e329de028"},"execution_count":14,"outputs":[{"output_type":"stream","name":"stdout","text":["✅ Fixed alignment headers with start=9, end=302, chain=A\n","✅ Modeller script created. Running it now...\n"]}]}]}