{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a5b2e0d4-d2bb-4a95-9337-592dfa1adffa",
   "metadata": {},
   "source": [
    "## ===================================== ##"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bc2f8c5-d12a-4d27-8270-f679853ef9f3",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 01 Clinvar"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c36c2572-aa77-466b-b3ae-1d5125e74d1d",
   "metadata": {},
   "source": [
    "## 1.1 Convert Clinvar txt data to excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e8449656-8ebe-4a4f-9f7d-9fde8507ec31",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clinvar_result.txt\n"
     ]
    }
   ],
   "source": [
    "!ls 00_data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ef8f4788-5d13-4310-8ed9-db7832d12ca7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the tab-separated .txt file\n",
    "df = pd.read_csv('00_data/clinvar_result.txt', sep='\\t', low_memory=False)\n",
    "\n",
    "# Save to Excel\n",
    "df.to_excel('00_data/clinvar_result.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f3812df5-ed86-40c9-9a6a-8915f230b07c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sheet Name: Sheet1\n",
      "Column Names: Index(['Name', 'Gene(s)', 'Protein change', 'Condition(s)', 'Accession',\n",
      "       'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome',\n",
      "       'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID',\n",
      "       'Canonical SPDI', 'Variant type', 'Molecular consequence',\n",
      "       'Germline classification', 'Germline date last evaluated',\n",
      "       'Germline review status', 'Somatic clinical impact',\n",
      "       'Somatic clinical impact date last evaluated',\n",
      "       'Somatic clinical impact review status', 'Oncogenicity classification',\n",
      "       'Oncogenicity date last evaluated', 'Oncogenicity review status',\n",
      "       'Unnamed: 24'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the Excel file and get sheet names\n",
    "xls = pd.ExcelFile('00_data/clinvar_result.xlsx')\n",
    "sheet_names = xls.sheet_names  # List of sheet names\n",
    "\n",
    "# Load the first sheet (you can specify a different one)\n",
    "df_excel = pd.read_excel(xls, sheet_name=sheet_names[0])\n",
    "\n",
    "# Get column names\n",
    "column_names = df_excel.columns\n",
    "\n",
    "# Print sheet name and column names\n",
    "print(f\"Sheet Name: {sheet_names[0]}\")\n",
    "print(\"Column Names:\", column_names)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db615c2b-f03f-4969-a8ee-ab722ed1ad4c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1.2 view 'Sheet1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c72fdd76-31d1-4346-97af-ca42d9e89b5a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>Protein change</th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GRCh37Chromosome</th>\n",
       "      <th>GRCh37Location</th>\n",
       "      <th>GRCh38Chromosome</th>\n",
       "      <th>GRCh38Location</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>...</th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>Germline date last evaluated</th>\n",
       "      <th>Germline review status</th>\n",
       "      <th>Somatic clinical impact</th>\n",
       "      <th>Somatic clinical impact date last evaluated</th>\n",
       "      <th>Somatic clinical impact review status</th>\n",
       "      <th>Oncogenicity classification</th>\n",
       "      <th>Oncogenicity date last evaluated</th>\n",
       "      <th>Oncogenicity review status</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh38/hg38 Xp22.33-q28(chrX:2757837-156030895)x2</td>\n",
       "      <td>ABCB7|ABCD1|ACE2|ACE2-DT|ACOT9|ACSL4|ACTRT1|AD...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Klinefelter syndrome</td>\n",
       "      <td>VCV003242589</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>2757837 - 156030895</td>\n",
       "      <td>3242589</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Jun 20, 2024</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRCh37/hg19 Xp22.33-q28(chrX:200855-155240074)</td>\n",
       "      <td>ABCB7|ABCD1|ACE2|ACOT9|ACSL4|ACTRT1|ADGRG2|ADG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Klinefelter syndrome</td>\n",
       "      <td>VCV002506546</td>\n",
       "      <td>X</td>\n",
       "      <td>200855 - 155240074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2506546</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>May 26, 2023</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRCh37/hg19 Xp22.2-q28(chrX:11522765-155233731)x1</td>\n",
       "      <td>ABCB7|ABCD1|ACE2|ACOT9|ACSL4|ACTRT1|ADGRG2|ADG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV001707442</td>\n",
       "      <td>X</td>\n",
       "      <td>11522765 - 155233731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1707442</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRCh37/hg19 Xp22.33-q28(chrX:61545-155226048)x2</td>\n",
       "      <td>ABCB7|ABCD1|ACE2|ACOT9|ACSL4|ACTRT1|ADGRG2|ADG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Klinefelter syndrome</td>\n",
       "      <td>VCV001704345</td>\n",
       "      <td>X</td>\n",
       "      <td>61545 - 155226048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1704345</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRCh37/hg19 Xp22.33-q28(chrX:1-155270560)</td>\n",
       "      <td>ABCB7|ABCD1|ACE2|ACOT9|ACSL4|ACTRT1|ADGRG2|ADG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Turner syndrome</td>\n",
       "      <td>VCV001703580</td>\n",
       "      <td>X</td>\n",
       "      <td>1 - 155270560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1703580</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2188</th>\n",
       "      <td>GRCh38/hg38 Xp22.33-11.22(chrX:10679-52213731)x1</td>\n",
       "      <td>TIMP1|TLR7|TLR8|TLR8-AS1|TMEM47|TMSB4X|TRAPPC2...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV000148018</td>\n",
       "      <td>X</td>\n",
       "      <td>60679 - 51948998</td>\n",
       "      <td>X</td>\n",
       "      <td>10679 - 52213731</td>\n",
       "      <td>148018</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Dec 22, 2010</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2189</th>\n",
       "      <td>GRCh37/hg19 Xp22.33-q28(chrX:71267-155224766)x1</td>\n",
       "      <td>TMEM185A|TMEM187|TMEM255A|TMEM31|TMEM35A|TMEM4...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV000253495</td>\n",
       "      <td>X</td>\n",
       "      <td>71267 - 155224766</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>253495</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Jan 20, 2016</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2190</th>\n",
       "      <td>GRCh37/hg19 Xp22.33-q28(chrX:70297-155246585)x1</td>\n",
       "      <td>TMLHE|TMSB15A|TMSB15B|TMSB4X|TNMD|TRAPPC2|TREX...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV000395718</td>\n",
       "      <td>X</td>\n",
       "      <td>70297 - 155246585</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>395718</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2191</th>\n",
       "      <td>GRCh38/hg38 Xp22.33-11.22(chrX:10701-53750424)x1</td>\n",
       "      <td>USP27X|USP27X-DT|USP9X|UXT|UXT-AS1|VCX|VCX2|VC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV000152052</td>\n",
       "      <td>X</td>\n",
       "      <td>60701 - 53776922</td>\n",
       "      <td>X</td>\n",
       "      <td>10701 - 53750424</td>\n",
       "      <td>152052</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Nov 26, 2012</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2192</th>\n",
       "      <td>GRCh37/hg19 Xp22.33-11.21(chrX:168546-57683964)x1</td>\n",
       "      <td>WDR45|WNK3|WWC3|ACE2|ACOT9|ADGRG2|AKAP4|ALAS2|...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV000441856</td>\n",
       "      <td>X</td>\n",
       "      <td>168546 - 57683964</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>441856</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Jul 7, 2014</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2193 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Name  \\\n",
       "0     GRCh38/hg38 Xp22.33-q28(chrX:2757837-156030895)x2   \n",
       "1        GRCh37/hg19 Xp22.33-q28(chrX:200855-155240074)   \n",
       "2     GRCh37/hg19 Xp22.2-q28(chrX:11522765-155233731)x1   \n",
       "3       GRCh37/hg19 Xp22.33-q28(chrX:61545-155226048)x2   \n",
       "4             GRCh37/hg19 Xp22.33-q28(chrX:1-155270560)   \n",
       "...                                                 ...   \n",
       "2188   GRCh38/hg38 Xp22.33-11.22(chrX:10679-52213731)x1   \n",
       "2189    GRCh37/hg19 Xp22.33-q28(chrX:71267-155224766)x1   \n",
       "2190    GRCh37/hg19 Xp22.33-q28(chrX:70297-155246585)x1   \n",
       "2191   GRCh38/hg38 Xp22.33-11.22(chrX:10701-53750424)x1   \n",
       "2192  GRCh37/hg19 Xp22.33-11.21(chrX:168546-57683964)x1   \n",
       "\n",
       "                                                Gene(s) Protein change  \\\n",
       "0     ABCB7|ABCD1|ACE2|ACE2-DT|ACOT9|ACSL4|ACTRT1|AD...            NaN   \n",
       "1     ABCB7|ABCD1|ACE2|ACOT9|ACSL4|ACTRT1|ADGRG2|ADG...            NaN   \n",
       "2     ABCB7|ABCD1|ACE2|ACOT9|ACSL4|ACTRT1|ADGRG2|ADG...            NaN   \n",
       "3     ABCB7|ABCD1|ACE2|ACOT9|ACSL4|ACTRT1|ADGRG2|ADG...            NaN   \n",
       "4     ABCB7|ABCD1|ACE2|ACOT9|ACSL4|ACTRT1|ADGRG2|ADG...            NaN   \n",
       "...                                                 ...            ...   \n",
       "2188  TIMP1|TLR7|TLR8|TLR8-AS1|TMEM47|TMSB4X|TRAPPC2...            NaN   \n",
       "2189  TMEM185A|TMEM187|TMEM255A|TMEM31|TMEM35A|TMEM4...            NaN   \n",
       "2190  TMLHE|TMSB15A|TMSB15B|TMSB4X|TNMD|TRAPPC2|TREX...            NaN   \n",
       "2191  USP27X|USP27X-DT|USP9X|UXT|UXT-AS1|VCX|VCX2|VC...            NaN   \n",
       "2192  WDR45|WNK3|WWC3|ACE2|ACOT9|ADGRG2|AKAP4|ALAS2|...            NaN   \n",
       "\n",
       "              Condition(s)     Accession GRCh37Chromosome  \\\n",
       "0     Klinefelter syndrome  VCV003242589              NaN   \n",
       "1     Klinefelter syndrome  VCV002506546                X   \n",
       "2                See cases  VCV001707442                X   \n",
       "3     Klinefelter syndrome  VCV001704345                X   \n",
       "4          Turner syndrome  VCV001703580                X   \n",
       "...                    ...           ...              ...   \n",
       "2188             See cases  VCV000148018                X   \n",
       "2189             See cases  VCV000253495                X   \n",
       "2190             See cases  VCV000395718                X   \n",
       "2191             See cases  VCV000152052                X   \n",
       "2192             See cases  VCV000441856                X   \n",
       "\n",
       "            GRCh37Location GRCh38Chromosome       GRCh38Location  VariationID  \\\n",
       "0                      NaN                X  2757837 - 156030895      3242589   \n",
       "1       200855 - 155240074              NaN                  NaN      2506546   \n",
       "2     11522765 - 155233731              NaN                  NaN      1707442   \n",
       "3        61545 - 155226048              NaN                  NaN      1704345   \n",
       "4            1 - 155270560              NaN                  NaN      1703580   \n",
       "...                    ...              ...                  ...          ...   \n",
       "2188      60679 - 51948998                X     10679 - 52213731       148018   \n",
       "2189     71267 - 155224766              NaN                  NaN       253495   \n",
       "2190     70297 - 155246585              NaN                  NaN       395718   \n",
       "2191      60701 - 53776922                X     10701 - 53750424       152052   \n",
       "2192     168546 - 57683964              NaN                  NaN       441856   \n",
       "\n",
       "      ...  Germline classification Germline date last evaluated  \\\n",
       "0     ...               Pathogenic                 Jun 20, 2024   \n",
       "1     ...               Pathogenic                 May 26, 2023   \n",
       "2     ...               Pathogenic                          NaN   \n",
       "3     ...               Pathogenic                          NaN   \n",
       "4     ...               Pathogenic                          NaN   \n",
       "...   ...                      ...                          ...   \n",
       "2188  ...               Pathogenic                 Dec 22, 2010   \n",
       "2189  ...               Pathogenic                 Jan 20, 2016   \n",
       "2190  ...               Pathogenic                          NaN   \n",
       "2191  ...               Pathogenic                 Nov 26, 2012   \n",
       "2192  ...               Pathogenic                  Jul 7, 2014   \n",
       "\n",
       "                   Germline review status Somatic clinical impact  \\\n",
       "0     criteria provided, single submitter                     NaN   \n",
       "1     criteria provided, single submitter                     NaN   \n",
       "2          no assertion criteria provided                     NaN   \n",
       "3          no assertion criteria provided                     NaN   \n",
       "4          no assertion criteria provided                     NaN   \n",
       "...                                   ...                     ...   \n",
       "2188       no assertion criteria provided                     NaN   \n",
       "2189  criteria provided, single submitter                     NaN   \n",
       "2190       no assertion criteria provided                     NaN   \n",
       "2191       no assertion criteria provided                     NaN   \n",
       "2192       no assertion criteria provided                     NaN   \n",
       "\n",
       "     Somatic clinical impact date last evaluated  \\\n",
       "0                                            NaN   \n",
       "1                                            NaN   \n",
       "2                                            NaN   \n",
       "3                                            NaN   \n",
       "4                                            NaN   \n",
       "...                                          ...   \n",
       "2188                                         NaN   \n",
       "2189                                         NaN   \n",
       "2190                                         NaN   \n",
       "2191                                         NaN   \n",
       "2192                                         NaN   \n",
       "\n",
       "     Somatic clinical impact review status Oncogenicity classification  \\\n",
       "0                                      NaN                         NaN   \n",
       "1                                      NaN                         NaN   \n",
       "2                                      NaN                         NaN   \n",
       "3                                      NaN                         NaN   \n",
       "4                                      NaN                         NaN   \n",
       "...                                    ...                         ...   \n",
       "2188                                   NaN                         NaN   \n",
       "2189                                   NaN                         NaN   \n",
       "2190                                   NaN                         NaN   \n",
       "2191                                   NaN                         NaN   \n",
       "2192                                   NaN                         NaN   \n",
       "\n",
       "     Oncogenicity date last evaluated  Oncogenicity review status  Unnamed: 24  \n",
       "0                                 NaN                         NaN          NaN  \n",
       "1                                 NaN                         NaN          NaN  \n",
       "2                                 NaN                         NaN          NaN  \n",
       "3                                 NaN                         NaN          NaN  \n",
       "4                                 NaN                         NaN          NaN  \n",
       "...                               ...                         ...          ...  \n",
       "2188                              NaN                         NaN          NaN  \n",
       "2189                              NaN                         NaN          NaN  \n",
       "2190                              NaN                         NaN          NaN  \n",
       "2191                              NaN                         NaN          NaN  \n",
       "2192                              NaN                         NaN          NaN  \n",
       "\n",
       "[2193 rows x 25 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the Excel file\n",
    "df_excel = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='Sheet1')\n",
    "\n",
    "# Display the dataset\n",
    "display(df_excel)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "623768c0-55e0-4105-9556-867b898dd68d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1.3 show columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "4b7657e7-67a4-4b5a-addd-af044abbc1ed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column Names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gene(s)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Protein change</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Condition(s)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Accession</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GRCh37Chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GRCh37Location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GRCh38Chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GRCh38Location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>VariationID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AlleleID(s)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>dbSNP ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Canonical SPDI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Variant type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Molecular consequence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Germline classification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Germline date last evaluated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Germline review status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Somatic clinical impact</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Somatic clinical impact date last evaluated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Somatic clinical impact review status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Oncogenicity classification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Oncogenicity date last evaluated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Oncogenicity review status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Unnamed: 24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Column Names\n",
       "0                                          Name\n",
       "1                                       Gene(s)\n",
       "2                                Protein change\n",
       "3                                  Condition(s)\n",
       "4                                     Accession\n",
       "5                              GRCh37Chromosome\n",
       "6                                GRCh37Location\n",
       "7                              GRCh38Chromosome\n",
       "8                                GRCh38Location\n",
       "9                                   VariationID\n",
       "10                                  AlleleID(s)\n",
       "11                                     dbSNP ID\n",
       "12                               Canonical SPDI\n",
       "13                                 Variant type\n",
       "14                        Molecular consequence\n",
       "15                      Germline classification\n",
       "16                 Germline date last evaluated\n",
       "17                       Germline review status\n",
       "18                      Somatic clinical impact\n",
       "19  Somatic clinical impact date last evaluated\n",
       "20        Somatic clinical impact review status\n",
       "21                  Oncogenicity classification\n",
       "22             Oncogenicity date last evaluated\n",
       "23                   Oncogenicity review status\n",
       "24                                  Unnamed: 24"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the Excel file\n",
    "df_excel = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='Sheet1')\n",
    "\n",
    "# Convert column names to DataFrame for better visualization\n",
    "column_names_df = pd.DataFrame(df_excel.columns, columns=['Column Names'])\n",
    "\n",
    "# Display the column names as a table\n",
    "display(column_names_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40120298-c689-448f-be5f-a41c830364a8",
   "metadata": {},
   "source": [
    "### 1.4 Value counts: 'Sheet1': "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "fb4e4581-2b96-4a45-943e-f5d0693cda37",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant type</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>1517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Deletion</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Duplication</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>copy number loss</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>copy number gain</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Microsatellite</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Insertion</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Indel</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Complex</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Inversion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Total Variant type</td>\n",
       "      <td>2193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Variant type  Count\n",
       "0   single nucleotide variant   1517\n",
       "1                    Deletion    269\n",
       "2                 Duplication    139\n",
       "3            copy number loss    107\n",
       "4            copy number gain     78\n",
       "5              Microsatellite     46\n",
       "6                   Insertion     17\n",
       "7                       Indel     16\n",
       "8                     Complex      2\n",
       "9                   Inversion      2\n",
       "10         Total Variant type   2193"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>not provided</td>\n",
       "      <td>578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Angelman syndrome-like|Developmental and epile...</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>See cases</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>Early infantile epileptic encephalopathy with ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Infantile spasms</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Neurodevelopmental disorder</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Total Condition(s)</td>\n",
       "      <td>2193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>153 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Condition(s)  Count\n",
       "0                                         not provided    578\n",
       "1    Developmental and epileptic encephalopathy, 2|...    363\n",
       "2    Angelman syndrome-like|Developmental and epile...    247\n",
       "3                                       CDKL5 disorder    224\n",
       "4                                            See cases    131\n",
       "..                                                 ...    ...\n",
       "148  Early infantile epileptic encephalopathy with ...      1\n",
       "149  Developmental and epileptic encephalopathy, 2|...      1\n",
       "150                                   Infantile spasms      1\n",
       "151                        Neurodevelopmental disorder      1\n",
       "152                                 Total Condition(s)   2193\n",
       "\n",
       "[153 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Molecular consequence</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>missense variant</td>\n",
       "      <td>485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>missense variant|intron variant</td>\n",
       "      <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>intron variant</td>\n",
       "      <td>295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>frameshift variant</td>\n",
       "      <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>synonymous variant</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nonsense</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>synonymous variant|intron variant</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>frameshift variant|intron variant</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>nonsense|intron variant</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>splice donor variant</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>splice acceptor variant</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>intron variant|splice acceptor variant</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>intron variant|splice donor variant</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>5 prime UTR variant</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>inframe_indel</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>inframe_indel|intron variant</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>3 prime UTR variant|intron variant</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>splice acceptor variant|splice donor variant</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>stop lost|intron variant</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>frameshift variant|nonsense|intron variant</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>nonsense|inframe_insertion</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>splice donor variant|intron variant</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>missense variant|splice donor variant</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Total Molecular consequence</td>\n",
       "      <td>1909</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Molecular consequence  Count\n",
       "0                               missense variant    485\n",
       "1                missense variant|intron variant    301\n",
       "2                                 intron variant    295\n",
       "3                             frameshift variant    234\n",
       "4                             synonymous variant    177\n",
       "5                                       nonsense    101\n",
       "6              synonymous variant|intron variant     97\n",
       "7              frameshift variant|intron variant     49\n",
       "8                        nonsense|intron variant     47\n",
       "9                           splice donor variant     36\n",
       "10                       splice acceptor variant     31\n",
       "11                              inframe_deletion     10\n",
       "12        intron variant|splice acceptor variant      8\n",
       "13           intron variant|splice donor variant      7\n",
       "14                           5 prime UTR variant      6\n",
       "15                                 inframe_indel      5\n",
       "16                  inframe_indel|intron variant      5\n",
       "17            3 prime UTR variant|intron variant      4\n",
       "18  splice acceptor variant|splice donor variant      3\n",
       "19                      stop lost|intron variant      2\n",
       "20    frameshift variant|nonsense|intron variant      2\n",
       "21                    nonsense|inframe_insertion      1\n",
       "22                             inframe_insertion      1\n",
       "23           splice donor variant|intron variant      1\n",
       "24         missense variant|splice donor variant      1\n",
       "25                   Total Molecular consequence   1909"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Likely benign</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Benign</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Conflicting classifications of pathogenicity</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>not provided</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>conflicting data from submitters</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Total Germline classification</td>\n",
       "      <td>2193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Germline classification  Count\n",
       "0                                     Pathogenic    698\n",
       "1                         Uncertain significance    473\n",
       "2                                  Likely benign    471\n",
       "3                              Likely pathogenic    229\n",
       "4                                         Benign    112\n",
       "5   Conflicting classifications of pathogenicity     68\n",
       "6                   Pathogenic/Likely pathogenic     55\n",
       "7                           Benign/Likely benign     44\n",
       "8                                   not provided     42\n",
       "9               conflicting data from submitters      1\n",
       "10                 Total Germline classification   2193"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the Excel file\n",
    "df_excel = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='Sheet1')\n",
    "\n",
    "# Define columns to count\n",
    "columns_to_count = ['Variant type', 'Condition(s)', 'Molecular consequence', 'Germline classification']\n",
    "\n",
    "# Get value counts for each column and display them\n",
    "for col in columns_to_count:\n",
    "    counts_df = df_excel[col].value_counts().reset_index()\n",
    "    counts_df.columns = [col, 'Count']\n",
    "    \n",
    "    # Add total count row\n",
    "    total_row = pd.DataFrame([[f'Total {col}', counts_df['Count'].sum()]], columns=[col, 'Count'])\n",
    "    counts_df = pd.concat([counts_df, total_row], ignore_index=True)\n",
    "\n",
    "    # Display the table in IPython format\n",
    "    display(counts_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe0499e8-8f3d-4933-977c-538c7605576b",
   "metadata": {},
   "source": [
    "## 1.5 Extract 'single nucleotide variant'from 'Variant type' and save as 'Single_Nucleotide_Variant' sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "e0611ce9-b2b3-4305-af9f-d4184ebf2639",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data added as 'Single_Nucleotide_Variant' in the same Excel file.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "# Load the existing Excel file\n",
    "file_path = '00_data/clinvar_result.xlsx'\n",
    "df_excel = pd.read_excel(file_path, sheet_name='Sheet1')\n",
    "\n",
    "# Filter for 'single nucleotide variant' entries\n",
    "snv_df = df_excel[df_excel['Variant type'] == 'single nucleotide variant']\n",
    "\n",
    "# Load the workbook and write the new sheet without overwriting existing data\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl') as writer:\n",
    "    snv_df.to_excel(writer, sheet_name='Single_Nucleotide_Variant', index=False)\n",
    "\n",
    "print(\"Filtered data added as 'Single_Nucleotide_Variant' in the same Excel file.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5f4cf4d-7c4a-44b5-97ad-1d84efd8ebbb",
   "metadata": {},
   "source": [
    "### 1.6 Value counts: 'Single_Nucleotid_variant' sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "32e44537-6986-4574-8654-405ff9fa2b49",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total occurrences of 'missense' in 'Molecular consequence' column: 783\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the Excel file\n",
    "df_excel = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='Single_Nucleotide_Variant')\n",
    "\n",
    "# Filter rows containing 'missense' in 'Molecular consequence' column\n",
    "missense_counts = df_excel[df_excel['Molecular consequence'].str.contains('missense variant', na=False)].shape[0]\n",
    "\n",
    "# Display the count\n",
    "print(\"Total occurrences of 'missense' in 'Molecular consequence' column:\", missense_counts)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bd8d44e-9633-4aae-8e3c-6242ccbc366c",
   "metadata": {},
   "source": [
    "## 1.7 Extract 'missense variant' from 'Molecular Consequence' column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "b118eb94-c7af-4b44-8db6-1b697134aea5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved as 'Missense_Variant' in the same Excel file.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "# Load the existing Excel file\n",
    "file_path = '00_data/clinvar_result.xlsx'\n",
    "df_snv = pd.read_excel(file_path, sheet_name='Single_Nucleotide_Variant')\n",
    "\n",
    "# Filter for 'missense variant' entries\n",
    "missense_df = df_snv[df_snv['Molecular consequence'].str.contains('missense variant', na=False)]\n",
    "\n",
    "# Load the workbook and write the new sheet without overwriting existing data\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl') as writer:\n",
    "    missense_df.to_excel(writer, sheet_name='Missense_Variant', index=False)\n",
    "\n",
    "print(\"Filtered data saved as 'Missense_Variant' in the same Excel file.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61ea3313-abea-4db1-9f0c-d589a55b5403",
   "metadata": {},
   "source": [
    "###  1.8 Value counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "aa3b21d8-3de8-48df-9ca2-54e6ac636131",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant type</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total Variant type</td>\n",
       "      <td>783</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Variant type  Count\n",
       "0  single nucleotide variant    783\n",
       "1         Total Variant type    783"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>not provided</td>\n",
       "      <td>205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Angelman syndrome-like|Developmental and epile...</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>not provided|CDKL5 disorder|Angelman syndrome-...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Inborn genetic diseases|Developmental and epil...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>not provided|not specified|Angelman syndrome-l...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>Retinoschisis|Retinal dystrophy|Juvenile retin...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Total Condition(s)</td>\n",
       "      <td>783</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Condition(s)  Count\n",
       "0                                        not provided    205\n",
       "1   Developmental and epileptic encephalopathy, 2|...    141\n",
       "2   Angelman syndrome-like|Developmental and epile...     86\n",
       "3                                      CDKL5 disorder     84\n",
       "4       Developmental and epileptic encephalopathy, 2     33\n",
       "..                                                ...    ...\n",
       "79  not provided|CDKL5 disorder|Angelman syndrome-...      1\n",
       "80  Inborn genetic diseases|Developmental and epil...      1\n",
       "81  not provided|not specified|Angelman syndrome-l...      1\n",
       "82  Retinoschisis|Retinal dystrophy|Juvenile retin...      1\n",
       "83                                 Total Condition(s)    783\n",
       "\n",
       "[84 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Molecular consequence</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>missense variant</td>\n",
       "      <td>485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>missense variant|intron variant</td>\n",
       "      <td>297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>missense variant|splice donor variant</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Total Molecular consequence</td>\n",
       "      <td>783</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Molecular consequence  Count\n",
       "0                       missense variant    485\n",
       "1        missense variant|intron variant    297\n",
       "2  missense variant|splice donor variant      1\n",
       "3            Total Molecular consequence    783"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Likely benign</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Conflicting classifications of pathogenicity</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Benign</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>not provided</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Total Germline classification</td>\n",
       "      <td>783</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Germline classification  Count\n",
       "0                        Uncertain significance    356\n",
       "1                             Likely pathogenic    109\n",
       "2                                    Pathogenic    103\n",
       "3                                 Likely benign     72\n",
       "4  Conflicting classifications of pathogenicity     50\n",
       "5                  Pathogenic/Likely pathogenic     28\n",
       "6                                        Benign     27\n",
       "7                                  not provided     25\n",
       "8                          Benign/Likely benign     13\n",
       "9                 Total Germline classification    783"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the Excel file\n",
    "df_excel = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='Missense_Variant')\n",
    "\n",
    "# Define columns to count\n",
    "columns_to_count = ['Variant type', 'Condition(s)', 'Molecular consequence', 'Germline classification']\n",
    "\n",
    "# Get value counts for each column and display them\n",
    "for col in columns_to_count:\n",
    "    counts_df = df_excel[col].value_counts().reset_index()\n",
    "    counts_df.columns = [col, 'Count']\n",
    "    \n",
    "    # Add total count row\n",
    "    total_row = pd.DataFrame([[f'Total {col}', counts_df['Count'].sum()]], columns=[col, 'Count'])\n",
    "    counts_df = pd.concat([counts_df, total_row], ignore_index=True)\n",
    "\n",
    "    # Display the table in IPython format\n",
    "    display(counts_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec094a41-266b-4781-9de6-cbd14c9ffaa7",
   "metadata": {},
   "source": [
    "## 1.9 Extract 'CDKL5' from 'Condition(s)' column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "1d10f8c7-ccfa-4fb9-88cb-a9cb7228bb68",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved as 'CDKL5_Condition' in the same Excel file.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "# Load the existing Excel file\n",
    "file_path = '00_data/clinvar_result.xlsx'\n",
    "df_missense = pd.read_excel(file_path, sheet_name='Missense_Variant')\n",
    "\n",
    "# Filter rows containing 'CDKL5' in 'Condition(s)' column\n",
    "cdkl5_df = df_missense[df_missense['Condition(s)'].str.contains('CDKL5', na=False)]\n",
    "\n",
    "# Save the filtered data to a new sheet in the same Excel file\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl') as writer:\n",
    "    cdkl5_df.to_excel(writer, sheet_name='CDKL5_Condition', index=False)\n",
    "\n",
    "print(\"Filtered data saved as 'CDKL5_Condition' in the same Excel file.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c031d14-b645-4034-939c-1bb6b8f63d28",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 1.10 Value counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "51cb48d1-9327-45e1-b43f-f2ab78d3779f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant type</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total Variant type</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Variant type  Count\n",
       "0  single nucleotide variant    139\n",
       "1         Total Variant type    139"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>not provided|CDKL5 disorder</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Angelman syndrome-like|Developmental and epile...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDKL5-related disorder</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CDKL5 disorder|Developmental and epileptic enc...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CDKL5 disorder|Angelman syndrome-like|Developm...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CDKL5 disorder|Inborn genetic diseases</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDKL5 disorder|not provided|Developmental and ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Angelman syndrome-like|Developmental and epile...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Epileptic encephalopathy|Developmental and epi...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CDKL5 disorder|Developmental and epileptic enc...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Angelman syndrome-like|Developmental and epile...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>not specified|Developmental and epileptic ence...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CDKL5 disorder|not specified|not provided|Deve...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Inborn genetic diseases|CDKL5 disorder|not pro...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>not provided|CDKL5 disorder|Angelman syndrome-...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CDKL5-related disorder|Developmental and epile...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CDKL5-related disorder|not provided</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Angelman syndrome-like|Developmental and epile...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CDKL5 disorder|Developmental and epileptic enc...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>West syndrome|Developmental and epileptic ence...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Total Condition(s)</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Condition(s)  Count\n",
       "0                                      CDKL5 disorder     84\n",
       "1   Developmental and epileptic encephalopathy, 2|...      7\n",
       "2                         not provided|CDKL5 disorder      5\n",
       "3   Angelman syndrome-like|Developmental and epile...      5\n",
       "4                              CDKL5-related disorder      5\n",
       "5   CDKL5 disorder|Developmental and epileptic enc...      4\n",
       "6   CDKL5 disorder|Angelman syndrome-like|Developm...      4\n",
       "7   Developmental and epileptic encephalopathy, 2|...      3\n",
       "8              CDKL5 disorder|Inborn genetic diseases      2\n",
       "9   CDKL5 disorder|not provided|Developmental and ...      2\n",
       "10  Developmental and epileptic encephalopathy, 2|...      1\n",
       "11  Developmental and epileptic encephalopathy, 2|...      1\n",
       "12  Angelman syndrome-like|Developmental and epile...      1\n",
       "13  Developmental and epileptic encephalopathy, 2|...      1\n",
       "14  Epileptic encephalopathy|Developmental and epi...      1\n",
       "15  CDKL5 disorder|Developmental and epileptic enc...      1\n",
       "16  Angelman syndrome-like|Developmental and epile...      1\n",
       "17  Developmental and epileptic encephalopathy, 2|...      1\n",
       "18  Developmental and epileptic encephalopathy, 2|...      1\n",
       "19  not specified|Developmental and epileptic ence...      1\n",
       "20  CDKL5 disorder|not specified|not provided|Deve...      1\n",
       "21  Inborn genetic diseases|CDKL5 disorder|not pro...      1\n",
       "22  not provided|CDKL5 disorder|Angelman syndrome-...      1\n",
       "23  CDKL5-related disorder|Developmental and epile...      1\n",
       "24                CDKL5-related disorder|not provided      1\n",
       "25  Angelman syndrome-like|Developmental and epile...      1\n",
       "26  CDKL5 disorder|Developmental and epileptic enc...      1\n",
       "27  West syndrome|Developmental and epileptic ence...      1\n",
       "28                                 Total Condition(s)    139"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Molecular consequence</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>missense variant</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>missense variant|intron variant</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Total Molecular consequence</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Molecular consequence  Count\n",
       "0                 missense variant    120\n",
       "1  missense variant|intron variant     19\n",
       "2      Total Molecular consequence    139"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Likely benign</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Conflicting classifications of pathogenicity</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Benign</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Total Germline classification</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Germline classification  Count\n",
       "0                        Uncertain significance     45\n",
       "1                             Likely pathogenic     22\n",
       "2                                 Likely benign     19\n",
       "3  Conflicting classifications of pathogenicity     16\n",
       "4                                        Benign     16\n",
       "5                                    Pathogenic      9\n",
       "6                  Pathogenic/Likely pathogenic      8\n",
       "7                          Benign/Likely benign      4\n",
       "8                 Total Germline classification    139"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the Excel file\n",
    "df_excel = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='CDKL5_Condition')\n",
    "\n",
    "# Define columns to count\n",
    "columns_to_count = ['Variant type', 'Condition(s)', 'Molecular consequence', 'Germline classification']\n",
    "\n",
    "# Get value counts for each column and display them\n",
    "for col in columns_to_count:\n",
    "    counts_df = df_excel[col].value_counts().reset_index()\n",
    "    counts_df.columns = [col, 'Count']\n",
    "    \n",
    "    # Add total count row\n",
    "    total_row = pd.DataFrame([[f'Total {col}', counts_df['Count'].sum()]], columns=[col, 'Count'])\n",
    "    counts_df = pd.concat([counts_df, total_row], ignore_index=True)\n",
    "\n",
    "    # Display the table in IPython format\n",
    "    display(counts_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b98c7ec7-46d8-436d-8305-a3002f57f748",
   "metadata": {},
   "source": [
    "## 1.11: only 'misssense variant' discard 'intron variant'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "49b2f82f-59f7-4556-906a-db4c5a436f95",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved as 'CDKL5_missense_only' in the same Excel file.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "# Load the existing Excel file\n",
    "file_path = '00_data/clinvar_result.xlsx'\n",
    "df_cdkl5_condition = pd.read_excel(file_path, sheet_name='CDKL5_Condition')\n",
    "\n",
    "# Remove rows containing 'intron variant' in 'Molecular consequence' column\n",
    "filtered_df = df_cdkl5_condition[~df_cdkl5_condition['Molecular consequence'].str.contains('intron variant', na=False)]\n",
    "\n",
    "# Save the cleaned data to a new sheet with the correct name\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl') as writer:\n",
    "    filtered_df.to_excel(writer, sheet_name='CDKL5_missense_only', index=False)\n",
    "\n",
    "print(\"Filtered data saved as 'CDKL5_missense_only' in the same Excel file.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50880505-eeca-4412-ae4f-6c1021d750c1",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 1.12 Value count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d47ce6a7-ce08-4996-8d52-382dafa9f82c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant type</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total Variant type</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Variant type  Count\n",
       "0  single nucleotide variant    120\n",
       "1         Total Variant type    120"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>not provided|CDKL5 disorder</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Angelman syndrome-like|Developmental and epile...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDKL5 disorder|Angelman syndrome-like|Developm...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CDKL5 disorder|Developmental and epileptic enc...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CDKL5-related disorder</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CDKL5 disorder|not provided|Developmental and ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDKL5 disorder|Developmental and epileptic enc...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Angelman syndrome-like|Developmental and epile...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Epileptic encephalopathy|Developmental and epi...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Developmental and epileptic encephalopathy, 2|...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CDKL5 disorder|not specified|not provided|Deve...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Inborn genetic diseases|CDKL5 disorder|not pro...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CDKL5 disorder|Inborn genetic diseases</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>not provided|CDKL5 disorder|Angelman syndrome-...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CDKL5-related disorder|Developmental and epile...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CDKL5 disorder|Developmental and epileptic enc...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Angelman syndrome-like|Developmental and epile...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Total Condition(s)</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Condition(s)  Count\n",
       "0                                      CDKL5 disorder     74\n",
       "1   Developmental and epileptic encephalopathy, 2|...      7\n",
       "2                         not provided|CDKL5 disorder      5\n",
       "3   Angelman syndrome-like|Developmental and epile...      5\n",
       "4   CDKL5 disorder|Angelman syndrome-like|Developm...      4\n",
       "5   CDKL5 disorder|Developmental and epileptic enc...      4\n",
       "6                              CDKL5-related disorder      4\n",
       "7   Developmental and epileptic encephalopathy, 2|...      2\n",
       "8   CDKL5 disorder|not provided|Developmental and ...      2\n",
       "9   CDKL5 disorder|Developmental and epileptic enc...      1\n",
       "10  Angelman syndrome-like|Developmental and epile...      1\n",
       "11  Developmental and epileptic encephalopathy, 2|...      1\n",
       "12  Developmental and epileptic encephalopathy, 2|...      1\n",
       "13  Epileptic encephalopathy|Developmental and epi...      1\n",
       "14  Developmental and epileptic encephalopathy, 2|...      1\n",
       "15  CDKL5 disorder|not specified|not provided|Deve...      1\n",
       "16  Inborn genetic diseases|CDKL5 disorder|not pro...      1\n",
       "17             CDKL5 disorder|Inborn genetic diseases      1\n",
       "18  not provided|CDKL5 disorder|Angelman syndrome-...      1\n",
       "19  CDKL5-related disorder|Developmental and epile...      1\n",
       "20  CDKL5 disorder|Developmental and epileptic enc...      1\n",
       "21  Angelman syndrome-like|Developmental and epile...      1\n",
       "22                                 Total Condition(s)    120"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Molecular consequence</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>missense variant</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total Molecular consequence</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Molecular consequence  Count\n",
       "0             missense variant    120\n",
       "1  Total Molecular consequence    120"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Conflicting classifications of pathogenicity</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Benign</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Likely benign</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Total Germline classification</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Germline classification  Count\n",
       "0                        Uncertain significance     43\n",
       "1                             Likely pathogenic     22\n",
       "2  Conflicting classifications of pathogenicity     13\n",
       "3                                        Benign     12\n",
       "4                                 Likely benign     10\n",
       "5                                    Pathogenic      9\n",
       "6                  Pathogenic/Likely pathogenic      8\n",
       "7                          Benign/Likely benign      3\n",
       "8                 Total Germline classification    120"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the Excel file\n",
    "df_excel = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='CDKL5_missense_only')\n",
    "\n",
    "# Define columns to count\n",
    "columns_to_count = ['Variant type', 'Condition(s)', 'Molecular consequence', 'Germline classification']\n",
    "\n",
    "# Get value counts for each column and display them\n",
    "for col in columns_to_count:\n",
    "    counts_df = df_excel[col].value_counts().reset_index()\n",
    "    counts_df.columns = [col, 'Count']\n",
    "    \n",
    "    # Add total count row\n",
    "    total_row = pd.DataFrame([[f'Total {col}', counts_df['Count'].sum()]], columns=[col, 'Count'])\n",
    "    counts_df = pd.concat([counts_df, total_row], ignore_index=True)\n",
    "\n",
    "    # Display the table in IPython format\n",
    "    display(counts_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6387cf8c-4667-4cda-8be4-fb5e068199d3",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1.13 make 'Mutation' column with 'wild' 'position' 'mutant'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67f0f615-f361-442a-92b4-9d3c309bf2a7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c7b17bcf-0b95-4a42-85a9-d79389c00af1",
   "metadata": {},
   "source": [
    "## =================================== ##"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a86c013-a5bd-49f8-b27f-df89e2648582",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 2 Comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faeea2ea-b0ca-45a7-ae29-8c48e20a74af",
   "metadata": {},
   "source": [
    "### 2.1 Comparison: Unique common old vs new clinvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "9cf614be-e049-4688-ba63-4f1885000e55",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparison results saved in 'protein_change_comparison.xlsx'.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the two datasets\n",
    "df_cdkl5_missense = pd.read_excel('00_data/clinvar_result_copy.xlsx', sheet_name='CDKL5_missense_AF')\n",
    "df_cdkl5_condition = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='CDKL5_Condition')\n",
    "\n",
    "# Extract 'Protein change' column from both datasets\n",
    "protein_missense = set(df_cdkl5_missense['Protein change'].dropna())\n",
    "protein_condition = set(df_cdkl5_condition['Protein change'].dropna())\n",
    "\n",
    "# Find common and unique values\n",
    "common_protein_changes = protein_missense.intersection(protein_condition)\n",
    "unique_missense = protein_missense - protein_condition\n",
    "unique_condition = protein_condition - protein_missense\n",
    "\n",
    "# Convert results to DataFrames for saving\n",
    "common_df = pd.DataFrame(list(common_protein_changes), columns=['Common Protein Changes'])\n",
    "unique_missense_df = pd.DataFrame(list(unique_missense), columns=['Unique to CDKL5_missense_AF'])\n",
    "unique_condition_df = pd.DataFrame(list(unique_condition), columns=['Unique to CDKL5_Condition'])\n",
    "\n",
    "# Save comparisons in a new Excel file\n",
    "comparison_file = \"00_data/protein_change_comparison.xlsx\"\n",
    "with pd.ExcelWriter(comparison_file, mode='w', engine='openpyxl') as writer:\n",
    "    common_df.to_excel(writer, sheet_name='Common_Protein_Changes', index=False)\n",
    "    unique_missense_df.to_excel(writer, sheet_name='Unique_CDKL5_missense_AF', index=False)\n",
    "    unique_condition_df.to_excel(writer, sheet_name='Unique_CDKL5_Condition', index=False)\n",
    "\n",
    "print(\"Comparison results saved in 'protein_change_comparison.xlsx'.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ebd7f7c-3bbd-4e75-87a6-6f3228747966",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 2.2 Comparison: Unique common old vs new clinvar after deleting 'intron variants'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "31f40eef-1870-4de5-8f30-3e3c56cb3826",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Comparison Summary:\n",
      "Total in CDKL5_missense_AF: 120\n",
      "Total in CDKL5_missense_only: 120\n",
      "Total Common: 120\n",
      "Total Unique to CDKL5_missense_AF: 0\n",
      "Total Unique to CDKL5_missense_only: 0\n",
      "\n",
      "Comparison results saved in 'protein_change_comparison.xlsx'.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the two datasets\n",
    "df_cdkl5_missense = pd.read_excel('00_data/clinvar_result_copy.xlsx', sheet_name='CDKL5_missense_AF')\n",
    "df_cdkl5_missense_only = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='CDKL5_missense_only')\n",
    "\n",
    "# Extract 'Protein change' column from both datasets\n",
    "protein_missense = set(df_cdkl5_missense['Protein change'].dropna())\n",
    "protein_missense_only = set(df_cdkl5_missense_only['Protein change'].dropna())\n",
    "\n",
    "# Find common and unique values\n",
    "common_protein_changes = protein_missense.intersection(protein_missense_only)\n",
    "unique_missense = protein_missense - protein_missense_only\n",
    "unique_missense_only = protein_missense_only - protein_missense\n",
    "\n",
    "# Create a summary dictionary\n",
    "summary = {\n",
    "    'Total in CDKL5_missense_AF': len(protein_missense),\n",
    "    'Total in CDKL5_missense_only': len(protein_missense_only),\n",
    "    'Total Common': len(common_protein_changes),\n",
    "    'Total Unique to CDKL5_missense_AF': len(unique_missense),\n",
    "    'Total Unique to CDKL5_missense_only': len(unique_missense_only)\n",
    "}\n",
    "\n",
    "# Print the summary\n",
    "print(\"\\nComparison Summary:\")\n",
    "for key, value in summary.items():\n",
    "    print(f\"{key}: {value}\")\n",
    "\n",
    "# Convert results to DataFrames for saving\n",
    "summary_df = pd.DataFrame(list(summary.items()), columns=['Category', 'Count'])\n",
    "common_df = pd.DataFrame(list(common_protein_changes), columns=['Common Protein Changes'])\n",
    "unique_missense_df = pd.DataFrame(list(unique_missense), columns=['Unique to CDKL5_missense_AF'])\n",
    "unique_missense_only_df = pd.DataFrame(list(unique_missense_only), columns=['Unique to CDKL5_missense_only'])\n",
    "\n",
    "# Save comparisons in a new Excel file\n",
    "comparison_file = \"00_data/protein_change_comparison.xlsx\"\n",
    "with pd.ExcelWriter(comparison_file, mode='w', engine='openpyxl') as writer:\n",
    "    summary_df.to_excel(writer, sheet_name='Comparison_Summary', index=False)\n",
    "    common_df.to_excel(writer, sheet_name='Common_Protein_Changes', index=False)\n",
    "    unique_missense_df.to_excel(writer, sheet_name='Unique_CDKL5_missense_AF', index=False)\n",
    "    unique_missense_only_df.to_excel(writer, sheet_name='Unique_CDKL5_missense_only', index=False)\n",
    "\n",
    "print(\"\\nComparison results saved in 'protein_change_comparison.xlsx'.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de6c91c8-838d-4843-a02c-729d25daccde",
   "metadata": {},
   "source": [
    "### 2.3 Comparison gnomAD vs Clinvar"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4dc9c67-cf5e-43a4-a3be-5855abc5f46a",
   "metadata": {},
   "source": [
    "#### 2.3.1 check clinvar sheets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7591eb36-6bdc-4249-b390-063c5504d5a7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Sheet1', 'Single_Nucleotide_Variant', 'Missense_Variant', 'CDKL5_Condition', 'CDKL5_missense_only']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# check sheet names\n",
    "excel_file = '00_data/clinvar_result.xlsx'\n",
    "sheet_names = pd.ExcelFile(excel_file).sheet_names\n",
    "\n",
    "print(sheet_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2a846a55-e92a-433d-bc0a-fbbe26933cc4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Column names: ['Name', 'Gene(s)', 'Protein change', 'Condition(s)', 'Accession', 'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome', 'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID', 'Canonical SPDI', 'Variant type', 'Molecular consequence', 'Germline classification', 'Germline date last evaluated', 'Germline review status', 'Somatic clinical impact', 'Somatic clinical impact date last evaluated', 'Somatic clinical impact review status', 'Oncogenicity classification', 'Oncogenicity date last evaluated', 'Oncogenicity review status', 'Unnamed: 24']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load a specific sheet\n",
    "df = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name = 'CDKL5_missense_only')\n",
    "\n",
    "print(\"Column names:\", df.columns.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b77f69e-1701-47b1-8692-de92a62a1237",
   "metadata": {},
   "source": [
    "### 2.4 compare the 'Protein change' column \n",
    "from clinvar_result.xlsx and the 'Protein change (One-Letter)' column from gnomAD.xlsx, and print the number of common and unique entries in each:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ac354bf3-9f30-4b98-913c-48f416340bf3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of common protein changes: 51\n",
      "Number of unique protein changes in clinvar_result.xlsx: 69\n",
      "Number of unique protein changes in gnomAD.xlsx: 215\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the ClinVar dataset\n",
    "df1 = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='CDKL5_missense_only')\n",
    "\n",
    "# Load the gnomAD dataset\n",
    "df2 = pd.read_excel('00_data/gnomAD.xlsx', sheet_name='ClinVar_Classified_Only')\n",
    "\n",
    "# Extract relevant columns and drop NaNs\n",
    "set1 = set(df1['Protein change'].dropna().astype(str))\n",
    "set2 = set(df2['Protein change (One-Letter)'].dropna().astype(str))\n",
    "\n",
    "# Find common and unique elements\n",
    "common = set1.intersection(set2)\n",
    "unique_to_df1 = set1 - set2\n",
    "unique_to_df2 = set2 - set1\n",
    "\n",
    "# Print the results\n",
    "print(f\"Number of common protein changes: {len(common)}\")\n",
    "print(f\"Number of unique protein changes in clinvar_result.xlsx: {len(unique_to_df1)}\")\n",
    "print(f\"Number of unique protein changes in gnomAD.xlsx: {len(unique_to_df2)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6236b47c-a508-48a2-ba5c-195593f4f29e",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 2 gnomAD"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "599b6a33-8c77-4d7c-89c3-29a3045410cb",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 2.1 csv to excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "7d5c66f6-a138-457b-ae2b-677ca60b57aa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CSV file converted to Excel: 00_data/gnomAD.xlsx\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the CSV file\n",
    "csv_file = \"00_data/gnomAD.csv\"\n",
    "df_gnomad = pd.read_csv(csv_file)\n",
    "\n",
    "# Save as an Excel file\n",
    "excel_file = \"00_data/gnomAD.xlsx\"\n",
    "df_gnomad.to_excel(excel_file, index=False, engine='openpyxl')\n",
    "\n",
    "print(f\"CSV file converted to Excel: {excel_file}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0864eb5-3a62-41ea-be6a-c05a58a4c8f8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 2.2 view gnomAD data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "a0fc0a94-ae57-4ce9-98c7-c67c0c67c0f9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gnomAD ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>rsIDs</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Alternate</th>\n",
       "      <th>Source</th>\n",
       "      <th>Filters - exomes</th>\n",
       "      <th>Filters - genomes</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>...</th>\n",
       "      <th>Homozygote Count Amish</th>\n",
       "      <th>Hemizygote Count Amish</th>\n",
       "      <th>Allele Count South Asian</th>\n",
       "      <th>Allele Number South Asian</th>\n",
       "      <th>Homozygote Count South Asian</th>\n",
       "      <th>Hemizygote Count South Asian</th>\n",
       "      <th>Allele Count Remaining</th>\n",
       "      <th>Allele Number Remaining</th>\n",
       "      <th>Homozygote Count Remaining</th>\n",
       "      <th>Hemizygote Count Remaining</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X-18507022-C-A</td>\n",
       "      <td>X</td>\n",
       "      <td>18507022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>AC0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000623535.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47228</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32581</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X-18507025-A-T</td>\n",
       "      <td>X</td>\n",
       "      <td>18507025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>AC0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000623535.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>48182</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>33314</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>X-18507026-C-T</td>\n",
       "      <td>X</td>\n",
       "      <td>18507026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>AS_VQSR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000623535.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>48511</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>33958</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X-18507028-G-A</td>\n",
       "      <td>X</td>\n",
       "      <td>18507028</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000623535.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>48917</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>34634</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>X-18507029-C-A</td>\n",
       "      <td>X</td>\n",
       "      <td>18507029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>AC0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000623535.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49033</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>34927</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4603</th>\n",
       "      <td>X-18653612-C-T</td>\n",
       "      <td>X</td>\n",
       "      <td>18653612</td>\n",
       "      <td>rs1602318427</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>AC0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000379989.6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>53064</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>45106</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4604</th>\n",
       "      <td>X-18653614-A-C</td>\n",
       "      <td>X</td>\n",
       "      <td>18653614</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000379989.6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>52930</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>44818</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4605</th>\n",
       "      <td>X-18653614-A-G</td>\n",
       "      <td>X</td>\n",
       "      <td>18653614</td>\n",
       "      <td>rs1244414729</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>gnomAD Exomes,gnomAD Genomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>ENST00000379989.6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>52802</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>44775</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4606</th>\n",
       "      <td>X-18653615-T-C</td>\n",
       "      <td>X</td>\n",
       "      <td>18653615</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000379989.6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>52926</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>44862</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4607</th>\n",
       "      <td>X-18653616-G-T</td>\n",
       "      <td>X</td>\n",
       "      <td>18653616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>AC0,AS_VQSR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000379989.6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>52811</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>44847</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4608 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gnomAD ID Chromosome  Position         rsIDs Reference Alternate  \\\n",
       "0     X-18507022-C-A          X  18507022           NaN         C         A   \n",
       "1     X-18507025-A-T          X  18507025           NaN         A         T   \n",
       "2     X-18507026-C-T          X  18507026           NaN         C         T   \n",
       "3     X-18507028-G-A          X  18507028           NaN         G         A   \n",
       "4     X-18507029-C-A          X  18507029           NaN         C         A   \n",
       "...              ...        ...       ...           ...       ...       ...   \n",
       "4603  X-18653612-C-T          X  18653612  rs1602318427         C         T   \n",
       "4604  X-18653614-A-C          X  18653614           NaN         A         C   \n",
       "4605  X-18653614-A-G          X  18653614  rs1244414729         A         G   \n",
       "4606  X-18653615-T-C          X  18653615           NaN         T         C   \n",
       "4607  X-18653616-G-T          X  18653616           NaN         G         T   \n",
       "\n",
       "                            Source Filters - exomes Filters - genomes  \\\n",
       "0                    gnomAD Exomes              AC0               NaN   \n",
       "1                    gnomAD Exomes              AC0               NaN   \n",
       "2                    gnomAD Exomes          AS_VQSR               NaN   \n",
       "3                    gnomAD Exomes             PASS               NaN   \n",
       "4                    gnomAD Exomes              AC0               NaN   \n",
       "...                            ...              ...               ...   \n",
       "4603                 gnomAD Exomes              AC0               NaN   \n",
       "4604                 gnomAD Exomes             PASS               NaN   \n",
       "4605  gnomAD Exomes,gnomAD Genomes             PASS              PASS   \n",
       "4606                 gnomAD Exomes             PASS               NaN   \n",
       "4607                 gnomAD Exomes      AC0,AS_VQSR               NaN   \n",
       "\n",
       "             Transcript  ... Homozygote Count Amish Hemizygote Count Amish  \\\n",
       "0     ENST00000623535.2  ...                      0                      0   \n",
       "1     ENST00000623535.2  ...                      0                      0   \n",
       "2     ENST00000623535.2  ...                      0                      0   \n",
       "3     ENST00000623535.2  ...                      0                      0   \n",
       "4     ENST00000623535.2  ...                      0                      0   \n",
       "...                 ...  ...                    ...                    ...   \n",
       "4603  ENST00000379989.6  ...                      0                      0   \n",
       "4604  ENST00000379989.6  ...                      0                      0   \n",
       "4605  ENST00000379989.6  ...                      0                      0   \n",
       "4606  ENST00000379989.6  ...                      0                      0   \n",
       "4607  ENST00000379989.6  ...                      0                      0   \n",
       "\n",
       "     Allele Count South Asian Allele Number South Asian  \\\n",
       "0                           0                     47228   \n",
       "1                           0                     48182   \n",
       "2                           1                     48511   \n",
       "3                           0                     48917   \n",
       "4                           0                     49033   \n",
       "...                       ...                       ...   \n",
       "4603                        0                     53064   \n",
       "4604                        0                     52930   \n",
       "4605                        0                     52802   \n",
       "4606                        1                     52926   \n",
       "4607                        0                     52811   \n",
       "\n",
       "     Homozygote Count South Asian  Hemizygote Count South Asian  \\\n",
       "0                               0                             0   \n",
       "1                               0                             0   \n",
       "2                               0                             1   \n",
       "3                               0                             0   \n",
       "4                               0                             0   \n",
       "...                           ...                           ...   \n",
       "4603                            0                             0   \n",
       "4604                            0                             0   \n",
       "4605                            0                             0   \n",
       "4606                            0                             0   \n",
       "4607                            0                             0   \n",
       "\n",
       "     Allele Count Remaining  Allele Number Remaining  \\\n",
       "0                         0                    32581   \n",
       "1                         0                    33314   \n",
       "2                         0                    33958   \n",
       "3                         1                    34634   \n",
       "4                         0                    34927   \n",
       "...                     ...                      ...   \n",
       "4603                      0                    45106   \n",
       "4604                      0                    44818   \n",
       "4605                      1                    44775   \n",
       "4606                      0                    44862   \n",
       "4607                      0                    44847   \n",
       "\n",
       "      Homozygote Count Remaining  Hemizygote Count Remaining  \n",
       "0                              0                           0  \n",
       "1                              0                           0  \n",
       "2                              0                           0  \n",
       "3                              0                           0  \n",
       "4                              0                           0  \n",
       "...                          ...                         ...  \n",
       "4603                           0                           0  \n",
       "4604                           0                           0  \n",
       "4605                           0                           0  \n",
       "4606                           0                           0  \n",
       "4607                           0                           0  \n",
       "\n",
       "[4608 rows x 72 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the Excel file\n",
    "df_gnomad = pd.read_excel('00_data/gnomAD.xlsx')\n",
    "\n",
    "# Display the dataset in IPython style\n",
    "display(df_gnomad)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "65345639-90e9-4fda-bc6e-53822c638ac7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column Names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gnomAD ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Position</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rsIDs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Hemizygote Count South Asian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>Allele Count Remaining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>Allele Number Remaining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>Homozygote Count Remaining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Hemizygote Count Remaining</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Column Names\n",
       "0                      gnomAD ID\n",
       "1                     Chromosome\n",
       "2                       Position\n",
       "3                          rsIDs\n",
       "4                      Reference\n",
       "..                           ...\n",
       "67  Hemizygote Count South Asian\n",
       "68        Allele Count Remaining\n",
       "69       Allele Number Remaining\n",
       "70    Homozygote Count Remaining\n",
       "71    Hemizygote Count Remaining\n",
       "\n",
       "[72 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the Excel file\n",
    "df_gnomad = pd.read_excel('00_data/gnomAD.xlsx', sheet_name='Sheet1')\n",
    "\n",
    "# Convert column names to DataFrame for better visualization\n",
    "column_names_df = pd.DataFrame(df_gnomad.columns, columns=['Column Names'])\n",
    "\n",
    "# Display the column names as a table\n",
    "display(column_names_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd52b46e-cccb-42b2-8ad3-fb4f5ccf7d11",
   "metadata": {},
   "source": [
    "## Value count gnomAD sheet1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f71eeae4-e877-4396-89d6-4e3493e5d205",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Value counts for 'Sheet1' sheet:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VEP Annotation</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>intron_variant</td>\n",
       "      <td>2563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>missense_variant</td>\n",
       "      <td>971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>splice_region_variant</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>stop_gained</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>splice_donor_variant</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>stop_lost</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>start_lost</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>stop_retained_variant</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             VEP Annotation  Count\n",
       "0            intron_variant   2563\n",
       "1          missense_variant    971\n",
       "2        synonymous_variant    483\n",
       "3       3_prime_UTR_variant    235\n",
       "4     splice_region_variant    146\n",
       "5       5_prime_UTR_variant     67\n",
       "6        frameshift_variant     37\n",
       "7               stop_gained     32\n",
       "8      splice_donor_variant     28\n",
       "9   splice_acceptor_variant     21\n",
       "10         inframe_deletion     12\n",
       "11        inframe_insertion      7\n",
       "12                stop_lost      4\n",
       "13               start_lost      1\n",
       "14    stop_retained_variant      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total entries in 'VEP Annotation': 4608\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ClinVar Germline Classification</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Likely benign</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Benign</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Conflicting classifications of pathogenicity</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                ClinVar Germline Classification  Count\n",
       "0                                           NaN   3994\n",
       "1                                 Likely benign    266\n",
       "2                        Uncertain significance    161\n",
       "3                                        Benign     70\n",
       "4                          Benign/Likely benign     39\n",
       "5  Conflicting classifications of pathogenicity     35\n",
       "6                                    Pathogenic     25\n",
       "7                             Likely pathogenic     11\n",
       "8                  Pathogenic/Likely pathogenic      7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total entries in 'ClinVar Germline Classification': 4608\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the Excel file\n",
    "df_gnomad = pd.read_excel('00_data/gnomAD.xlsx', sheet_name='Sheet1')\n",
    "\n",
    "# Define columns to count\n",
    "columns_to_count = ['VEP Annotation', 'ClinVar Germline Classification']\n",
    "\n",
    "# Function to display value counts with total entries including NaNs\n",
    "def show_value_counts_with_total(df, sheet_name, columns):\n",
    "    print(f\"\\nValue counts for '{sheet_name}' sheet:\")\n",
    "    for col in columns:\n",
    "        counts = df[col].value_counts(dropna=False).reset_index()\n",
    "        counts.columns = [col, 'Count']\n",
    "        display(counts)\n",
    "        total = len(df[col])\n",
    "        print(f\"Total entries in '{col}': {total}\\n\")\n",
    "\n",
    "# Show value counts for 'Sheet1' with totals\n",
    "show_value_counts_with_total(df_gnomad, 'Sheet1', columns_to_count)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2ec7f82-0b47-4ae1-8593-8a025e6e79ca",
   "metadata": {},
   "source": [
    "## 2.3 Missense_Variant "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "8ffcb0bc-f1ce-4068-b29f-e26e17340876",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved as 'Missense_Variant' in the same Excel file.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "# Load the existing Excel file\n",
    "file_path = '00_data/gnomAD.xlsx'\n",
    "df_gnomad = pd.read_excel(file_path, sheet_name='Sheet1')\n",
    "\n",
    "# Filter for 'missense_variant' entries\n",
    "missense_df = df_gnomad[df_gnomad['VEP Annotation'].str.contains('missense_variant', na=False)]\n",
    "\n",
    "# Save the filtered data to a new sheet with an appropriate name\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl') as writer:\n",
    "    missense_df.to_excel(writer, sheet_name='Missense_Variant', index=False)\n",
    "\n",
    "print(\"Filtered data saved as 'Missense_Variant' in the same Excel file.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "49d78485-ab7c-43f5-b643-008ff58ccac1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Value counts for 'Sheet1' sheet:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VEP Annotation</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>missense_variant</td>\n",
       "      <td>971</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     VEP Annotation  Count\n",
       "0  missense_variant    971"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total entries in 'VEP Annotation': 971\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ClinVar Germline Classification</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Likely benign</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Conflicting classifications of pathogenicity</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Benign</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                ClinVar Germline Classification  Count\n",
       "0                                           NaN    705\n",
       "1                        Uncertain significance    138\n",
       "2                                 Likely benign     56\n",
       "3  Conflicting classifications of pathogenicity     24\n",
       "4                                        Benign     21\n",
       "5                          Benign/Likely benign     12\n",
       "6                             Likely pathogenic      7\n",
       "7                                    Pathogenic      6\n",
       "8                  Pathogenic/Likely pathogenic      2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total entries in 'ClinVar Germline Classification': 971\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the Excel file\n",
    "df_gnomad = pd.read_excel('00_data/gnomAD.xlsx', sheet_name='Missense_Variant')\n",
    "\n",
    "# Define columns to count\n",
    "columns_to_count = ['VEP Annotation', 'ClinVar Germline Classification']\n",
    "\n",
    "# Function to display value counts with total entries including NaNs\n",
    "def show_value_counts_with_total(df, sheet_name, columns):\n",
    "    print(f\"\\nValue counts for '{sheet_name}' sheet:\")\n",
    "    for col in columns:\n",
    "        counts = df[col].value_counts(dropna=False).reset_index()\n",
    "        counts.columns = [col, 'Count']\n",
    "        display(counts)\n",
    "        total = len(df[col])\n",
    "        print(f\"Total entries in '{col}': {total}\\n\")\n",
    "\n",
    "# Show value counts for 'Sheet1' with totals\n",
    "show_value_counts_with_total(df_gnomad, 'Sheet1', columns_to_count)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04292537-84f3-4831-84d3-5994d4c10104",
   "metadata": {},
   "source": [
    "## 2.4 Protein_change_converted: p.Lys2Arg to K2R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "73a66839-e3b7-4acd-b93c-ca80fd44f4f1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein changes converted to one-letter format and saved in existing sheet 'Protein_Change_Converted'.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "# Full mapping of three-letter amino acid codes to one-letter codes\n",
    "aa_mapping = {\n",
    "    \"Ala\": \"A\", \"Arg\": \"R\", \"Asn\": \"N\", \"Asp\": \"D\",\n",
    "    \"Cys\": \"C\", \"Glu\": \"E\", \"Gln\": \"Q\", \"Gly\": \"G\",\n",
    "    \"His\": \"H\", \"Ile\": \"I\", \"Leu\": \"L\", \"Lys\": \"K\",\n",
    "    \"Met\": \"M\", \"Phe\": \"F\", \"Pro\": \"P\", \"Ser\": \"S\",\n",
    "    \"Thr\": \"T\", \"Trp\": \"W\", \"Tyr\": \"Y\", \"Val\": \"V\",\n",
    "    \"Sec\": \"U\", \"Pyl\": \"O\", \"Ter\": \"*\"\n",
    "}\n",
    "\n",
    "# Load the Excel file and the specified sheet\n",
    "file_path = '00_data/gnomAD.xlsx'\n",
    "df_gnomad = pd.read_excel(file_path, sheet_name='Protein_Change_Converted')\n",
    "\n",
    "# Strip 'p.' and store in a new column for parsing\n",
    "df_gnomad['Protein Consequence (Three-Letter)'] = df_gnomad['Protein Consequence'].str.replace(\"p.\", \"\", regex=False)\n",
    "\n",
    "# Function to convert three-letter amino acid change to one-letter format\n",
    "def convert_to_one_letter(protein_consequence):\n",
    "    try:\n",
    "        wt = protein_consequence[:3]                         # Wild-type (3-letter)\n",
    "        mut = protein_consequence[-3:]                       # Mutant (3-letter)\n",
    "        pos = protein_consequence[3:-3]                      # Position\n",
    "        \n",
    "        wt_1 = aa_mapping.get(wt, '?')\n",
    "        mut_1 = aa_mapping.get(mut, '?')\n",
    "\n",
    "        return f\"{wt_1}{pos}{mut_1}\"\n",
    "    except Exception:\n",
    "        return None\n",
    "\n",
    "# Apply the function to create the corrected column\n",
    "df_gnomad['Protein change (One-Letter)'] = df_gnomad['Protein Consequence (Three-Letter)'].apply(convert_to_one_letter)\n",
    "\n",
    "# Save changes back to the same sheet (overwriting it)\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl', if_sheet_exists='replace') as writer:\n",
    "    df_gnomad.to_excel(writer, sheet_name='Protein_Change_Converted', index=False)\n",
    "\n",
    "print(\"Protein changes converted to one-letter format and saved in existing sheet 'Protein_Change_Converted'.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4657814-18b3-4bd6-8a07-7615631b148b",
   "metadata": {},
   "source": [
    "### Sanity check one letter code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a43443a0-bd00-487c-9e90-b9f31eae55b5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "✅ Valid entries: 971\n",
      "❌ Invalid entries: 0\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# Load the updated sheet\n",
    "file_path = '00_data/gnomAD.xlsx'\n",
    "df = pd.read_excel(file_path, sheet_name='Protein_Change_Converted')\n",
    "\n",
    "# Define set of valid one-letter amino acid codes (including stop codon '*')\n",
    "valid_aa = set('ACDEFGHIKLMNPQRSTVWY*')\n",
    "\n",
    "# Sanity check function\n",
    "def is_valid_change(change):\n",
    "    if not isinstance(change, str):\n",
    "        return False\n",
    "    match = re.match(r'^([A-Z*])(\\d+)([A-Z*])$', change)\n",
    "    if match:\n",
    "        wt, pos, mut = match.groups()\n",
    "        return wt in valid_aa and mut in valid_aa\n",
    "    return False\n",
    "\n",
    "# Apply sanity check\n",
    "df['Is_Valid_One_Letter_Change'] = df['Protein change (One-Letter)'].apply(is_valid_change)\n",
    "\n",
    "# Report results\n",
    "valid_count = df['Is_Valid_One_Letter_Change'].sum()\n",
    "invalid_count = (~df['Is_Valid_One_Letter_Change']).sum()\n",
    "\n",
    "print(f\"\\n✅ Valid entries: {valid_count}\")\n",
    "print(f\"❌ Invalid entries: {invalid_count}\")\n",
    "\n",
    "# Show examples of invalid entries\n",
    "if invalid_count > 0:\n",
    "    print(\"\\nExamples of invalid entries:\")\n",
    "    print(df.loc[~df['Is_Valid_One_Letter_Change'], ['Protein Consequence', 'Protein change (One-Letter)']].head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "50292d14-67f5-4f50-8bae-af187cb0078a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gnomAD ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>rsIDs</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Alternate</th>\n",
       "      <th>Source</th>\n",
       "      <th>Filters - exomes</th>\n",
       "      <th>Filters - genomes</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>...</th>\n",
       "      <th>Allele Count South Asian</th>\n",
       "      <th>Allele Number South Asian</th>\n",
       "      <th>Homozygote Count South Asian</th>\n",
       "      <th>Hemizygote Count South Asian</th>\n",
       "      <th>Allele Count Remaining</th>\n",
       "      <th>Allele Number Remaining</th>\n",
       "      <th>Homozygote Count Remaining</th>\n",
       "      <th>Hemizygote Count Remaining</th>\n",
       "      <th>Protein Consequence (Three-Letter)</th>\n",
       "      <th>Protein change (One-Letter)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X-18507101-A-G</td>\n",
       "      <td>X</td>\n",
       "      <td>18507101</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>AC0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000623535.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>56517</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47048</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Lys2Arg</td>\n",
       "      <td>K2R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X-18507103-A-T</td>\n",
       "      <td>X</td>\n",
       "      <td>18507103</td>\n",
       "      <td>rs138539908</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>gnomAD Exomes,gnomAD Genomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>ENST00000623535.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>56357</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47093</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ile3Phe</td>\n",
       "      <td>I3F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>X-18507107-C-T</td>\n",
       "      <td>X</td>\n",
       "      <td>18507107</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000623535.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>56593</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>47210</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Pro4Leu</td>\n",
       "      <td>P4L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X-18507109-A-G</td>\n",
       "      <td>X</td>\n",
       "      <td>18507109</td>\n",
       "      <td>rs767844474</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>gnomAD Exomes,gnomAD Genomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>ENST00000623535.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>56443</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>47190</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Asn5Asp</td>\n",
       "      <td>N5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>X-18507113-T-C</td>\n",
       "      <td>X</td>\n",
       "      <td>18507113</td>\n",
       "      <td>rs746036179</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000623535.2</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>56668</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>47290</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Ile6Thr</td>\n",
       "      <td>I6T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>966</th>\n",
       "      <td>X-18653536-G-A</td>\n",
       "      <td>X</td>\n",
       "      <td>18653536</td>\n",
       "      <td>rs201765217</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gnomAD Exomes,gnomAD Genomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>ENST00000379989.6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>56824</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>47593</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Gly1029Ser</td>\n",
       "      <td>G1029S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>967</th>\n",
       "      <td>X-18653536-G-C</td>\n",
       "      <td>X</td>\n",
       "      <td>18653536</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>AS_VQSR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000379989.6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>56952</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47638</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Gly1029Arg</td>\n",
       "      <td>G1029R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>968</th>\n",
       "      <td>X-18653537-G-T</td>\n",
       "      <td>X</td>\n",
       "      <td>18653537</td>\n",
       "      <td>rs1250531842</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000379989.6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>56932</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47638</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Gly1029Val</td>\n",
       "      <td>G1029V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>969</th>\n",
       "      <td>X-18653539-A-G</td>\n",
       "      <td>X</td>\n",
       "      <td>18653539</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000379989.6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>56959</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47634</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Lys1030Glu</td>\n",
       "      <td>K1030E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>X-18653540-A-T</td>\n",
       "      <td>X</td>\n",
       "      <td>18653540</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>gnomAD Exomes</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000379989.6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>56928</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47628</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Lys1030Met</td>\n",
       "      <td>K1030M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>971 rows × 74 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gnomAD ID Chromosome  Position         rsIDs Reference Alternate  \\\n",
       "0    X-18507101-A-G          X  18507101           NaN         A         G   \n",
       "1    X-18507103-A-T          X  18507103   rs138539908         A         T   \n",
       "2    X-18507107-C-T          X  18507107           NaN         C         T   \n",
       "3    X-18507109-A-G          X  18507109   rs767844474         A         G   \n",
       "4    X-18507113-T-C          X  18507113   rs746036179         T         C   \n",
       "..              ...        ...       ...           ...       ...       ...   \n",
       "966  X-18653536-G-A          X  18653536   rs201765217         G         A   \n",
       "967  X-18653536-G-C          X  18653536           NaN         G         C   \n",
       "968  X-18653537-G-T          X  18653537  rs1250531842         G         T   \n",
       "969  X-18653539-A-G          X  18653539           NaN         A         G   \n",
       "970  X-18653540-A-T          X  18653540           NaN         A         T   \n",
       "\n",
       "                           Source Filters - exomes Filters - genomes  \\\n",
       "0                   gnomAD Exomes              AC0               NaN   \n",
       "1    gnomAD Exomes,gnomAD Genomes             PASS              PASS   \n",
       "2                   gnomAD Exomes             PASS               NaN   \n",
       "3    gnomAD Exomes,gnomAD Genomes             PASS              PASS   \n",
       "4                   gnomAD Exomes             PASS               NaN   \n",
       "..                            ...              ...               ...   \n",
       "966  gnomAD Exomes,gnomAD Genomes             PASS              PASS   \n",
       "967                 gnomAD Exomes          AS_VQSR               NaN   \n",
       "968                 gnomAD Exomes             PASS               NaN   \n",
       "969                 gnomAD Exomes             PASS               NaN   \n",
       "970                 gnomAD Exomes             PASS               NaN   \n",
       "\n",
       "            Transcript  ... Allele Count South Asian  \\\n",
       "0    ENST00000623535.2  ...                        0   \n",
       "1    ENST00000623535.2  ...                        0   \n",
       "2    ENST00000623535.2  ...                        0   \n",
       "3    ENST00000623535.2  ...                        0   \n",
       "4    ENST00000623535.2  ...                        6   \n",
       "..                 ...  ...                      ...   \n",
       "966  ENST00000379989.6  ...                        0   \n",
       "967  ENST00000379989.6  ...                        0   \n",
       "968  ENST00000379989.6  ...                        0   \n",
       "969  ENST00000379989.6  ...                        0   \n",
       "970  ENST00000379989.6  ...                        0   \n",
       "\n",
       "    Allele Number South Asian Homozygote Count South Asian  \\\n",
       "0                       56517                            0   \n",
       "1                       56357                            0   \n",
       "2                       56593                            0   \n",
       "3                       56443                            0   \n",
       "4                       56668                            0   \n",
       "..                        ...                          ...   \n",
       "966                     56824                            0   \n",
       "967                     56952                            0   \n",
       "968                     56932                            0   \n",
       "969                     56959                            0   \n",
       "970                     56928                            0   \n",
       "\n",
       "    Hemizygote Count South Asian Allele Count Remaining  \\\n",
       "0                              0                      0   \n",
       "1                              0                      0   \n",
       "2                              0                      1   \n",
       "3                              0                      1   \n",
       "4                              3                      0   \n",
       "..                           ...                    ...   \n",
       "966                            0                      1   \n",
       "967                            0                      0   \n",
       "968                            0                      0   \n",
       "969                            0                      0   \n",
       "970                            0                      0   \n",
       "\n",
       "     Allele Number Remaining  Homozygote Count Remaining  \\\n",
       "0                      47048                           0   \n",
       "1                      47093                           0   \n",
       "2                      47210                           0   \n",
       "3                      47190                           0   \n",
       "4                      47290                           0   \n",
       "..                       ...                         ...   \n",
       "966                    47593                           0   \n",
       "967                    47638                           0   \n",
       "968                    47638                           0   \n",
       "969                    47634                           0   \n",
       "970                    47628                           0   \n",
       "\n",
       "     Hemizygote Count Remaining  Protein Consequence (Three-Letter)  \\\n",
       "0                             0                             Lys2Arg   \n",
       "1                             0                             Ile3Phe   \n",
       "2                             1                             Pro4Leu   \n",
       "3                             0                             Asn5Asp   \n",
       "4                             0                             Ile6Thr   \n",
       "..                          ...                                 ...   \n",
       "966                           1                          Gly1029Ser   \n",
       "967                           0                          Gly1029Arg   \n",
       "968                           0                          Gly1029Val   \n",
       "969                           0                          Lys1030Glu   \n",
       "970                           0                          Lys1030Met   \n",
       "\n",
       "     Protein change (One-Letter)  \n",
       "0                            K2R  \n",
       "1                            I3F  \n",
       "2                            P4L  \n",
       "3                            N5D  \n",
       "4                            I6T  \n",
       "..                           ...  \n",
       "966                       G1029S  \n",
       "967                       G1029R  \n",
       "968                       G1029V  \n",
       "969                       K1030E  \n",
       "970                       K1030M  \n",
       "\n",
       "[971 rows x 74 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the updated Excel file\n",
    "df_gnomad = pd.read_excel('00_data/gnomAD.xlsx', sheet_name='Protein_Change_Converted')\n",
    "\n",
    "# Display the dataset in a clean format\n",
    "display(df_gnomad)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "af9eb819-66eb-4108-8a88-66f7e5536905",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Column Names in 'Protein_Change_Converted' sheet:\n",
      "Index(['gnomAD ID', 'Chromosome', 'Position', 'rsIDs', 'Reference',\n",
      "       'Alternate', 'Source', 'Filters - exomes', 'Filters - genomes',\n",
      "       'Transcript', 'HGVS Consequence', 'Protein Consequence',\n",
      "       'Transcript Consequence', 'VEP Annotation',\n",
      "       'ClinVar Germline Classification', 'ClinVar Variation ID', 'Flags',\n",
      "       'Allele Count', 'Allele Number', 'Allele Frequency', 'Homozygote Count',\n",
      "       'Hemizygote Count', 'Filters - joint', 'GroupMax FAF group',\n",
      "       'GroupMax FAF frequency', 'cadd', 'revel_max', 'spliceai_ds_max',\n",
      "       'pangolin_largest_ds', 'phylop', 'sift_max', 'polyphen_max',\n",
      "       'Allele Count African/African American',\n",
      "       'Allele Number African/African American',\n",
      "       'Homozygote Count African/African American',\n",
      "       'Hemizygote Count African/African American',\n",
      "       'Allele Count Admixed American', 'Allele Number Admixed American',\n",
      "       'Homozygote Count Admixed American',\n",
      "       'Hemizygote Count Admixed American', 'Allele Count Ashkenazi Jewish',\n",
      "       'Allele Number Ashkenazi Jewish', 'Homozygote Count Ashkenazi Jewish',\n",
      "       'Hemizygote Count Ashkenazi Jewish', 'Allele Count East Asian',\n",
      "       'Allele Number East Asian', 'Homozygote Count East Asian',\n",
      "       'Hemizygote Count East Asian', 'Allele Count European (Finnish)',\n",
      "       'Allele Number European (Finnish)',\n",
      "       'Homozygote Count European (Finnish)',\n",
      "       'Hemizygote Count European (Finnish)', 'Allele Count Middle Eastern',\n",
      "       'Allele Number Middle Eastern', 'Homozygote Count Middle Eastern',\n",
      "       'Hemizygote Count Middle Eastern',\n",
      "       'Allele Count European (non-Finnish)',\n",
      "       'Allele Number European (non-Finnish)',\n",
      "       'Homozygote Count European (non-Finnish)',\n",
      "       'Hemizygote Count European (non-Finnish)', 'Allele Count Amish',\n",
      "       'Allele Number Amish', 'Homozygote Count Amish',\n",
      "       'Hemizygote Count Amish', 'Allele Count South Asian',\n",
      "       'Allele Number South Asian', 'Homozygote Count South Asian',\n",
      "       'Hemizygote Count South Asian', 'Allele Count Remaining',\n",
      "       'Allele Number Remaining', 'Homozygote Count Remaining',\n",
      "       'Hemizygote Count Remaining', 'Protein Consequence (Three-Letter)',\n",
      "       'Protein change (One-Letter)'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the updated Excel file\n",
    "file_path = '00_data/gnomAD.xlsx'\n",
    "df_protein_change = pd.read_excel(file_path, sheet_name='Protein_Change_Converted')\n",
    "\n",
    "# Print column names\n",
    "print(\"\\nColumn Names in 'Protein_Change_Converted' sheet:\")\n",
    "print(df_protein_change.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9f33b23d-99e6-4d53-8b75-d9365e10acd4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Column Names in 'Protein_Change_Converted' sheet:\n",
      "Index(['gnomAD ID', 'Chromosome', 'Position', 'rsIDs', 'Reference',\n",
      "       'Alternate', 'Source', 'Filters - exomes', 'Filters - genomes',\n",
      "       'Transcript', 'HGVS Consequence', 'Protein Consequence',\n",
      "       'Transcript Consequence', 'VEP Annotation',\n",
      "       'ClinVar Germline Classification', 'ClinVar Variation ID', 'Flags',\n",
      "       'Allele Count', 'Allele Number', 'Allele Frequency', 'Homozygote Count',\n",
      "       'Hemizygote Count', 'Filters - joint', 'GroupMax FAF group',\n",
      "       'GroupMax FAF frequency', 'cadd', 'revel_max', 'spliceai_ds_max',\n",
      "       'pangolin_largest_ds', 'phylop', 'sift_max', 'polyphen_max',\n",
      "       'Allele Count African/African American',\n",
      "       'Allele Number African/African American',\n",
      "       'Homozygote Count African/African American',\n",
      "       'Hemizygote Count African/African American',\n",
      "       'Allele Count Admixed American', 'Allele Number Admixed American',\n",
      "       'Homozygote Count Admixed American',\n",
      "       'Hemizygote Count Admixed American', 'Allele Count Ashkenazi Jewish',\n",
      "       'Allele Number Ashkenazi Jewish', 'Homozygote Count Ashkenazi Jewish',\n",
      "       'Hemizygote Count Ashkenazi Jewish', 'Allele Count East Asian',\n",
      "       'Allele Number East Asian', 'Homozygote Count East Asian',\n",
      "       'Hemizygote Count East Asian', 'Allele Count European (Finnish)',\n",
      "       'Allele Number European (Finnish)',\n",
      "       'Homozygote Count European (Finnish)',\n",
      "       'Hemizygote Count European (Finnish)', 'Allele Count Middle Eastern',\n",
      "       'Allele Number Middle Eastern', 'Homozygote Count Middle Eastern',\n",
      "       'Hemizygote Count Middle Eastern',\n",
      "       'Allele Count European (non-Finnish)',\n",
      "       'Allele Number European (non-Finnish)',\n",
      "       'Homozygote Count European (non-Finnish)',\n",
      "       'Hemizygote Count European (non-Finnish)', 'Allele Count Amish',\n",
      "       'Allele Number Amish', 'Homozygote Count Amish',\n",
      "       'Hemizygote Count Amish', 'Allele Count South Asian',\n",
      "       'Allele Number South Asian', 'Homozygote Count South Asian',\n",
      "       'Hemizygote Count South Asian', 'Allele Count Remaining',\n",
      "       'Allele Number Remaining', 'Homozygote Count Remaining',\n",
      "       'Hemizygote Count Remaining', 'Protein Consequence (Three-Letter)',\n",
      "       'Protein change (One-Letter)'],\n",
      "      dtype='object')\n",
      "\n",
      "Value Counts for 'VEP Annotation':\n",
      "VEP Annotation\n",
      "missense_variant    971\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Value Counts for 'ClinVar Germline Classification':\n",
      "ClinVar Germline Classification\n",
      "NaN                                             705\n",
      "Uncertain significance                          138\n",
      "Likely benign                                    56\n",
      "Conflicting classifications of pathogenicity     24\n",
      "Benign                                           21\n",
      "Benign/Likely benign                             12\n",
      "Likely pathogenic                                 7\n",
      "Pathogenic                                        6\n",
      "Pathogenic/Likely pathogenic                      2\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the Excel sheet\n",
    "file_path = '00_data/gnomAD.xlsx'\n",
    "df_protein_change = pd.read_excel(file_path, sheet_name='Protein_Change_Converted')\n",
    "\n",
    "# Print column names (optional for verification)\n",
    "print(\"\\nColumn Names in 'Protein_Change_Converted' sheet:\")\n",
    "print(df_protein_change.columns)\n",
    "\n",
    "# Value counts for 'VEP Annotation'\n",
    "print(\"\\nValue Counts for 'VEP Annotation':\")\n",
    "print(df_protein_change['VEP Annotation'].value_counts(dropna=False))\n",
    "\n",
    "# Value counts for 'ClinVar Germline Classification'\n",
    "print(\"\\nValue Counts for 'ClinVar Germline Classification':\")\n",
    "print(df_protein_change['ClinVar Germline Classification'].value_counts(dropna=False))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "249e4cbb-c21e-4ece-a7c3-8580168ed72c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 2.5 make 'mutation' column with 'wild' 'position' 'mutant'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30973d5a-92b6-45e4-8a67-cb46f64c9f84",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 2.5 New sheet 'NaN' discarded on 'ClinVar Germline Classification'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "961efdda-7ea0-4551-80c8-f5ac7775411f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved to new sheet 'ClinVar_Classified_Only' in gnomAD.xlsx.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the Excel file and the specific sheet\n",
    "file_path = '00_data/gnomAD.xlsx'\n",
    "df = pd.read_excel(file_path, sheet_name='Protein_Change_Converted')\n",
    "\n",
    "# Filter out rows where 'ClinVar Germline Classification' is not NaN\n",
    "df_filtered = df[df['ClinVar Germline Classification'].notna()]\n",
    "\n",
    "# Save to a new sheet in the same Excel file\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl', if_sheet_exists='replace') as writer:\n",
    "    df_filtered.to_excel(writer, sheet_name='ClinVar_Classified_Only', index=False)\n",
    "\n",
    "print(\"Filtered data saved to new sheet 'ClinVar_Classified_Only' in gnomAD.xlsx.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d632fa6e-7a8f-4cfb-a53c-07b078e83c49",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Preview of 'ClinVar_Classified_Only' sheet:\n",
      "        gnomAD ID Chromosome  Position         rsIDs Reference Alternate  \\\n",
      "0  X-18507103-A-T          X  18507103   rs138539908         A         T   \n",
      "1  X-18507109-A-G          X  18507109   rs767844474         A         G   \n",
      "2  X-18507113-T-C          X  18507113   rs746036179         T         C   \n",
      "3  X-18507122-T-C          X  18507122  rs1569197962         T         C   \n",
      "4  X-18507154-G-A          X  18507154   rs267608418         G         A   \n",
      "\n",
      "                         Source Filters - exomes Filters - genomes  \\\n",
      "0  gnomAD Exomes,gnomAD Genomes             PASS              PASS   \n",
      "1  gnomAD Exomes,gnomAD Genomes             PASS              PASS   \n",
      "2                 gnomAD Exomes             PASS               NaN   \n",
      "3                 gnomAD Exomes             PASS               NaN   \n",
      "4                 gnomAD Exomes      AC0,AS_VQSR               NaN   \n",
      "\n",
      "          Transcript  ... Allele Count South Asian Allele Number South Asian  \\\n",
      "0  ENST00000623535.2  ...                        0                     56357   \n",
      "1  ENST00000623535.2  ...                        0                     56443   \n",
      "2  ENST00000623535.2  ...                        6                     56668   \n",
      "3  ENST00000623535.2  ...                        0                     56785   \n",
      "4  ENST00000623535.2  ...                        0                     56426   \n",
      "\n",
      "  Homozygote Count South Asian Hemizygote Count South Asian  \\\n",
      "0                            0                            0   \n",
      "1                            0                            0   \n",
      "2                            0                            3   \n",
      "3                            0                            0   \n",
      "4                            0                            0   \n",
      "\n",
      "  Allele Count Remaining  Allele Number Remaining  Homozygote Count Remaining  \\\n",
      "0                      0                    47093                           0   \n",
      "1                      1                    47190                           0   \n",
      "2                      0                    47290                           0   \n",
      "3                      0                    47403                           0   \n",
      "4                      0                    46927                           0   \n",
      "\n",
      "   Hemizygote Count Remaining  Protein Consequence (Three-Letter)  \\\n",
      "0                           0                             Ile3Phe   \n",
      "1                           0                             Asn5Asp   \n",
      "2                           0                             Ile6Thr   \n",
      "3                           0                             Val9Ala   \n",
      "4                           0                            Gly20Ser   \n",
      "\n",
      "   Protein change (One-Letter)  \n",
      "0                          I3F  \n",
      "1                          N5D  \n",
      "2                          I6T  \n",
      "3                          V9A  \n",
      "4                         G20S  \n",
      "\n",
      "[5 rows x 74 columns]\n",
      "\n",
      "Value Counts for 'ClinVar Germline Classification':\n",
      "ClinVar Germline Classification\n",
      "Uncertain significance                          138\n",
      "Likely benign                                    56\n",
      "Conflicting classifications of pathogenicity     24\n",
      "Benign                                           21\n",
      "Benign/Likely benign                             12\n",
      "Likely pathogenic                                 7\n",
      "Pathogenic                                        6\n",
      "Pathogenic/Likely pathogenic                      2\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Total entries in 'ClinVar Germline Classification': 266\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the new sheet\n",
    "file_path = '00_data/gnomAD.xlsx'\n",
    "df_filtered = pd.read_excel(file_path, sheet_name='ClinVar_Classified_Only')\n",
    "\n",
    "# Function to get value counts with total count\n",
    "def value_counts_with_total(df, column_name):\n",
    "    counts = df[column_name].value_counts(dropna=False)\n",
    "    total = len(df[column_name])\n",
    "    print(counts)\n",
    "    print(f\"\\nTotal entries in '{column_name}': {total}\")\n",
    "\n",
    "# Display the first few rows (optional)\n",
    "print(\"\\nPreview of 'ClinVar_Classified_Only' sheet:\")\n",
    "print(df_filtered.head())\n",
    "\n",
    "# Use the function for 'ClinVar Germline Classification'\n",
    "print(\"\\nValue Counts for 'ClinVar Germline Classification':\")\n",
    "value_counts_with_total(df_filtered, 'ClinVar Germline Classification')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f440c988-baca-44d3-92fb-b792ce4e8c62",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 3 Comparison: ClinVar vs gnomAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2355feac-d27f-4201-a156-3b816adb28b8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of common entries: 51\n",
      "Number of unique entries in ClinVar: 69\n",
      "Number of unique entries in gnomAD: 913\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load datasets from Excel files\n",
    "df_clinvar = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='CDKL5_missense_only')\n",
    "df_gnomad = pd.read_excel('00_data/gnomAD.xlsx', sheet_name='Protein_Change_Converted')\n",
    "\n",
    "# Extract relevant columns and clean (drop NaN, convert to string)\n",
    "set_clinvar = set(df_clinvar['Protein change'].dropna().astype(str))\n",
    "set_gnomad = set(df_gnomad['Protein change (One-Letter)'].dropna().astype(str))\n",
    "\n",
    "# Calculate common and unique protein changes\n",
    "common = set_clinvar.intersection(set_gnomad)\n",
    "unique_clinvar = set_clinvar - set_gnomad\n",
    "unique_gnomad = set_gnomad - set_clinvar\n",
    "\n",
    "# Print results with ClinVar first\n",
    "print(f\"Number of common entries: {len(common)}\")\n",
    "print(f\"Number of unique entries in ClinVar: {len(unique_clinvar)}\")\n",
    "print(f\"Number of unique entries in gnomAD: {len(unique_gnomad)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "f99354e5-37a0-4ecd-99d5-6a7003cefd09",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== Dataset Comparison Information ===\n",
      "ClinVar File           : 00_data/clinvar_result.xlsx\n",
      "ClinVar Sheet          : CDKL5_missense_only\n",
      "ClinVar Column         : Protein change\n",
      "gnomAD File            : 00_data/gnomAD.xlsx\n",
      "gnomAD Sheet           : Protein_Change_Converted\n",
      "gnomAD Column          : Protein change (One-Letter)\n",
      "======================================\n",
      "\n",
      "Number of common entries: 51\n",
      "Number of unique entries in ClinVar: 69\n",
      "Number of unique entries in gnomAD (set difference): 913\n",
      "Number of duplicate entries in unique gnomAD dataset: 7\n",
      "No extra entries found. Sets and dataframe filtering consistent.\n",
      "\n",
      "Value counts of 'ClinVar Germline Classification' for unique gnomAD entries:\n",
      "ClinVar Germline Classification\n",
      "NaN                                             699\n",
      "Uncertain significance                          125\n",
      "Likely benign                                    46\n",
      "Conflicting classifications of pathogenicity     20\n",
      "Benign/Likely benign                              9\n",
      "Benign                                            8\n",
      "Likely pathogenic                                 3\n",
      "Pathogenic                                        3\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Total unique gnomAD entries counted: 913\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def print_comparison_info(file1, sheet1, col1, file2, sheet2, col2):\n",
    "    \"\"\"Prints the details of what is being compared.\"\"\"\n",
    "    print(\"=== Dataset Comparison Information ===\")\n",
    "    print(f\"ClinVar File           : {file1}\")\n",
    "    print(f\"ClinVar Sheet          : {sheet1}\")\n",
    "    print(f\"ClinVar Column         : {col1}\")\n",
    "    print(f\"gnomAD File            : {file2}\")\n",
    "    print(f\"gnomAD Sheet           : {sheet2}\")\n",
    "    print(f\"gnomAD Column          : {col2}\")\n",
    "    print(\"======================================\\n\")\n",
    "\n",
    "# File, sheet, and column metadata\n",
    "clinvar_file = '00_data/clinvar_result.xlsx'\n",
    "clinvar_sheet = 'CDKL5_missense_only'\n",
    "clinvar_column = 'Protein change'\n",
    "\n",
    "gnomad_file = '00_data/gnomAD.xlsx'\n",
    "gnomad_sheet = 'Protein_Change_Converted'\n",
    "gnomad_column = 'Protein change (One-Letter)'\n",
    "\n",
    "# Print comparison info\n",
    "print_comparison_info(clinvar_file, clinvar_sheet, clinvar_column, gnomad_file, gnomad_sheet, gnomad_column)\n",
    "\n",
    "# Load datasets\n",
    "df_clinvar = pd.read_excel(clinvar_file, sheet_name=clinvar_sheet)\n",
    "df_gnomad = pd.read_excel(gnomad_file, sheet_name=gnomad_sheet)\n",
    "\n",
    "# Extract sets of protein changes\n",
    "set_clinvar = set(df_clinvar[clinvar_column].dropna().astype(str))\n",
    "set_gnomad = set(df_gnomad[gnomad_column].dropna().astype(str))\n",
    "\n",
    "# Compare sets\n",
    "common = set_clinvar.intersection(set_gnomad)\n",
    "unique_clinvar = set_clinvar - set_gnomad\n",
    "unique_gnomad = set_gnomad - set_clinvar\n",
    "\n",
    "# Print basic stats\n",
    "print(f\"Number of common entries: {len(common)}\")\n",
    "print(f\"Number of unique entries in ClinVar: {len(unique_clinvar)}\")\n",
    "print(f\"Number of unique entries in gnomAD (set difference): {len(unique_gnomad)}\")\n",
    "\n",
    "# Filter gnomAD for unique entries\n",
    "unique_gnomad_df = df_gnomad[df_gnomad[gnomad_column].isin(unique_gnomad)]\n",
    "\n",
    "# Check duplicates\n",
    "duplicates = unique_gnomad_df.duplicated(subset=[gnomad_column]).sum()\n",
    "print(f\"Number of duplicate entries in unique gnomAD dataset: {duplicates}\")\n",
    "\n",
    "# Drop duplicates\n",
    "unique_gnomad_df = unique_gnomad_df.drop_duplicates(subset=[gnomad_column])\n",
    "\n",
    "# Final consistency check\n",
    "if len(unique_gnomad_df) == len(unique_gnomad):\n",
    "    print(\"No extra entries found. Sets and dataframe filtering consistent.\")\n",
    "else:\n",
    "    print(\"⚠️ Inconsistency found: check for repeated rows.\")\n",
    "\n",
    "# Classification value counts\n",
    "classification_counts = unique_gnomad_df['ClinVar Germline Classification'].value_counts(dropna=False)\n",
    "print(\"\\nValue counts of 'ClinVar Germline Classification' for unique gnomAD entries:\")\n",
    "print(classification_counts)\n",
    "\n",
    "# Total entries\n",
    "print(f\"\\nTotal unique gnomAD entries counted: {classification_counts.sum()}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99bb3817-566d-4915-ad73-edabe7b451f7",
   "metadata": {},
   "source": [
    "# 4 just last one: how many AF pulled in previous clinvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "2eacbf45-0b2d-4b42-a212-3c56cd9b8c34",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clinvar_result_copy.xlsx  clinvar_result.xlsx  gnomAD.xlsx\n",
      "clinvar_result.txt\t  gnomAD.csv\t       protein_change_comparison.xlsx\n"
     ]
    }
   ],
   "source": [
    "!ls 00_data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "005110d1-cccb-40a1-a4a3-801f4455648b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sheet names in the Excel file:\n",
      "['Sheet1', 'CDKL5_rows', 'CDKL5_missense', 'CDKL5_missense_only', 'CDKL5_missense_AF']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the Excel file and list sheet names\n",
    "file_path = '00_data/clinvar_result_copy.xlsx'\n",
    "xls = pd.ExcelFile(file_path)\n",
    "\n",
    "# Print all sheet names\n",
    "print(\"Sheet names in the Excel file:\")\n",
    "print(xls.sheet_names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ce1a4d4b-4d9b-4ec0-8fe5-25a7e73b87a0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns in the 'CDKL5_missense_AF' sheet:\n",
      "['Name', 'Gene(s)', 'Protein change', 'gnomAD_AF', 'Condition(s)', 'Accession', 'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome', 'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID', 'Canonical SPDI', 'Variant type', 'Molecular consequence', 'Germline classification', 'Germline date last evaluated', 'Germline review status', 'Somatic clinical impact', 'Somatic clinical impact date last evaluated', 'Somatic clinical impact review status', 'Oncogenicity classification', 'Oncogenicity date last evaluated', 'Oncogenicity review status', 'Unnamed: 24']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Define the file path and sheet name\n",
    "file_path = '00_data/clinvar_result_copy.xlsx'\n",
    "sheet_name = 'CDKL5_missense_AF'\n",
    "\n",
    "# Load the sheet\n",
    "df = pd.read_excel(file_path, sheet_name=sheet_name)\n",
    "\n",
    "# Show the column names\n",
    "print(\"Columns in the 'CDKL5_missense_AF' sheet:\")\n",
    "print(df.columns.tolist())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "be219b93-766f-42d1-b578-f99be8e8124d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of entries with a value in 'gnomAD_AF': 48\n",
      "Number of empty (NaN) entries in 'gnomAD_AF': 72\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the sheet\n",
    "df = pd.read_excel('00_data/clinvar_result_copy.xlsx', sheet_name='CDKL5_missense_AF')\n",
    "\n",
    "# Count non-empty and empty values in 'gnomAD_AF' column\n",
    "non_empty_count = df['gnomAD_AF'].notna().sum()\n",
    "empty_count = df['gnomAD_AF'].isna().sum()\n",
    "\n",
    "# Print the results\n",
    "print(f\"Number of entries with a value in 'gnomAD_AF': {non_empty_count}\")\n",
    "print(f\"Number of empty (NaN) entries in 'gnomAD_AF': {empty_count}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ac703d8-5340-4f1e-831c-936d99171259",
   "metadata": {},
   "source": [
    "# =============================== #"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a96ac9a8-2c3f-4e5f-9d69-d7806dc0d18f",
   "metadata": {},
   "source": [
    "# 5 The 1000 Genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01f825ae-069f-4f35-bef0-4d357df69f9f",
   "metadata": {},
   "source": [
    "## 5.1 1KGP cdkl5 grch38 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "64d0badf-b187-4066-963c-85af816db154",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>RefGenome</th>\n",
       "      <th>AF</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18425583</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>ENST00000379996.7:c.-275G&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18425593</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000530</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>ENST00000379996.7:c.-265C&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18425637</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.002384</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>ENST00000379996.7:c.-221C&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18425667</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>ENST00000379996.7:c.-191C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18425805</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENST00000379996.7:c.-163+110A&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4475</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18653534</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>RS1</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENST00000379984.4:c.184+3119G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4476</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18653535</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.001324</td>\n",
       "      <td>RS1</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENST00000379984.4:c.184+3118C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4477</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18653548</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>RS1</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENST00000379984.4:c.184+3105G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4478</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18653620</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000795</td>\n",
       "      <td>RS1</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENST00000379984.4:c.184+3033G&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4479</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18653623</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>RS1</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ENST00000379984.4:c.184+3030G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4480 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM       POS REF ALT RefGenome        AF   Gene          Consequence  \\\n",
       "0     chrX  18425583   G   C    GRCh38  0.000265  CDKL5  5_prime_UTR_variant   \n",
       "1     chrX  18425593   C   G    GRCh38  0.000530  CDKL5  5_prime_UTR_variant   \n",
       "2     chrX  18425637   C   G    GRCh38  0.002384  CDKL5  5_prime_UTR_variant   \n",
       "3     chrX  18425667   C   T    GRCh38  0.000265  CDKL5  5_prime_UTR_variant   \n",
       "4     chrX  18425805   A   T    GRCh38  0.000265  CDKL5       intron_variant   \n",
       "...    ...       ...  ..  ..       ...       ...    ...                  ...   \n",
       "4475  chrX  18653534   C   T    GRCh38  0.000265    RS1       intron_variant   \n",
       "4476  chrX  18653535   G   A    GRCh38  0.001324    RS1       intron_variant   \n",
       "4477  chrX  18653548   C   T    GRCh38  0.000265    RS1       intron_variant   \n",
       "4478  chrX  18653620   C   G    GRCh38  0.000795    RS1       intron_variant   \n",
       "4479  chrX  18653623   C   T    GRCh38  0.000265    RS1       intron_variant   \n",
       "\n",
       "                                HGVSc HGVSp  \n",
       "0         ENST00000379996.7:c.-275G>C   NaN  \n",
       "1         ENST00000379996.7:c.-265C>G   NaN  \n",
       "2         ENST00000379996.7:c.-221C>G   NaN  \n",
       "3         ENST00000379996.7:c.-191C>T   NaN  \n",
       "4     ENST00000379996.7:c.-163+110A>T   NaN  \n",
       "...                               ...   ...  \n",
       "4475  ENST00000379984.4:c.184+3119G>A   NaN  \n",
       "4476  ENST00000379984.4:c.184+3118C>T   NaN  \n",
       "4477  ENST00000379984.4:c.184+3105G>A   NaN  \n",
       "4478  ENST00000379984.4:c.184+3033G>C   NaN  \n",
       "4479  ENST00000379984.4:c.184+3030G>A   NaN  \n",
       "\n",
       "[4480 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the dataset\n",
    "file_path = '00_data/1kgp_cdkl5_grch38.xlsx'\n",
    "df_1kgp = pd.read_excel(file_path, sheet_name='Sheet1')\n",
    "\n",
    "# Display the entire dataset\n",
    "display(df_1kgp)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "078bee20-da92-46fe-877f-18d239ea80ff",
   "metadata": {},
   "source": [
    "## 5.2 columns in Sheet1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a40418ca-96fa-4fea-ba28-38eeff23afb6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns in Sheet1:\n",
      "['CHROM', 'POS', 'REF', 'ALT', 'RefGenome', 'AF', 'Gene', 'Consequence', 'HGVSc', 'HGVSp']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the dataset\n",
    "file_path = '00_data/1kgp_cdkl5_grch38.xlsx'\n",
    "df_1kgp = pd.read_excel(file_path, sheet_name='Sheet1')\n",
    "\n",
    "# View column names\n",
    "print(\"Columns in Sheet1:\")\n",
    "print(df_1kgp.columns.tolist())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b719fd18-c123-440d-95aa-2236380e04b3",
   "metadata": {},
   "source": [
    "## 5.3 value counts 'Consequence'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "273c3764-f609-4d6c-959a-81eb4d7b5af0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Value counts for 'Consequence':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>intron_variant</td>\n",
       "      <td>4163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>missense_variant</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>splice_polypyrimidine_tract_variant&amp;intron_var...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>splice_region_variant&amp;5_prime_UTR_variant</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>splice_region_variant&amp;intron_variant</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>splice_donor_5th_base_variant&amp;intron_variant</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Consequence  Count\n",
       "0                                      intron_variant   4163\n",
       "1                               upstream_gene_variant    114\n",
       "2                             downstream_gene_variant     86\n",
       "3                                 3_prime_UTR_variant     56\n",
       "4                                  synonymous_variant     27\n",
       "5                                    missense_variant     19\n",
       "6                                 5_prime_UTR_variant      8\n",
       "7   splice_polypyrimidine_tract_variant&intron_var...      4\n",
       "8           splice_region_variant&5_prime_UTR_variant      1\n",
       "9                splice_region_variant&intron_variant      1\n",
       "10       splice_donor_5th_base_variant&intron_variant      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Total number of entries in 'Consequence': 4480\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the dataset\n",
    "df_1kgp = pd.read_excel('00_data/1kgp_cdkl5_grch38.xlsx', sheet_name='Sheet1')\n",
    "\n",
    "# Get value counts as a DataFrame\n",
    "consequence_counts = df_1kgp['Consequence'].value_counts(dropna=False).rename_axis('Consequence').reset_index(name='Count')\n",
    "\n",
    "# Display the table\n",
    "print(\"Value counts for 'Consequence':\")\n",
    "display(consequence_counts)\n",
    "\n",
    "# Print total count\n",
    "print(f\"\\nTotal number of entries in 'Consequence': {consequence_counts['Count'].sum()}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ad41821-49db-46e9-a6b0-1bb9f4b4fe36",
   "metadata": {},
   "source": [
    "## 5.4 pull out 'missense_variant' from 'Consequence' column of 'Sheet1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "53da45c1-f7ba-409e-8ee3-033a3ec02663",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved 19 'missense_variant' entries to sheet 'missense_only' in 00_data/1kgp_cdkl5_grch38.xlsx\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Define file path\n",
    "file_path = '00_data/1kgp_cdkl5_grch38.xlsx'\n",
    "\n",
    "# Load original Excel file\n",
    "with pd.ExcelFile(file_path) as xls:\n",
    "    df_all = xls.parse('Sheet1')  # Load Sheet1\n",
    "\n",
    "# Filter for missense_variant rows\n",
    "missense_df = df_all[df_all['Consequence'] == 'missense_variant']\n",
    "\n",
    "# Save to a new sheet in the same Excel file\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl', if_sheet_exists='replace') as writer:\n",
    "    missense_df.to_excel(writer, sheet_name='missense_only', index=False)\n",
    "\n",
    "print(f\"Saved {len(missense_df)} 'missense_variant' entries to sheet 'missense_only' in {file_path}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4e562e0-9721-4a5e-95ca-2f292c65c42e",
   "metadata": {},
   "source": [
    "## 5.5 view the new sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4122e132-9672-4ec6-9819-2690129f0b2f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>RefGenome</th>\n",
       "      <th>AF</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18507103</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.7A&gt;T</td>\n",
       "      <td>ENSP00000369325.3:p.Ile3Phe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18507109</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.13A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn5Asp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18575447</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.239G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg80His</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18588118</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.719G&gt;C</td>\n",
       "      <td>ENSP00000369325.3:p.Ser240Thr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18603988</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1064G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg355Gln</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604012</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000530</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1088A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn363Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604366</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1442C&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Ser481Cys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604686</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1762C&gt;T</td>\n",
       "      <td>ENSP00000369325.3:p.Pro588Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18613214</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2215C&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Leu739Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18613214</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2215C&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Leu739Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18613242</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2243A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn748Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18619962</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.016689</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2372A&gt;C</td>\n",
       "      <td>ENSP00000369325.3:p.Gln791Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18628543</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2669G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg890Gln</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18642131</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000795</td>\n",
       "      <td>RS1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379984.4:c.548C&gt;T</td>\n",
       "      <td>ENSP00000369320.3:p.Thr183Ile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18644519</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000530</td>\n",
       "      <td>RS1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379984.4:c.433G&gt;A</td>\n",
       "      <td>ENSP00000369320.3:p.Asp145Asn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18644580</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>RS1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379984.4:c.372G&gt;T</td>\n",
       "      <td>ENSP00000369320.3:p.Gln124His</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18647222</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.015099</td>\n",
       "      <td>RS1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379984.4:c.295A&gt;G</td>\n",
       "      <td>ENSP00000369320.3:p.Asn99Asp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18647284</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>RS1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379984.4:c.233C&gt;T</td>\n",
       "      <td>ENSP00000369320.3:p.Pro78Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18647329</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>RS1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379984.4:c.188G&gt;A</td>\n",
       "      <td>ENSP00000369320.3:p.Cys63Tyr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM       POS REF ALT RefGenome        AF   Gene       Consequence  \\\n",
       "0   chrX  18507103   A   T    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "1   chrX  18507109   A   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "2   chrX  18575447   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "3   chrX  18588118   G   C    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "4   chrX  18603988   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "5   chrX  18604012   A   G    GRCh38  0.000530  CDKL5  missense_variant   \n",
       "6   chrX  18604366   C   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "7   chrX  18604686   C   T    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "8   chrX  18613214   C   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "9   chrX  18613214   C   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "10  chrX  18613242   A   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "11  chrX  18619962   A   C    GRCh38  0.016689  CDKL5  missense_variant   \n",
       "12  chrX  18628543   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "13  chrX  18642131   G   A    GRCh38  0.000795    RS1  missense_variant   \n",
       "14  chrX  18644519   C   T    GRCh38  0.000530    RS1  missense_variant   \n",
       "15  chrX  18644580   C   A    GRCh38  0.000265    RS1  missense_variant   \n",
       "16  chrX  18647222   T   C    GRCh38  0.015099    RS1  missense_variant   \n",
       "17  chrX  18647284   G   A    GRCh38  0.000265    RS1  missense_variant   \n",
       "18  chrX  18647329   C   T    GRCh38  0.000265    RS1  missense_variant   \n",
       "\n",
       "                          HGVSc                          HGVSp  \n",
       "0      ENST00000379989.6:c.7A>T    ENSP00000369325.3:p.Ile3Phe  \n",
       "1     ENST00000379989.6:c.13A>G    ENSP00000369325.3:p.Asn5Asp  \n",
       "2    ENST00000379989.6:c.239G>A   ENSP00000369325.3:p.Arg80His  \n",
       "3    ENST00000379989.6:c.719G>C  ENSP00000369325.3:p.Ser240Thr  \n",
       "4   ENST00000379989.6:c.1064G>A  ENSP00000369325.3:p.Arg355Gln  \n",
       "5   ENST00000379989.6:c.1088A>G  ENSP00000369325.3:p.Asn363Ser  \n",
       "6   ENST00000379989.6:c.1442C>G  ENSP00000369325.3:p.Ser481Cys  \n",
       "7   ENST00000379989.6:c.1762C>T  ENSP00000369325.3:p.Pro588Ser  \n",
       "8   ENST00000379989.6:c.2215C>G  ENSP00000369325.3:p.Leu739Val  \n",
       "9   ENST00000379989.6:c.2215C>G  ENSP00000369325.3:p.Leu739Val  \n",
       "10  ENST00000379989.6:c.2243A>G  ENSP00000369325.3:p.Asn748Ser  \n",
       "11  ENST00000379989.6:c.2372A>C  ENSP00000369325.3:p.Gln791Pro  \n",
       "12  ENST00000379989.6:c.2669G>A  ENSP00000369325.3:p.Arg890Gln  \n",
       "13   ENST00000379984.4:c.548C>T  ENSP00000369320.3:p.Thr183Ile  \n",
       "14   ENST00000379984.4:c.433G>A  ENSP00000369320.3:p.Asp145Asn  \n",
       "15   ENST00000379984.4:c.372G>T  ENSP00000369320.3:p.Gln124His  \n",
       "16   ENST00000379984.4:c.295A>G   ENSP00000369320.3:p.Asn99Asp  \n",
       "17   ENST00000379984.4:c.233C>T   ENSP00000369320.3:p.Pro78Leu  \n",
       "18   ENST00000379984.4:c.188G>A   ENSP00000369320.3:p.Cys63Tyr  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the new sheet 'missense_only' from the Excel file\n",
    "df_missense = pd.read_excel('00_data/1kgp_cdkl5_grch38.xlsx', sheet_name='missense_only')\n",
    "\n",
    "# Display the dataframe in a nice table format\n",
    "display(df_missense)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9345224b-1b58-401f-acc8-e731c565e26c",
   "metadata": {},
   "source": [
    "## 5.6 value count 'Gene'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8dcf177e-2fae-43c1-acba-0f47f8cf7449",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Value counts for 'Gene':\n",
      "Gene\n",
      "CDKL5    13\n",
      "RS1       6\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Total number of entries: 19\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the 'missense_only' sheet\n",
    "df_missense = pd.read_excel('00_data/1kgp_cdkl5_grch38.xlsx', sheet_name='missense_only')\n",
    "\n",
    "# Value counts for 'Gene' column\n",
    "gene_counts = df_missense['Gene'].value_counts(dropna=False)\n",
    "\n",
    "print(\"Value counts for 'Gene':\")\n",
    "print(gene_counts)\n",
    "\n",
    "# Also print the total count (number of rows)\n",
    "print(f\"\\nTotal number of entries: {len(df_missense)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82599ead-71ad-448b-949c-9bd97296f96a",
   "metadata": {},
   "source": [
    "## 5.7 pull out 'CDKL5' from 'Gene'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "27e96d4b-7247-474b-a404-ab8173428462",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sheets in the Excel file:\n",
      "- Sheet1\n",
      "- missense_only\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_path = '00_data/1kgp_cdkl5_grch38.xlsx'\n",
    "\n",
    "# Load Excel file\n",
    "xls = pd.ExcelFile(file_path)\n",
    "\n",
    "# Print sheet names\n",
    "print(\"Sheets in the Excel file:\")\n",
    "for sheet in xls.sheet_names:\n",
    "    print(\"-\", sheet)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ca355dcd-1f6f-45c0-a68b-419fb4f17a84",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# File path\n",
    "file_path = '00_data/1kgp_cdkl5_grch38.xlsx'\n",
    "\n",
    "# Load the missense sheet\n",
    "df = pd.read_excel(file_path, sheet_name='missense_only')\n",
    "\n",
    "# Filter for Gene == 'CDKL5'\n",
    "df_cdkl5 = df[df['Gene'] == 'CDKL5']\n",
    "\n",
    "# Save to the same Excel file, appending as new sheet 'CDKL5_only'\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl', if_sheet_exists='replace') as writer:\n",
    "    df_cdkl5.to_excel(writer, sheet_name='missense_cdkl5_only', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "5eed8277-5e43-45df-9edd-4fa1bd1cb0ac",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>RefGenome</th>\n",
       "      <th>AF</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18507103</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.7A&gt;T</td>\n",
       "      <td>ENSP00000369325.3:p.Ile3Phe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18507109</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.13A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn5Asp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18575447</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.239G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg80His</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18588118</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.719G&gt;C</td>\n",
       "      <td>ENSP00000369325.3:p.Ser240Thr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18603988</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1064G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg355Gln</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604012</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000530</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1088A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn363Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604366</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1442C&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Ser481Cys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604686</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1762C&gt;T</td>\n",
       "      <td>ENSP00000369325.3:p.Pro588Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18613214</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2215C&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Leu739Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18613214</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2215C&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Leu739Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18613242</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2243A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn748Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18619962</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.016689</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2372A&gt;C</td>\n",
       "      <td>ENSP00000369325.3:p.Gln791Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18628543</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2669G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg890Gln</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM       POS REF ALT RefGenome        AF   Gene       Consequence  \\\n",
       "0   chrX  18507103   A   T    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "1   chrX  18507109   A   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "2   chrX  18575447   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "3   chrX  18588118   G   C    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "4   chrX  18603988   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "5   chrX  18604012   A   G    GRCh38  0.000530  CDKL5  missense_variant   \n",
       "6   chrX  18604366   C   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "7   chrX  18604686   C   T    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "8   chrX  18613214   C   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "9   chrX  18613214   C   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "10  chrX  18613242   A   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "11  chrX  18619962   A   C    GRCh38  0.016689  CDKL5  missense_variant   \n",
       "12  chrX  18628543   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "\n",
       "                          HGVSc                          HGVSp  \n",
       "0      ENST00000379989.6:c.7A>T    ENSP00000369325.3:p.Ile3Phe  \n",
       "1     ENST00000379989.6:c.13A>G    ENSP00000369325.3:p.Asn5Asp  \n",
       "2    ENST00000379989.6:c.239G>A   ENSP00000369325.3:p.Arg80His  \n",
       "3    ENST00000379989.6:c.719G>C  ENSP00000369325.3:p.Ser240Thr  \n",
       "4   ENST00000379989.6:c.1064G>A  ENSP00000369325.3:p.Arg355Gln  \n",
       "5   ENST00000379989.6:c.1088A>G  ENSP00000369325.3:p.Asn363Ser  \n",
       "6   ENST00000379989.6:c.1442C>G  ENSP00000369325.3:p.Ser481Cys  \n",
       "7   ENST00000379989.6:c.1762C>T  ENSP00000369325.3:p.Pro588Ser  \n",
       "8   ENST00000379989.6:c.2215C>G  ENSP00000369325.3:p.Leu739Val  \n",
       "9   ENST00000379989.6:c.2215C>G  ENSP00000369325.3:p.Leu739Val  \n",
       "10  ENST00000379989.6:c.2243A>G  ENSP00000369325.3:p.Asn748Ser  \n",
       "11  ENST00000379989.6:c.2372A>C  ENSP00000369325.3:p.Gln791Pro  \n",
       "12  ENST00000379989.6:c.2669G>A  ENSP00000369325.3:p.Arg890Gln  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the new sheet 'missense_only' from the Excel file\n",
    "df_missense = pd.read_excel('00_data/1kgp_cdkl5_grch38.xlsx', sheet_name='missense_cdkl5_only')\n",
    "\n",
    "# Display the dataframe in a nice table format\n",
    "display(df_missense)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7f012bc-3243-4b53-821b-d0f06dd34c53",
   "metadata": {},
   "source": [
    "## 5.8 Check for any duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c160f3a5-1c1a-4b78-bae9-2d40339068d3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicate rows: 1\n",
      "Saved unique entries in sheet 'missense_cdkl5_unique'.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_path = '00_data/1kgp_cdkl5_grch38.xlsx'\n",
    "\n",
    "# Load the Excel file and the sheet with missense_variant data (assuming it's called 'missense_variant')\n",
    "xls = pd.ExcelFile(file_path)\n",
    "df = pd.read_excel(xls, sheet_name='missense_cdkl5_only')\n",
    "\n",
    "# Check for duplicates based on all columns (or choose specific columns if preferred)\n",
    "duplicates = df.duplicated()\n",
    "num_duplicates = duplicates.sum()\n",
    "print(f\"Number of duplicate rows: {num_duplicates}\")\n",
    "\n",
    "# Remove duplicates\n",
    "df_unique = df.drop_duplicates()\n",
    "\n",
    "# Save the unique dataframe as a new sheet\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl', if_sheet_exists='replace') as writer:\n",
    "    df_unique.to_excel(writer, sheet_name='missense_cdkl5_unique', index=False)\n",
    "\n",
    "print(\"Saved unique entries in sheet 'missense_cdkl5_unique'.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4e6124f8-ec75-4cc6-807b-633c325dfda5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>RefGenome</th>\n",
       "      <th>AF</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18507103</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.7A&gt;T</td>\n",
       "      <td>ENSP00000369325.3:p.Ile3Phe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18507109</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.13A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn5Asp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18575447</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.239G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg80His</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18588118</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.719G&gt;C</td>\n",
       "      <td>ENSP00000369325.3:p.Ser240Thr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18603988</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1064G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg355Gln</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604012</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000530</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1088A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn363Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604366</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1442C&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Ser481Cys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604686</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1762C&gt;T</td>\n",
       "      <td>ENSP00000369325.3:p.Pro588Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18613214</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2215C&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Leu739Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18613242</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2243A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn748Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18619962</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.016689</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2372A&gt;C</td>\n",
       "      <td>ENSP00000369325.3:p.Gln791Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18628543</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2669G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg890Gln</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM       POS REF ALT RefGenome        AF   Gene       Consequence  \\\n",
       "0   chrX  18507103   A   T    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "1   chrX  18507109   A   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "2   chrX  18575447   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "3   chrX  18588118   G   C    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "4   chrX  18603988   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "5   chrX  18604012   A   G    GRCh38  0.000530  CDKL5  missense_variant   \n",
       "6   chrX  18604366   C   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "7   chrX  18604686   C   T    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "8   chrX  18613214   C   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "9   chrX  18613242   A   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "10  chrX  18619962   A   C    GRCh38  0.016689  CDKL5  missense_variant   \n",
       "11  chrX  18628543   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "\n",
       "                          HGVSc                          HGVSp  \n",
       "0      ENST00000379989.6:c.7A>T    ENSP00000369325.3:p.Ile3Phe  \n",
       "1     ENST00000379989.6:c.13A>G    ENSP00000369325.3:p.Asn5Asp  \n",
       "2    ENST00000379989.6:c.239G>A   ENSP00000369325.3:p.Arg80His  \n",
       "3    ENST00000379989.6:c.719G>C  ENSP00000369325.3:p.Ser240Thr  \n",
       "4   ENST00000379989.6:c.1064G>A  ENSP00000369325.3:p.Arg355Gln  \n",
       "5   ENST00000379989.6:c.1088A>G  ENSP00000369325.3:p.Asn363Ser  \n",
       "6   ENST00000379989.6:c.1442C>G  ENSP00000369325.3:p.Ser481Cys  \n",
       "7   ENST00000379989.6:c.1762C>T  ENSP00000369325.3:p.Pro588Ser  \n",
       "8   ENST00000379989.6:c.2215C>G  ENSP00000369325.3:p.Leu739Val  \n",
       "9   ENST00000379989.6:c.2243A>G  ENSP00000369325.3:p.Asn748Ser  \n",
       "10  ENST00000379989.6:c.2372A>C  ENSP00000369325.3:p.Gln791Pro  \n",
       "11  ENST00000379989.6:c.2669G>A  ENSP00000369325.3:p.Arg890Gln  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the new sheet 'missense_only' from the Excel file\n",
    "df_missense = pd.read_excel('00_data/1kgp_cdkl5_grch38.xlsx', sheet_name='missense_cdkl5_unique')\n",
    "\n",
    "# Display the dataframe in a nice table format\n",
    "display(df_missense)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a73767f5-6a37-41c4-9cf0-480b4633fd61",
   "metadata": {},
   "source": [
    "## 5.9 missense_cdkl5_unique_prot_change_1L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b2b9d852-f90c-43cc-8c66-06169fcfccbf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved updated dataframe with protein change columns to sheet 'missense_cdkl5_unique_prot_chan'.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# File path and sheet names\n",
    "file_path = '00_data/1kgp_cdkl5_grch38.xlsx'\n",
    "input_sheet = 'missense_cdkl5_unique'\n",
    "output_sheet = 'missense_cdkl5_unique_prot_chan'  # shortened sheet name, max 31 chars\n",
    "\n",
    "# Load the Excel file and data\n",
    "xls = pd.ExcelFile(file_path)\n",
    "df = pd.read_excel(xls, sheet_name=input_sheet)\n",
    "\n",
    "# Extract protein change info from 'HGVSp' column after ':p.'\n",
    "df['Protein change 3L'] = df['HGVSp'].str.extract(r':p\\.(.+)')\n",
    "\n",
    "# Three-letter to one-letter AA code dictionary\n",
    "aa_3to1 = {\n",
    "    'Ala':'A', 'Arg':'R', 'Asn':'N', 'Asp':'D', 'Cys':'C',\n",
    "    'Glu':'E', 'Gln':'Q', 'Gly':'G', 'His':'H', 'Ile':'I',\n",
    "    'Leu':'L', 'Lys':'K', 'Met':'M', 'Phe':'F', 'Pro':'P',\n",
    "    'Ser':'S', 'Thr':'T', 'Trp':'W', 'Tyr':'Y', 'Val':'V',\n",
    "    'Sec':'U', 'Pyl':'O', 'Asx':'B', 'Glx':'Z', 'Xaa':'X'\n",
    "}\n",
    "\n",
    "# Function to convert 3-letter code protein change to 1-letter code\n",
    "import re\n",
    "def convert_3to1(prot_change_3L):\n",
    "    if pd.isna(prot_change_3L):\n",
    "        return prot_change_3L\n",
    "    # Regex pattern to extract: wildtype (3 letters), position (number), mutant (3 letters)\n",
    "    m = re.match(r'([A-Z][a-z]{2})(\\d+)([A-Z][a-z]{2})', prot_change_3L)\n",
    "    if not m:\n",
    "        return prot_change_3L  # return as-is if pattern doesn't match\n",
    "    wt_3, pos, mut_3 = m.groups()\n",
    "    wt_1 = aa_3to1.get(wt_3, wt_3)\n",
    "    mut_1 = aa_3to1.get(mut_3, mut_3)\n",
    "    return f\"{wt_1}{pos}{mut_1}\"\n",
    "\n",
    "# Apply conversion and add new column\n",
    "df['Protein change'] = df['Protein change 3L'].apply(convert_3to1)\n",
    "\n",
    "# Save to the same Excel file, replacing the output sheet if it exists\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl', if_sheet_exists='replace') as writer:\n",
    "    df.to_excel(writer, sheet_name=output_sheet, index=False)\n",
    "\n",
    "print(f\"Saved updated dataframe with protein change columns to sheet '{output_sheet}'.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "5d58b7e9-b460-4f88-90e6-11647d073b31",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>RefGenome</th>\n",
       "      <th>AF</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>Protein change 3L</th>\n",
       "      <th>Protein change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18507103</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.7A&gt;T</td>\n",
       "      <td>ENSP00000369325.3:p.Ile3Phe</td>\n",
       "      <td>Ile3Phe</td>\n",
       "      <td>I3F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18507109</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.13A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn5Asp</td>\n",
       "      <td>Asn5Asp</td>\n",
       "      <td>N5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18575447</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.239G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg80His</td>\n",
       "      <td>Arg80His</td>\n",
       "      <td>R80H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18588118</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.719G&gt;C</td>\n",
       "      <td>ENSP00000369325.3:p.Ser240Thr</td>\n",
       "      <td>Ser240Thr</td>\n",
       "      <td>S240T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18603988</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1064G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg355Gln</td>\n",
       "      <td>Arg355Gln</td>\n",
       "      <td>R355Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604012</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000530</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1088A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn363Ser</td>\n",
       "      <td>Asn363Ser</td>\n",
       "      <td>N363S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604366</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1442C&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Ser481Cys</td>\n",
       "      <td>Ser481Cys</td>\n",
       "      <td>S481C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18604686</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.1762C&gt;T</td>\n",
       "      <td>ENSP00000369325.3:p.Pro588Ser</td>\n",
       "      <td>Pro588Ser</td>\n",
       "      <td>P588S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18613214</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2215C&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Leu739Val</td>\n",
       "      <td>Leu739Val</td>\n",
       "      <td>L739V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18613242</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2243A&gt;G</td>\n",
       "      <td>ENSP00000369325.3:p.Asn748Ser</td>\n",
       "      <td>Asn748Ser</td>\n",
       "      <td>N748S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18619962</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.016689</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2372A&gt;C</td>\n",
       "      <td>ENSP00000369325.3:p.Gln791Pro</td>\n",
       "      <td>Gln791Pro</td>\n",
       "      <td>Q791P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18628543</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ENST00000379989.6:c.2669G&gt;A</td>\n",
       "      <td>ENSP00000369325.3:p.Arg890Gln</td>\n",
       "      <td>Arg890Gln</td>\n",
       "      <td>R890Q</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM       POS REF ALT RefGenome        AF   Gene       Consequence  \\\n",
       "0   chrX  18507103   A   T    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "1   chrX  18507109   A   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "2   chrX  18575447   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "3   chrX  18588118   G   C    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "4   chrX  18603988   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "5   chrX  18604012   A   G    GRCh38  0.000530  CDKL5  missense_variant   \n",
       "6   chrX  18604366   C   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "7   chrX  18604686   C   T    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "8   chrX  18613214   C   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "9   chrX  18613242   A   G    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "10  chrX  18619962   A   C    GRCh38  0.016689  CDKL5  missense_variant   \n",
       "11  chrX  18628543   G   A    GRCh38  0.000265  CDKL5  missense_variant   \n",
       "\n",
       "                          HGVSc                          HGVSp  \\\n",
       "0      ENST00000379989.6:c.7A>T    ENSP00000369325.3:p.Ile3Phe   \n",
       "1     ENST00000379989.6:c.13A>G    ENSP00000369325.3:p.Asn5Asp   \n",
       "2    ENST00000379989.6:c.239G>A   ENSP00000369325.3:p.Arg80His   \n",
       "3    ENST00000379989.6:c.719G>C  ENSP00000369325.3:p.Ser240Thr   \n",
       "4   ENST00000379989.6:c.1064G>A  ENSP00000369325.3:p.Arg355Gln   \n",
       "5   ENST00000379989.6:c.1088A>G  ENSP00000369325.3:p.Asn363Ser   \n",
       "6   ENST00000379989.6:c.1442C>G  ENSP00000369325.3:p.Ser481Cys   \n",
       "7   ENST00000379989.6:c.1762C>T  ENSP00000369325.3:p.Pro588Ser   \n",
       "8   ENST00000379989.6:c.2215C>G  ENSP00000369325.3:p.Leu739Val   \n",
       "9   ENST00000379989.6:c.2243A>G  ENSP00000369325.3:p.Asn748Ser   \n",
       "10  ENST00000379989.6:c.2372A>C  ENSP00000369325.3:p.Gln791Pro   \n",
       "11  ENST00000379989.6:c.2669G>A  ENSP00000369325.3:p.Arg890Gln   \n",
       "\n",
       "   Protein change 3L Protein change  \n",
       "0            Ile3Phe            I3F  \n",
       "1            Asn5Asp            N5D  \n",
       "2           Arg80His           R80H  \n",
       "3          Ser240Thr          S240T  \n",
       "4          Arg355Gln          R355Q  \n",
       "5          Asn363Ser          N363S  \n",
       "6          Ser481Cys          S481C  \n",
       "7          Pro588Ser          P588S  \n",
       "8          Leu739Val          L739V  \n",
       "9          Asn748Ser          N748S  \n",
       "10         Gln791Pro          Q791P  \n",
       "11         Arg890Gln          R890Q  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the new sheet 'missense_only' from the Excel file\n",
    "df_missense = pd.read_excel('00_data/1kgp_cdkl5_grch38.xlsx', sheet_name='missense_cdkl5_unique_prot_chan')\n",
    "\n",
    "# Display the dataframe in a nice table format\n",
    "display(df_missense)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "156d9f45-7785-4440-bcd8-927e63b7a806",
   "metadata": {
    "tags": []
   },
   "source": [
    "# ====================================== #"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a4b45b5-4dde-4310-870b-310ed6506c89",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 6 Hector2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8808c407-18ff-430a-9cac-d2b414989832",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 6.1 view data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "687ad72d-2eec-46c8-b902-1da1cdd802d9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H36R</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>PMID_29264392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T56M</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>PMID_29264392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R65Q</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>PMID_29264392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R80H</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>PMID_29264392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R80L</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>PMID_29264392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>Y262H</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>PMID_29264392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>L271P</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>PMID_29264392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>R285S</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>PMID_29264392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>T288I</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>PMID_29264392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>C291Y</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>PMID_29264392</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mutation                   Consequence         Source\n",
       "0      H36R          Benign/Likely benign  PMID_29264392\n",
       "1      T56M          Benign/Likely benign  PMID_29264392\n",
       "2      R65Q          Benign/Likely benign  PMID_29264392\n",
       "3      R80H          Benign/Likely benign  PMID_29264392\n",
       "4      R80L          Benign/Likely benign  PMID_29264392\n",
       "..      ...                           ...            ...\n",
       "68    Y262H  Pathogenic/Likely pathogenic  PMID_29264392\n",
       "69    L271P  Pathogenic/Likely pathogenic  PMID_29264392\n",
       "70    R285S  Pathogenic/Likely pathogenic  PMID_29264392\n",
       "71    T288I  Pathogenic/Likely pathogenic  PMID_29264392\n",
       "72    C291Y  Pathogenic/Likely pathogenic  PMID_29264392\n",
       "\n",
       "[73 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display\n",
    "import pandas as pd\n",
    "\n",
    "# Load the Excel file and the specified sheet\n",
    "hector_file_path = \"00_data/hector2017.xlsx\"\n",
    "hector_df = pd.read_excel(hector_file_path, sheet_name=\"Sheet1\")\n",
    "\n",
    "# Display the dataframe using IPython display\n",
    "display(hector_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af7c68df-cb10-45ec-8e8e-5aabf65b99f7",
   "metadata": {
    "tags": []
   },
   "source": [
    "# ===================================== #"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd40389a-f64e-4b34-85e5-8982d53dfd9a",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 7 Combined dataset (Clinvar, 1kgp & hector2017)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e5a620d-9613-42b7-809b-57be91f798d0",
   "metadata": {},
   "source": [
    "## 7.1 columns for these datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b61a9842-e4ee-4130-a5f9-88404902a448",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1kgp_cdkl5_grch38.xlsx\tgnomAD.csv\t protein_change_comparison.xlsx\n",
      "clinvar_result.txt\tgnomAD.xlsx\n",
      "clinvar_result.xlsx\thector2017.xlsx\n"
     ]
    }
   ],
   "source": [
    "!ls 00_data/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c49e8b20-d649-407f-ac4e-22676235c57a",
   "metadata": {},
   "source": [
    "## 7.2 sheets name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "14703dd2-9339-4cf9-883f-7cb3e2a86d88",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sheets in clinvar_result.xlsx:\n",
      "['Sheet1', 'Single_Nucleotide_Variant', 'Missense_Variant', 'CDKL5_Condition', 'CDKL5_missense_only']\n",
      "\n",
      "Sheets in 1kgp_cdkl5_grch38.xlsx:\n",
      "['Sheet1', 'missense_only', 'missense_cdkl5_only', 'missense_cdkl5_unique', 'missense_cdkl5_unique_prot_chan']\n",
      "\n",
      "Sheets in hector2017.xlsx:\n",
      "['Sheet1']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# File paths\n",
    "clinvar_path = '00_data/clinvar_result.xlsx'\n",
    "kgp_path = '00_data/1kgp_cdkl5_grch38.xlsx'\n",
    "hector_path = '00_data/hector2017.xlsx'\n",
    "\n",
    "# Load and print sheet names\n",
    "print(\"Sheets in clinvar_result.xlsx:\")\n",
    "print(pd.ExcelFile(clinvar_path).sheet_names)\n",
    "\n",
    "print(\"\\nSheets in 1kgp_cdkl5_grch38.xlsx:\")\n",
    "print(pd.ExcelFile(kgp_path).sheet_names)\n",
    "\n",
    "print(\"\\nSheets in hector2017.xlsx:\")\n",
    "print(pd.ExcelFile(hector_path).sheet_names)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b811287-25c7-49a5-9ed4-594938c2b66f",
   "metadata": {},
   "source": [
    "## 7.3 columns name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "192aa83b-2a27-47ec-8397-a54abd94ef07",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns in 'CDKL5_missense_only' from clinvar_result.xlsx:\n",
      "['Name', 'Gene(s)', 'Protein change', 'Condition(s)', 'Accession', 'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome', 'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID', 'Canonical SPDI', 'Variant type', 'Molecular consequence', 'Germline classification', 'Germline date last evaluated', 'Germline review status', 'Somatic clinical impact', 'Somatic clinical impact date last evaluated', 'Somatic clinical impact review status', 'Oncogenicity classification', 'Oncogenicity date last evaluated', 'Oncogenicity review status', 'Unnamed: 24']\n",
      "\n",
      "Columns in 'missense_cdkl5_unique_prot_chan' from 1kgp_cdkl5_grch38.xlsx:\n",
      "['CHROM', 'POS', 'REF', 'ALT', 'RefGenome', 'AF', 'Gene', 'Consequence', 'HGVSc', 'HGVSp', 'Protein change 3L', 'Protein change']\n",
      "\n",
      "Columns in 'Sheet1' from hector2017.xlsx:\n",
      "['Mutation', 'Consequence', 'Source']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# File paths\n",
    "clinvar_path = '00_data/clinvar_result.xlsx'\n",
    "kgp_path = '00_data/1kgp_cdkl5_grch38.xlsx'\n",
    "hector_path = '00_data/hector2017.xlsx'\n",
    "\n",
    "# Load the specific sheets\n",
    "clinvar_df = pd.read_excel(clinvar_path, sheet_name='CDKL5_missense_only')\n",
    "kgp_df = pd.read_excel(kgp_path, sheet_name='missense_cdkl5_unique_prot_chan')\n",
    "hector_df = pd.read_excel(hector_path, sheet_name='Sheet1')\n",
    "\n",
    "# Display column names\n",
    "print(\"Columns in 'CDKL5_missense_only' from clinvar_result.xlsx:\")\n",
    "print(clinvar_df.columns.tolist())\n",
    "\n",
    "print(\"\\nColumns in 'missense_cdkl5_unique_prot_chan' from 1kgp_cdkl5_grch38.xlsx:\")\n",
    "print(kgp_df.columns.tolist())\n",
    "\n",
    "print(\"\\nColumns in 'Sheet1' from hector2017.xlsx:\")\n",
    "print(hector_df.columns.tolist())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a49f941a-9a49-4b77-be32-262bf2f74058",
   "metadata": {},
   "source": [
    "## 7.4 Unique and common entries among them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "5b127973-3956-447a-bec9-ea64270d4934",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using ClinVar as reference:\n",
      "ClinVar total entries: 120\n",
      "\n",
      "1KGP:\n",
      "  Entries in ClinVar (common): 4\n",
      "  Entries NOT in ClinVar (unique): 8\n",
      "\n",
      "Hector2017:\n",
      "  Entries in ClinVar (common): 43\n",
      "  Entries NOT in ClinVar (unique): 30\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load datasets\n",
    "clinvar_df = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='CDKL5_missense_only')\n",
    "kgp_df = pd.read_excel('00_data/1kgp_cdkl5_grch38.xlsx', sheet_name='missense_cdkl5_unique_prot_chan')\n",
    "hector_df = pd.read_excel('00_data/hector2017.xlsx', sheet_name='Sheet1')\n",
    "\n",
    "# Clean and extract relevant sets\n",
    "clinvar_set = set(clinvar_df['Protein change'].dropna().unique())\n",
    "kgp_set = set(kgp_df['Protein change'].dropna().unique())\n",
    "hector_set = set(hector_df['Mutation'].dropna().unique())\n",
    "\n",
    "# Comparison using ClinVar as reference\n",
    "kgp_common = kgp_set & clinvar_set\n",
    "kgp_unique_wrt_clinvar = kgp_set - clinvar_set\n",
    "\n",
    "hector_common = hector_set & clinvar_set\n",
    "hector_unique_wrt_clinvar = hector_set - clinvar_set\n",
    "\n",
    "# Print counts\n",
    "print(\"Using ClinVar as reference:\")\n",
    "print(f\"ClinVar total entries: {len(clinvar_set)}\\n\")\n",
    "\n",
    "print(f\"1KGP:\")\n",
    "print(f\"  Entries in ClinVar (common): {len(kgp_common)}\")\n",
    "print(f\"  Entries NOT in ClinVar (unique): {len(kgp_unique_wrt_clinvar)}\\n\")\n",
    "\n",
    "print(f\"Hector2017:\")\n",
    "print(f\"  Entries in ClinVar (common): {len(hector_common)}\")\n",
    "print(f\"  Entries NOT in ClinVar (unique): {len(hector_unique_wrt_clinvar)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "378d221f-f85b-4437-91cf-a5fe0826491a",
   "metadata": {},
   "source": [
    "## 7.5 Save a new excel sheet contains Clinvar, 1KGP & Hector2017 (unique), "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a00392cf-10fb-4488-aafb-ba57172b0222",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved output to '00_data/clinvar_kgp_hector_unique_summary.xlsx' with sheets: ClinVar, 1KGP_unique, Hector2017_unique.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load data\n",
    "clinvar_df = pd.read_excel('00_data/clinvar_result.xlsx', sheet_name='CDKL5_missense_only')\n",
    "kgp_df = pd.read_excel('00_data/1kgp_cdkl5_grch38.xlsx', sheet_name='missense_cdkl5_unique_prot_chan')\n",
    "hector_df = pd.read_excel('00_data/hector2017.xlsx', sheet_name='Sheet1')\n",
    "\n",
    "# Extract relevant values\n",
    "clinvar_set = set(clinvar_df['Protein change'].dropna().unique())\n",
    "kgp_set = set(kgp_df['Protein change'].dropna().unique())\n",
    "hector_set = set(hector_df['Mutation'].dropna().unique())\n",
    "\n",
    "# Unique entries w.r.t. ClinVar\n",
    "kgp_unique_wrt_clinvar = kgp_set - clinvar_set\n",
    "hector_unique_wrt_clinvar = hector_set - clinvar_set\n",
    "\n",
    "# Create filtered DataFrames\n",
    "kgp_unique_df = kgp_df[kgp_df['Protein change'].isin(kgp_unique_wrt_clinvar)]\n",
    "hector_unique_df = hector_df[hector_df['Mutation'].isin(hector_unique_wrt_clinvar)]\n",
    "\n",
    "# Save to Excel\n",
    "output_path = '00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "with pd.ExcelWriter(output_path, engine='openpyxl') as writer:\n",
    "    clinvar_df.to_excel(writer, sheet_name='ClinVar', index=False)\n",
    "    kgp_unique_df.to_excel(writer, sheet_name='1KGP_unique', index=False)\n",
    "    hector_unique_df.to_excel(writer, sheet_name='Hector2017_unique', index=False)\n",
    "\n",
    "print(f\"Saved output to '{output_path}' with sheets: ClinVar, 1KGP_unique, Hector2017_unique.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deb6b763-770b-4351-bbcb-3e4f55db3329",
   "metadata": {
    "tags": []
   },
   "source": [
    "# =========================== #"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54825dcd-4007-4d5e-bd35-f4be05d1a6fc",
   "metadata": {},
   "source": [
    "# 8 Manually merge three unique datasets \n",
    "Now i have manually made a new sheet in the same 'clinvar_kgp_hector_unique_summary.xlsx' file and named the sheet: 'clinvar_1kgp_hector'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "979c7d6e-e205-4f26-bc60-63f0ad48f2e1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>Protein change</th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GRCh37Chromosome</th>\n",
       "      <th>GRCh37Location</th>\n",
       "      <th>GRCh38Chromosome</th>\n",
       "      <th>GRCh38Location</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>...</th>\n",
       "      <th>Source</th>\n",
       "      <th>Germline date last evaluated</th>\n",
       "      <th>Germline review status</th>\n",
       "      <th>Somatic clinical impact</th>\n",
       "      <th>Somatic clinical impact date last evaluated</th>\n",
       "      <th>Somatic clinical impact review status</th>\n",
       "      <th>Oncogenicity classification</th>\n",
       "      <th>Oncogenicity date last evaluated</th>\n",
       "      <th>Oncogenicity review status</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.320T&gt;A (p.Val107Asp)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>V107D</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003775037</td>\n",
       "      <td>X</td>\n",
       "      <td>18598005.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18579885.0</td>\n",
       "      <td>3775037.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Clinvar</td>\n",
       "      <td>Feb 28, 2025</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.592G&gt;C (p.Gly198Arg)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>G198R</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003775036</td>\n",
       "      <td>X</td>\n",
       "      <td>18606111.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18587991.0</td>\n",
       "      <td>3775036.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Clinvar</td>\n",
       "      <td>Feb 28, 2025</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.1468G&gt;A (p.Gly490Arg)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>G490R</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003775035</td>\n",
       "      <td>X</td>\n",
       "      <td>18622512.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18604392.0</td>\n",
       "      <td>3775035.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Clinvar</td>\n",
       "      <td>Feb 28, 2025</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.226A&gt;G (p.Lys76Glu)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>K76E</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003602048</td>\n",
       "      <td>X</td>\n",
       "      <td>18593554.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18575434.0</td>\n",
       "      <td>3602048.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Clinvar</td>\n",
       "      <td>Dec 24, 2024</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.104C&gt;T (p.Thr35Ile)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>T35I</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003602046</td>\n",
       "      <td>X</td>\n",
       "      <td>18582601.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18564481.0</td>\n",
       "      <td>3602046.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Clinvar</td>\n",
       "      <td>Dec 24, 2024</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>PMID_29264392</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>G202E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PMID_29264417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>PMID_29264392</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>P209R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PMID_29264418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>PMID_29264392</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>G213R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PMID_29264419</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>PMID_29264392</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>R248W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PMID_29264420</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>PMID_29264392</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>Y262H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PMID_29264421</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>158 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              Name Gene(s) Protein change  \\\n",
       "0     NM_001323289.2(CDKL5):c.320T>A (p.Val107Asp)   CDKL5          V107D   \n",
       "1     NM_001323289.2(CDKL5):c.592G>C (p.Gly198Arg)   CDKL5          G198R   \n",
       "2    NM_001323289.2(CDKL5):c.1468G>A (p.Gly490Arg)   CDKL5          G490R   \n",
       "3      NM_001323289.2(CDKL5):c.226A>G (p.Lys76Glu)   CDKL5           K76E   \n",
       "4      NM_001323289.2(CDKL5):c.104C>T (p.Thr35Ile)   CDKL5           T35I   \n",
       "..                                             ...     ...            ...   \n",
       "153                                  PMID_29264392   CDKL5          G202E   \n",
       "154                                  PMID_29264392   CDKL5          P209R   \n",
       "155                                  PMID_29264392   CDKL5          G213R   \n",
       "156                                  PMID_29264392   CDKL5          R248W   \n",
       "157                                  PMID_29264392   CDKL5          Y262H   \n",
       "\n",
       "       Condition(s)     Accession GRCh37Chromosome  GRCh37Location  \\\n",
       "0    CDKL5 disorder  VCV003775037                X      18598005.0   \n",
       "1    CDKL5 disorder  VCV003775036                X      18606111.0   \n",
       "2    CDKL5 disorder  VCV003775035                X      18622512.0   \n",
       "3    CDKL5 disorder  VCV003602048                X      18593554.0   \n",
       "4    CDKL5 disorder  VCV003602046                X      18582601.0   \n",
       "..              ...           ...              ...             ...   \n",
       "153             NaN           NaN              NaN             NaN   \n",
       "154             NaN           NaN              NaN             NaN   \n",
       "155             NaN           NaN              NaN             NaN   \n",
       "156             NaN           NaN              NaN             NaN   \n",
       "157             NaN           NaN              NaN             NaN   \n",
       "\n",
       "    GRCh38Chromosome  GRCh38Location  VariationID  ...         Source  \\\n",
       "0                  X      18579885.0    3775037.0  ...        Clinvar   \n",
       "1                  X      18587991.0    3775036.0  ...        Clinvar   \n",
       "2                  X      18604392.0    3775035.0  ...        Clinvar   \n",
       "3                  X      18575434.0    3602048.0  ...        Clinvar   \n",
       "4                  X      18564481.0    3602046.0  ...        Clinvar   \n",
       "..               ...             ...          ...  ...            ...   \n",
       "153              NaN             NaN          NaN  ...  PMID_29264417   \n",
       "154              NaN             NaN          NaN  ...  PMID_29264418   \n",
       "155              NaN             NaN          NaN  ...  PMID_29264419   \n",
       "156              NaN             NaN          NaN  ...  PMID_29264420   \n",
       "157              NaN             NaN          NaN  ...  PMID_29264421   \n",
       "\n",
       "    Germline date last evaluated               Germline review status  \\\n",
       "0                   Feb 28, 2025             reviewed by expert panel   \n",
       "1                   Feb 28, 2025             reviewed by expert panel   \n",
       "2                   Feb 28, 2025             reviewed by expert panel   \n",
       "3                   Dec 24, 2024  criteria provided, single submitter   \n",
       "4                   Dec 24, 2024  criteria provided, single submitter   \n",
       "..                           ...                                  ...   \n",
       "153                          NaN                                  NaN   \n",
       "154                          NaN                                  NaN   \n",
       "155                          NaN                                  NaN   \n",
       "156                          NaN                                  NaN   \n",
       "157                          NaN                                  NaN   \n",
       "\n",
       "    Somatic clinical impact Somatic clinical impact date last evaluated  \\\n",
       "0                       NaN                                         NaN   \n",
       "1                       NaN                                         NaN   \n",
       "2                       NaN                                         NaN   \n",
       "3                       NaN                                         NaN   \n",
       "4                       NaN                                         NaN   \n",
       "..                      ...                                         ...   \n",
       "153                     NaN                                         NaN   \n",
       "154                     NaN                                         NaN   \n",
       "155                     NaN                                         NaN   \n",
       "156                     NaN                                         NaN   \n",
       "157                     NaN                                         NaN   \n",
       "\n",
       "    Somatic clinical impact review status Oncogenicity classification  \\\n",
       "0                                     NaN                         NaN   \n",
       "1                                     NaN                         NaN   \n",
       "2                                     NaN                         NaN   \n",
       "3                                     NaN                         NaN   \n",
       "4                                     NaN                         NaN   \n",
       "..                                    ...                         ...   \n",
       "153                                   NaN                         NaN   \n",
       "154                                   NaN                         NaN   \n",
       "155                                   NaN                         NaN   \n",
       "156                                   NaN                         NaN   \n",
       "157                                   NaN                         NaN   \n",
       "\n",
       "    Oncogenicity date last evaluated Oncogenicity review status  Unnamed: 24  \n",
       "0                                NaN                        NaN          NaN  \n",
       "1                                NaN                        NaN          NaN  \n",
       "2                                NaN                        NaN          NaN  \n",
       "3                                NaN                        NaN          NaN  \n",
       "4                                NaN                        NaN          NaN  \n",
       "..                               ...                        ...          ...  \n",
       "153                              NaN                        NaN          NaN  \n",
       "154                              NaN                        NaN          NaN  \n",
       "155                              NaN                        NaN          NaN  \n",
       "156                              NaN                        NaN          NaN  \n",
       "157                              NaN                        NaN          NaN  \n",
       "\n",
       "[158 rows x 26 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Load the merged sheet\n",
    "merged_df = pd.read_excel('00_data/clinvar_kgp_hector_unique_summary.xlsx', sheet_name='clinvar_1kgp_hector')\n",
    "\n",
    "# Display the dataframe\n",
    "display(merged_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc40632b-18d1-4468-a6e3-b9e13ea09447",
   "metadata": {},
   "source": [
    "## if any duplication in the 'Protein change' column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "a879f959-b55d-44c0-bb71-b30b746c3be4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns in 'clinvar_1kgp_hector':\n",
      "['Name', 'Gene(s)', 'Protein change', 'Condition(s)', 'Accession', 'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome', 'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID', 'Canonical SPDI', 'Variant type', 'Molecular consequence', 'Germline classification', 'Source', 'Germline date last evaluated', 'Germline review status', 'Somatic clinical impact', 'Somatic clinical impact date last evaluated', 'Somatic clinical impact review status', 'Oncogenicity classification', 'Oncogenicity date last evaluated', 'Oncogenicity review status', 'Unnamed: 24']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the sheet\n",
    "merged_df = pd.read_excel('00_data/clinvar_kgp_hector_unique_summary.xlsx', sheet_name='clinvar_1kgp_hector')\n",
    "\n",
    "# Print column names\n",
    "print(\"Columns in 'clinvar_1kgp_hector':\")\n",
    "print(merged_df.columns.tolist())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "3a0fbc49-2a36-4cbb-9af7-e6a6197bca72",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicate entries in 'Protein change': 0\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the sheet\n",
    "merged_df = pd.read_excel('00_data/clinvar_kgp_hector_unique_summary.xlsx', sheet_name='clinvar_1kgp_hector')\n",
    "\n",
    "# Check duplicates in 'Protein change' column\n",
    "duplicates = merged_df['Protein change'].duplicated()\n",
    "num_duplicates = duplicates.sum()\n",
    "\n",
    "print(f\"Number of duplicate entries in 'Protein change': {num_duplicates}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42b2e1f9-2855-421e-99e9-c6b52385f875",
   "metadata": {},
   "source": [
    "## get allele frequency from gnomAD"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37d2504d-de9f-4127-9c28-ad6a3e4b3c7e",
   "metadata": {},
   "source": [
    "## first view the gnomAD sheets & columns \n",
    "file: 00_data/gnomAD.xlsx\n",
    "sheet: 'Protein_Change_Converted'\n",
    "Column: 'Protein change (One-Letter)'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "e85d2ce8-6e54-4494-8a7c-f7e0709cf84b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1kgp_cdkl5_grch38.xlsx\t\t\tgnomAD.csv\n",
      "clinvar_kgp_hector_unique_summary.xlsx\tgnomAD.xlsx\n",
      "clinvar_result.txt\t\t\thector2017.xlsx\n",
      "clinvar_result.xlsx\t\t\tprotein_change_comparison.xlsx\n"
     ]
    }
   ],
   "source": [
    "!ls 00_data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "2e0c038f-6277-45a0-a649-06754997dd68",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sheets in gnomAD.xlsx:\n",
      "['Sheet1', 'Missense_Variant', 'Protein_Change_Converted', 'ClinVar_Classified_Only']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_path = '00_data/gnomAD.xlsx'\n",
    "xls = pd.ExcelFile(file_path)\n",
    "print(\"Sheets in gnomAD.xlsx:\")\n",
    "print(xls.sheet_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "46f533b4-d608-4888-a951-82c132b258a6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns in 'Protein_Change_Converted':\n",
      "['gnomAD ID', 'Chromosome', 'Position', 'rsIDs', 'Reference', 'Alternate', 'Source', 'Filters - exomes', 'Filters - genomes', 'Transcript', 'HGVS Consequence', 'Protein Consequence', 'Transcript Consequence', 'VEP Annotation', 'ClinVar Germline Classification', 'ClinVar Variation ID', 'Flags', 'Allele Count', 'Allele Number', 'Allele Frequency', 'Homozygote Count', 'Hemizygote Count', 'Filters - joint', 'GroupMax FAF group', 'GroupMax FAF frequency', 'cadd', 'revel_max', 'spliceai_ds_max', 'pangolin_largest_ds', 'phylop', 'sift_max', 'polyphen_max', 'Allele Count African/African American', 'Allele Number African/African American', 'Homozygote Count African/African American', 'Hemizygote Count African/African American', 'Allele Count Admixed American', 'Allele Number Admixed American', 'Homozygote Count Admixed American', 'Hemizygote Count Admixed American', 'Allele Count Ashkenazi Jewish', 'Allele Number Ashkenazi Jewish', 'Homozygote Count Ashkenazi Jewish', 'Hemizygote Count Ashkenazi Jewish', 'Allele Count East Asian', 'Allele Number East Asian', 'Homozygote Count East Asian', 'Hemizygote Count East Asian', 'Allele Count European (Finnish)', 'Allele Number European (Finnish)', 'Homozygote Count European (Finnish)', 'Hemizygote Count European (Finnish)', 'Allele Count Middle Eastern', 'Allele Number Middle Eastern', 'Homozygote Count Middle Eastern', 'Hemizygote Count Middle Eastern', 'Allele Count European (non-Finnish)', 'Allele Number European (non-Finnish)', 'Homozygote Count European (non-Finnish)', 'Hemizygote Count European (non-Finnish)', 'Allele Count Amish', 'Allele Number Amish', 'Homozygote Count Amish', 'Hemizygote Count Amish', 'Allele Count South Asian', 'Allele Number South Asian', 'Homozygote Count South Asian', 'Hemizygote Count South Asian', 'Allele Count Remaining', 'Allele Number Remaining', 'Homozygote Count Remaining', 'Hemizygote Count Remaining', 'Protein Consequence (Three-Letter)', 'Protein change (One-Letter)']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_path = '00_data/gnomAD.xlsx'\n",
    "sheet_name = 'Protein_Change_Converted'\n",
    "\n",
    "df = pd.read_excel(file_path, sheet_name=sheet_name)\n",
    "print(f\"Columns in '{sheet_name}':\")\n",
    "print(df.columns.tolist())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "203ea1d5-e265-480c-aa74-4ed478d34c54",
   "metadata": {},
   "source": [
    "## pull gnomAD allele frequency\n",
    "data: clinvar\n",
    "file: 00_data/clinvar_kgp_hector_unique_summary.xlsx\n",
    "sheet: 'clinvar_1kgp_hector'\n",
    "Protein change column: 'Protein change'\n",
    "allele frequency column: 'Allele Frequency'\n",
    "\n",
    "data: gnomAD\n",
    "file: 00_data/gnomAD.xlsx\n",
    "sheet: 'Protein_Change_Converted'\n",
    "Protein change Column: 'Protein change (One-Letter)'\n",
    "\n",
    "So what i want is:\n",
    "read clinvar data, sheet and protein change column\n",
    "then match with the entries found in 'Protein_Change_Converted' sheet of gnomAD data with the column 'Protien change (One-Letter)'\n",
    "\n",
    "and pull out corresponding 'Allele Frequency'\n",
    "and save it as a new sheet in the clinvar excel file with proper name\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "f3a34bd8-e51a-4a63-b19d-6e5cd9ad89fa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved merged data with gnomAD Allele Frequency to sheet 'clinvar_1kgp_hector_gaf'.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# File paths\n",
    "clinvar_file = '00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "gnomad_file = '00_data/gnomAD.xlsx'\n",
    "\n",
    "# Read ClinVar data\n",
    "clinvar_df = pd.read_excel(clinvar_file, sheet_name='clinvar_1kgp_hector')\n",
    "# Column with protein changes (one-letter code)\n",
    "clinvar_col = 'Protein change'\n",
    "\n",
    "# Read gnomAD data\n",
    "gnomad_df = pd.read_excel(gnomad_file, sheet_name='Protein_Change_Converted')\n",
    "gnomad_protein_col = 'Protein change (One-Letter)'\n",
    "gnomad_af_col = 'Allele Frequency'\n",
    "\n",
    "# Merge ClinVar with gnomAD on protein change (one-letter code)\n",
    "merged_df = pd.merge(\n",
    "    clinvar_df,\n",
    "    gnomad_df[[gnomad_protein_col, gnomad_af_col]],\n",
    "    left_on=clinvar_col,\n",
    "    right_on=gnomad_protein_col,\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "# Rename the gnomAD Allele Frequency column for clarity\n",
    "merged_df.rename(columns={gnomad_af_col: 'gnomAD Allele Frequency'}, inplace=True)\n",
    "\n",
    "# Drop the redundant gnomAD protein change column after merge\n",
    "merged_df.drop(columns=[gnomad_protein_col], inplace=True)\n",
    "\n",
    "# Save to a new sheet in the same ClinVar Excel file\n",
    "with pd.ExcelWriter(clinvar_file, mode='a', engine='openpyxl', if_sheet_exists='replace') as writer:\n",
    "    merged_df.to_excel(writer, sheet_name='clinvar_1kgp_hector_gaf', index=False)\n",
    "\n",
    "print(\"Saved merged data with gnomAD Allele Frequency to sheet 'clinvar_1kgp_hector_gaf'.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f0917d7-5f2f-42bf-b068-04028e29398c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## view the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "674e11c2-ce14-408c-8627-f01dd82872b5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>Protein change</th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GRCh37Chromosome</th>\n",
       "      <th>GRCh37Location</th>\n",
       "      <th>GRCh38Chromosome</th>\n",
       "      <th>GRCh38Location</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>...</th>\n",
       "      <th>Germline date last evaluated</th>\n",
       "      <th>Germline review status</th>\n",
       "      <th>Somatic clinical impact</th>\n",
       "      <th>Somatic clinical impact date last evaluated</th>\n",
       "      <th>Somatic clinical impact review status</th>\n",
       "      <th>Oncogenicity classification</th>\n",
       "      <th>Oncogenicity date last evaluated</th>\n",
       "      <th>Oncogenicity review status</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "      <th>gnomAD Allele Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.320T&gt;A (p.Val107Asp)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>V107D</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003775037</td>\n",
       "      <td>X</td>\n",
       "      <td>18598005.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18579885.0</td>\n",
       "      <td>3775037.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Feb 28, 2025</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.592G&gt;C (p.Gly198Arg)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>G198R</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003775036</td>\n",
       "      <td>X</td>\n",
       "      <td>18606111.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18587991.0</td>\n",
       "      <td>3775036.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Feb 28, 2025</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.1468G&gt;A (p.Gly490Arg)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>G490R</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003775035</td>\n",
       "      <td>X</td>\n",
       "      <td>18622512.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18604392.0</td>\n",
       "      <td>3775035.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Feb 28, 2025</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.226A&gt;G (p.Lys76Glu)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>K76E</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003602048</td>\n",
       "      <td>X</td>\n",
       "      <td>18593554.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18575434.0</td>\n",
       "      <td>3602048.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Dec 24, 2024</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.104C&gt;T (p.Thr35Ile)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>T35I</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003602046</td>\n",
       "      <td>X</td>\n",
       "      <td>18582601.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18564481.0</td>\n",
       "      <td>3602046.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Dec 24, 2024</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.551T&gt;A (p.Leu184His)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>L184H</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003602045</td>\n",
       "      <td>X</td>\n",
       "      <td>18602470.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18584350.0</td>\n",
       "      <td>3602045.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Dec 23, 2024</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.530A&gt;C (p.Tyr177Ser)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>Y177S</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003602044</td>\n",
       "      <td>X</td>\n",
       "      <td>18602449.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18584329.0</td>\n",
       "      <td>3602044.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Dec 23, 2024</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.464G&gt;A (p.Gly155Asp)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>G155D</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003602043</td>\n",
       "      <td>X</td>\n",
       "      <td>18602383.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18584263.0</td>\n",
       "      <td>3602043.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Dec 23, 2024</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.607G&gt;A (p.Glu203Lys)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>E203K</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003602042</td>\n",
       "      <td>X</td>\n",
       "      <td>18606126.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18588006.0</td>\n",
       "      <td>3602042.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Dec 20, 2024</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NM_001323289.2(CDKL5):c.521C&gt;A (p.Thr174Asn)</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>T174N</td>\n",
       "      <td>CDKL5 disorder</td>\n",
       "      <td>VCV003602018</td>\n",
       "      <td>X</td>\n",
       "      <td>18602440.0</td>\n",
       "      <td>X</td>\n",
       "      <td>18584320.0</td>\n",
       "      <td>3602018.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Dec 17, 2024</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Name Gene(s) Protein change  \\\n",
       "0   NM_001323289.2(CDKL5):c.320T>A (p.Val107Asp)   CDKL5          V107D   \n",
       "1   NM_001323289.2(CDKL5):c.592G>C (p.Gly198Arg)   CDKL5          G198R   \n",
       "2  NM_001323289.2(CDKL5):c.1468G>A (p.Gly490Arg)   CDKL5          G490R   \n",
       "3    NM_001323289.2(CDKL5):c.226A>G (p.Lys76Glu)   CDKL5           K76E   \n",
       "4    NM_001323289.2(CDKL5):c.104C>T (p.Thr35Ile)   CDKL5           T35I   \n",
       "5   NM_001323289.2(CDKL5):c.551T>A (p.Leu184His)   CDKL5          L184H   \n",
       "6   NM_001323289.2(CDKL5):c.530A>C (p.Tyr177Ser)   CDKL5          Y177S   \n",
       "7   NM_001323289.2(CDKL5):c.464G>A (p.Gly155Asp)   CDKL5          G155D   \n",
       "8   NM_001323289.2(CDKL5):c.607G>A (p.Glu203Lys)   CDKL5          E203K   \n",
       "9   NM_001323289.2(CDKL5):c.521C>A (p.Thr174Asn)   CDKL5          T174N   \n",
       "\n",
       "     Condition(s)     Accession GRCh37Chromosome  GRCh37Location  \\\n",
       "0  CDKL5 disorder  VCV003775037                X      18598005.0   \n",
       "1  CDKL5 disorder  VCV003775036                X      18606111.0   \n",
       "2  CDKL5 disorder  VCV003775035                X      18622512.0   \n",
       "3  CDKL5 disorder  VCV003602048                X      18593554.0   \n",
       "4  CDKL5 disorder  VCV003602046                X      18582601.0   \n",
       "5  CDKL5 disorder  VCV003602045                X      18602470.0   \n",
       "6  CDKL5 disorder  VCV003602044                X      18602449.0   \n",
       "7  CDKL5 disorder  VCV003602043                X      18602383.0   \n",
       "8  CDKL5 disorder  VCV003602042                X      18606126.0   \n",
       "9  CDKL5 disorder  VCV003602018                X      18602440.0   \n",
       "\n",
       "  GRCh38Chromosome  GRCh38Location  VariationID  ...  \\\n",
       "0                X      18579885.0    3775037.0  ...   \n",
       "1                X      18587991.0    3775036.0  ...   \n",
       "2                X      18604392.0    3775035.0  ...   \n",
       "3                X      18575434.0    3602048.0  ...   \n",
       "4                X      18564481.0    3602046.0  ...   \n",
       "5                X      18584350.0    3602045.0  ...   \n",
       "6                X      18584329.0    3602044.0  ...   \n",
       "7                X      18584263.0    3602043.0  ...   \n",
       "8                X      18588006.0    3602042.0  ...   \n",
       "9                X      18584320.0    3602018.0  ...   \n",
       "\n",
       "   Germline date last evaluated               Germline review status  \\\n",
       "0                  Feb 28, 2025             reviewed by expert panel   \n",
       "1                  Feb 28, 2025             reviewed by expert panel   \n",
       "2                  Feb 28, 2025             reviewed by expert panel   \n",
       "3                  Dec 24, 2024  criteria provided, single submitter   \n",
       "4                  Dec 24, 2024  criteria provided, single submitter   \n",
       "5                  Dec 23, 2024  criteria provided, single submitter   \n",
       "6                  Dec 23, 2024  criteria provided, single submitter   \n",
       "7                  Dec 23, 2024  criteria provided, single submitter   \n",
       "8                  Dec 20, 2024  criteria provided, single submitter   \n",
       "9                  Dec 17, 2024  criteria provided, single submitter   \n",
       "\n",
       "  Somatic clinical impact Somatic clinical impact date last evaluated  \\\n",
       "0                     NaN                                         NaN   \n",
       "1                     NaN                                         NaN   \n",
       "2                     NaN                                         NaN   \n",
       "3                     NaN                                         NaN   \n",
       "4                     NaN                                         NaN   \n",
       "5                     NaN                                         NaN   \n",
       "6                     NaN                                         NaN   \n",
       "7                     NaN                                         NaN   \n",
       "8                     NaN                                         NaN   \n",
       "9                     NaN                                         NaN   \n",
       "\n",
       "  Somatic clinical impact review status Oncogenicity classification  \\\n",
       "0                                   NaN                         NaN   \n",
       "1                                   NaN                         NaN   \n",
       "2                                   NaN                         NaN   \n",
       "3                                   NaN                         NaN   \n",
       "4                                   NaN                         NaN   \n",
       "5                                   NaN                         NaN   \n",
       "6                                   NaN                         NaN   \n",
       "7                                   NaN                         NaN   \n",
       "8                                   NaN                         NaN   \n",
       "9                                   NaN                         NaN   \n",
       "\n",
       "  Oncogenicity date last evaluated Oncogenicity review status Unnamed: 24  \\\n",
       "0                              NaN                        NaN         NaN   \n",
       "1                              NaN                        NaN         NaN   \n",
       "2                              NaN                        NaN         NaN   \n",
       "3                              NaN                        NaN         NaN   \n",
       "4                              NaN                        NaN         NaN   \n",
       "5                              NaN                        NaN         NaN   \n",
       "6                              NaN                        NaN         NaN   \n",
       "7                              NaN                        NaN         NaN   \n",
       "8                              NaN                        NaN         NaN   \n",
       "9                              NaN                        NaN         NaN   \n",
       "\n",
       "   gnomAD Allele Frequency  \n",
       "0                      NaN  \n",
       "1                      NaN  \n",
       "2                 0.000002  \n",
       "3                      NaN  \n",
       "4                 0.000000  \n",
       "5                      NaN  \n",
       "6                      NaN  \n",
       "7                      NaN  \n",
       "8                      NaN  \n",
       "9                 0.000000  \n",
       "\n",
       "[10 rows x 27 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clinvar_file = '00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "\n",
    "# Load the new sheet\n",
    "df_new = pd.read_excel(clinvar_file, sheet_name='clinvar_1kgp_hector_gaf')\n",
    "\n",
    "# Display the dataframe nicely in IPython\n",
    "from IPython.display import display\n",
    "display(df_new.head(10))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eab6b99-ff72-46ee-b3cc-51e07c7a262b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## how many gnomAD AF and NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "dc4c3d8a-f9e5-4bca-abfa-52682d9dac14",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of non-NaN entries: 73\n",
      "Number of NaN entries: 85\n"
     ]
    }
   ],
   "source": [
    "# Count non-NaN values\n",
    "non_nan_count = df_new['gnomAD Allele Frequency'].notna().sum()\n",
    "\n",
    "# Count NaN values\n",
    "nan_count = df_new['gnomAD Allele Frequency'].isna().sum()\n",
    "\n",
    "print(f\"Number of non-NaN entries: {non_nan_count}\")\n",
    "print(f\"Number of NaN entries: {nan_count}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "80494f6a-7a10-4bf3-a995-f44e129d0bc7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Name', 'Gene(s)', 'Protein change', 'Condition(s)', 'Accession', 'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome', 'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID', 'Canonical SPDI', 'Variant type', 'Molecular consequence', 'Germline classification', 'Source', 'Germline date last evaluated', 'Germline review status', 'Somatic clinical impact', 'Somatic clinical impact date last evaluated', 'Somatic clinical impact review status', 'Oncogenicity classification', 'Oncogenicity date last evaluated', 'Oncogenicity review status', 'Unnamed: 24', 'gnomAD Allele Frequency']\n"
     ]
    }
   ],
   "source": [
    "print(df_new.columns.tolist())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e266a952-3dc7-49bf-aafd-79b0bc0a30ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "## value counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "b0390478-4330-4696-90a6-f25bd6a77bab",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Value counts for 'Variant type':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant type</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>single nucleotide variant</th>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Count\n",
       "Variant type                    \n",
       "single nucleotide variant    120\n",
       "NaN                           38"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total entries in 'Variant type': 158\n",
      "\n",
      "Value counts for 'Molecular consequence':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Molecular consequence</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>missense variant</th>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Count\n",
       "Molecular consequence       \n",
       "missense variant         128\n",
       "NaN                       30"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total entries in 'Molecular consequence': 158\n",
      "\n",
      "Value counts for 'Germline classification':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Germline classification</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Uncertain significance</th>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic/Likely pathogenic</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Likely pathogenic</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Benign</th>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Benign/Likely benign</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Conflicting classifications of pathogenicity</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Likely benign</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              Count\n",
       "Germline classification                            \n",
       "Uncertain significance                           43\n",
       "Pathogenic/Likely pathogenic                     26\n",
       "Likely pathogenic                                22\n",
       "Benign                                           20\n",
       "Benign/Likely benign                             15\n",
       "Conflicting classifications of pathogenicity     13\n",
       "Likely benign                                    10\n",
       "Pathogenic                                        9"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total entries in 'Germline classification': 158\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from IPython.display import display\n",
    "\n",
    "for col in ['Variant type', 'Molecular consequence', 'Germline classification']:\n",
    "    print(f\"Value counts for '{col}':\")\n",
    "    counts = df_new[col].value_counts(dropna=False)\n",
    "    display(counts.to_frame(name='Count'))\n",
    "    print(f\"Total entries in '{col}': {len(df_new[col])}\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "764a0cb2-4b86-42db-93f8-ed765f8c1524",
   "metadata": {},
   "source": [
    "## 'Protein change' column sanity check with Uniprot seq\n",
    "'Protein change' column represents \n",
    "'one_letter_code_of_wild_type_amino_acid-position-one_letter_code_of_mutant_amino_acid''\n",
    "for example 'V107D'\n",
    "\n",
    "here \n",
    "V:one_letter_code_of_wild_type_amino_acid\n",
    "107:position\n",
    "D:one_letter_code_of_mutant_amino_acid\n",
    "\n",
    "now cdkl5 uniprot id is: 'O76039'\n",
    "pull out amino acid sequence using the uniprot id\n",
    "\n",
    "then match with the \n",
    "one_letter_code_of_wild_type_amino_acid\n",
    "and \n",
    "position\n",
    "\n",
    "and print summary of the result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "71b9da1b-0dc2-4f9b-a188-173f7d22d4a3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of CDKL5 sequence: 960\n",
      "match_status\n",
      "Match                156\n",
      "Mismatch (Seq: H)      1\n",
      "Mismatch (Seq: D)      1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Mismatches:\n",
      "    Protein change wt_aa  pos mut_aa       match_status\n",
      "147          Y145H     Y  145      H  Mismatch (Seq: H)\n",
      "149          A153G     A  153      G  Mismatch (Seq: D)\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "# Step 1: Fetch CDKL5 sequence from UniProt (FASTA format)\n",
    "uniprot_id = \"O76039\"\n",
    "url = f\"https://www.uniprot.org/uniprot/{uniprot_id}.fasta\"\n",
    "response = requests.get(url)\n",
    "fasta_text = response.text\n",
    "\n",
    "# Parse FASTA sequence (skip header line)\n",
    "lines = fasta_text.split('\\n')\n",
    "sequence = ''.join(line.strip() for line in lines if not line.startswith('>'))\n",
    "\n",
    "print(f\"Length of CDKL5 sequence: {len(sequence)}\")\n",
    "\n",
    "# Step 2: Load your data with the 'Protein change' column\n",
    "file_path = '00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "sheet_name = 'clinvar_1kgp_hector_gaf'\n",
    "\n",
    "df = pd.read_excel(file_path, sheet_name=sheet_name)\n",
    "\n",
    "# Step 3: Extract wild type AA, position, mutant AA from 'Protein change'\n",
    "def parse_protein_change(pc):\n",
    "    # Example format: V107D\n",
    "    if isinstance(pc, str) and len(pc) >= 3:\n",
    "        wt = pc[0]\n",
    "        mut = pc[-1]\n",
    "        pos = pc[1:-1]\n",
    "        if pos.isdigit():\n",
    "            return wt, int(pos), mut\n",
    "    return None, None, None\n",
    "\n",
    "df[['wt_aa', 'pos', 'mut_aa']] = df['Protein change'].apply(\n",
    "    lambda x: pd.Series(parse_protein_change(x))\n",
    ")\n",
    "\n",
    "# Step 4: Check match against UniProt sequence\n",
    "def check_match(row):\n",
    "    wt = row['wt_aa']\n",
    "    pos = row['pos']\n",
    "    if wt is None or pos is None or pos > len(sequence) or pos < 1:\n",
    "        return 'Invalid'\n",
    "    # UniProt sequence is 1-based index, python 0-based\n",
    "    seq_aa = sequence[pos-1]\n",
    "    return 'Match' if seq_aa == wt else f'Mismatch (Seq: {seq_aa})'\n",
    "\n",
    "df['match_status'] = df.apply(check_match, axis=1)\n",
    "\n",
    "# Step 5: Summary\n",
    "summary = df['match_status'].value_counts()\n",
    "print(summary)\n",
    "\n",
    "# Optional: print mismatch rows to inspect\n",
    "print(\"\\nMismatches:\")\n",
    "print(df[df['match_status'].str.contains('Mismatch')][['Protein change', 'wt_aa', 'pos', 'mut_aa', 'match_status']])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "3fac6d12-e93d-4fef-803f-9dd7cba22225",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Final dataset saved to sheet 'clinvar_1kgp_hector_gaf_final' in '00_data/clinvar_kgp_hector_unique_summary.xlsx'.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# File path and sheet names\n",
    "file_path = '00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "input_sheet = 'clinvar_1kgp_hector_gaf'\n",
    "output_sheet = 'clinvar_1kgp_hector_gaf_final'\n",
    "\n",
    "# Load the data\n",
    "df = pd.read_excel(file_path, sheet_name=input_sheet)\n",
    "\n",
    "# Remove mismatches based on exact 'Protein change'\n",
    "mismatches = ['Y145H', 'A153G']\n",
    "df_filtered = df[~df['Protein change'].isin(mismatches)].copy()\n",
    "\n",
    "# Copy 'Protein change' column to new column 'Mutation'\n",
    "df_filtered['Mutation'] = df_filtered['Protein change']\n",
    "\n",
    "# Extract wild, position, and mutant from 'Mutation'\n",
    "df_filtered['wild'] = df_filtered['Mutation'].str.extract(r'^([A-Z])')\n",
    "df_filtered['position'] = df_filtered['Mutation'].str.extract(r'(\\d+)').astype(int)\n",
    "df_filtered['mutant'] = df_filtered['Mutation'].str.extract(r'([A-Z])$')\n",
    "\n",
    "# Sort by position\n",
    "df_filtered = df_filtered.sort_values(by='position')\n",
    "\n",
    "# Save to a new sheet\n",
    "with pd.ExcelWriter(file_path, mode='a', engine='openpyxl', if_sheet_exists='replace') as writer:\n",
    "    df_filtered.to_excel(writer, sheet_name=output_sheet, index=False)\n",
    "\n",
    "print(f\"✅ Final dataset saved to sheet '{output_sheet}' in '{file_path}'.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26bdebb8-0350-4b62-a075-4168bbee4606",
   "metadata": {},
   "outputs": [],
   "source": [
    "## columns in final sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "e34d7187-013f-4d0f-823b-2197eab8babe",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Columns in the final sheet:\n",
      "['Name', 'Gene(s)', 'Protein change', 'Condition(s)', 'Accession', 'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome', 'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID', 'Canonical SPDI', 'Variant type', 'Molecular consequence', 'Germline classification', 'Source', 'Germline date last evaluated', 'Germline review status', 'Somatic clinical impact', 'Somatic clinical impact date last evaluated', 'Somatic clinical impact review status', 'Oncogenicity classification', 'Oncogenicity date last evaluated', 'Oncogenicity review status', 'Unnamed: 24', 'gnomAD Allele Frequency', 'Mutation', 'wild', 'position', 'mutant']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# File path and sheet name\n",
    "file_path = '00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "sheet_name = 'clinvar_1kgp_hector_gaf_final'\n",
    "\n",
    "# Load the sheet\n",
    "df_final = pd.read_excel(file_path, sheet_name=sheet_name)\n",
    "\n",
    "# Display column names\n",
    "print(\"✅ Columns in the final sheet:\")\n",
    "print(df_final.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d23944d-b857-4a2d-8209-98382ffd0fee",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Value count | NaN-non-NaN values for AF | AF range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "1d521bd3-b102-4e79-b13a-fd945c17bd07",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Variant type (Total: 120):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant type</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>single nucleotide variant</th>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           count\n",
       "Variant type                    \n",
       "single nucleotide variant    120\n",
       "NaN                           36"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "✅ Molecular consequence (Total: 128):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Molecular consequence</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>missense variant</th>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       count\n",
       "Molecular consequence       \n",
       "missense variant         128\n",
       "NaN                       28"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "✅ Germline classification (Total: 156):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Germline classification</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Uncertain significance</th>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic/Likely pathogenic</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Likely pathogenic</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Benign</th>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Benign/Likely benign</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Conflicting classifications of pathogenicity</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Likely benign</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              count\n",
       "Germline classification                            \n",
       "Uncertain significance                           43\n",
       "Pathogenic/Likely pathogenic                     24\n",
       "Likely pathogenic                                22\n",
       "Benign                                           20\n",
       "Benign/Likely benign                             15\n",
       "Conflicting classifications of pathogenicity     13\n",
       "Likely benign                                    10\n",
       "Pathogenic                                        9"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "✅ gnomAD Allele Frequency:\n",
      "  Non-NaN values: 73\n",
      "  NaN values: 83\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from IPython.display import display\n",
    "\n",
    "# File path and sheet name\n",
    "file_path = '00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "sheet_name = 'clinvar_1kgp_hector_gaf_final'\n",
    "\n",
    "# Load data\n",
    "df = pd.read_excel(file_path, sheet_name=sheet_name)\n",
    "\n",
    "# Variant type value counts\n",
    "print(f\"✅ Variant type (Total: {df['Variant type'].notna().sum()}):\")\n",
    "display(df['Variant type'].value_counts(dropna=False).to_frame(name='count'))\n",
    "\n",
    "# Molecular consequence value counts\n",
    "print(f\"\\n✅ Molecular consequence (Total: {df['Molecular consequence'].notna().sum()}):\")\n",
    "display(df['Molecular consequence'].value_counts(dropna=False).to_frame(name='count'))\n",
    "\n",
    "# Germline classification value counts\n",
    "print(f\"\\n✅ Germline classification (Total: {df['Germline classification'].notna().sum()}):\")\n",
    "display(df['Germline classification'].value_counts(dropna=False).to_frame(name='count'))\n",
    "\n",
    "# gnomAD Allele Frequency NaN/non-NaN\n",
    "non_nan_gaf = df['gnomAD Allele Frequency'].notna().sum()\n",
    "nan_gaf = df['gnomAD Allele Frequency'].isna().sum()\n",
    "print(f\"\\n✅ gnomAD Allele Frequency:\")\n",
    "print(f\"  Non-NaN values: {non_nan_gaf}\")\n",
    "print(f\"  NaN values: {nan_gaf}\")\n",
    "\n",
    "# Add gaf_filled column where NaNs are replaced with 0\n",
    "df['gaf_filled'] = df['gnomAD Allele Frequency'].fillna(0)\n",
    "\n",
    "# Range of gnomAD Allele Frequency by Germline classification\n",
    "#print(\"\\n✅ gnomAD Allele Frequency range by Germline classification (NaN → 0):\")\n",
    "#display(df.groupby('Germline classification')['gaf_filled'].agg(['min', 'max']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "22081445-76cd-4de6-bdbc-2138a04ecd41",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ gnomAD Allele Frequency stats by Germline classification (NaN → 0):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>min_gaf</th>\n",
       "      <th>max_gaf</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Germline classification</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Uncertain significance</th>\n",
       "      <td>43</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>9.106661e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic/Likely pathogenic</th>\n",
       "      <td>24</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Likely pathogenic</th>\n",
       "      <td>22</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>8.256234e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Benign</th>\n",
       "      <td>20</td>\n",
       "      <td>8.275823e-07</td>\n",
       "      <td>4.387166e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Benign/Likely benign</th>\n",
       "      <td>15</td>\n",
       "      <td>8.266034e-07</td>\n",
       "      <td>2.958245e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Conflicting classifications of pathogenicity</th>\n",
       "      <td>13</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.561751e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Likely benign</th>\n",
       "      <td>10</td>\n",
       "      <td>8.253277e-07</td>\n",
       "      <td>6.248724e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>9</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              count       min_gaf  \\\n",
       "Germline classification                                             \n",
       "Uncertain significance                           43  0.000000e+00   \n",
       "Pathogenic/Likely pathogenic                     24  0.000000e+00   \n",
       "Likely pathogenic                                22  0.000000e+00   \n",
       "Benign                                           20  8.275823e-07   \n",
       "Benign/Likely benign                             15  8.266034e-07   \n",
       "Conflicting classifications of pathogenicity     13  0.000000e+00   \n",
       "Likely benign                                    10  8.253277e-07   \n",
       "Pathogenic                                        9  0.000000e+00   \n",
       "\n",
       "                                                   max_gaf  \n",
       "Germline classification                                     \n",
       "Uncertain significance                        9.106661e-06  \n",
       "Pathogenic/Likely pathogenic                  0.000000e+00  \n",
       "Likely pathogenic                             8.256234e-07  \n",
       "Benign                                        4.387166e-02  \n",
       "Benign/Likely benign                          2.958245e-04  \n",
       "Conflicting classifications of pathogenicity  2.561751e-05  \n",
       "Likely benign                                 6.248724e-05  \n",
       "Pathogenic                                    0.000000e+00  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# File path and sheet name\n",
    "file_path = '00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "sheet_name = 'clinvar_1kgp_hector_gaf_final'\n",
    "\n",
    "# Load data\n",
    "df = pd.read_excel(file_path, sheet_name=sheet_name)\n",
    "\n",
    "# Fill NaN gAF with 0 for analysis\n",
    "df['gaf_filled'] = df['gnomAD Allele Frequency'].fillna(0)\n",
    "\n",
    "# Group by 'Germline classification' and calculate count, min, and max\n",
    "summary = df.groupby('Germline classification').agg(\n",
    "    count=('Germline classification', 'count'),\n",
    "    min_gaf=('gaf_filled', 'min'),\n",
    "    max_gaf=('gaf_filled', 'max')\n",
    ").sort_values(by='count', ascending=False)\n",
    "\n",
    "# Display summary\n",
    "print(\"✅ gnomAD Allele Frequency stats by Germline classification (NaN → 0):\")\n",
    "display(summary)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5353dc85-70a9-4386-b7eb-6e5a31793085",
   "metadata": {},
   "source": [
    "## one final check 'Mutation' old vs new unique common"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "bc6d9d83-95d9-474b-9697-dafdfc01d601",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using current file as reference:\n",
      "Total mutations in current file: 156\n",
      "Entries common with old file: 156\n",
      "Entries unique to current file: 0\n",
      "Entries unique to old file: 0\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Paths and sheets\n",
    "old_file = '00_data/01_cdkl5_clinvar_gaf_1kgp_hctr_comb_unq_af_old.xlsx'\n",
    "old_sheet = 'Sheet1_with_AF'  # use actual sheet name, or read first sheet dynamically if unsure\n",
    "\n",
    "current_file = '00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "current_sheet = 'clinvar_1kgp_hector_gaf_final'\n",
    "\n",
    "# Load data\n",
    "df_old = pd.read_excel(old_file, sheet_name=old_sheet)\n",
    "df_current = pd.read_excel(current_file, sheet_name=current_sheet)\n",
    "\n",
    "# Extract 'Mutation' columns as sets\n",
    "mutations_old = set(df_old['mutation'].dropna().unique())\n",
    "mutations_current = set(df_current['Mutation'].dropna().unique())\n",
    "\n",
    "# Compare\n",
    "common = mutations_current.intersection(mutations_old)\n",
    "unique_current = mutations_current.difference(mutations_old)\n",
    "unique_old = mutations_old.difference(mutations_current)\n",
    "\n",
    "# Print results\n",
    "print(f\"Using current file as reference:\")\n",
    "print(f\"Total mutations in current file: {len(mutations_current)}\")\n",
    "print(f\"Entries common with old file: {len(common)}\")\n",
    "print(f\"Entries unique to current file: {len(unique_current)}\")\n",
    "print(f\"Entries unique to old file: {len(unique_old)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25f34123-1899-44df-b3b8-6f83232f4a9d",
   "metadata": {},
   "source": [
    "# 9 final value count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d67888e8-6523-49cd-aeeb-80fdd3749787",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ClinVar', '1KGP_unique', 'Hector2017_unique', 'clinvar_1kgp_hector', 'clinvar_1kgp_hector_gaf', 'clinvar_1kgp_hector_gaf_final']\n",
      "['Name', 'Gene(s)', 'Protein change', 'Condition(s)', 'Accession', 'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome', 'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID', 'Canonical SPDI', 'Variant type', 'Molecular consequence', 'Germline classification', 'Source', 'Germline date last evaluated', 'Germline review status', 'Somatic clinical impact', 'Somatic clinical impact date last evaluated', 'Somatic clinical impact review status', 'Oncogenicity classification', 'Oncogenicity date last evaluated', 'Oncogenicity review status', 'Unnamed: 24', 'gnomAD Allele Frequency', 'Mutation', 'wild', 'position', 'mutant']\n",
      "Germline classification\n",
      "Uncertain significance                          43\n",
      "Pathogenic/Likely pathogenic                    24\n",
      "Likely pathogenic                               22\n",
      "Benign                                          20\n",
      "Benign/Likely benign                            15\n",
      "Conflicting classifications of pathogenicity    13\n",
      "Likely benign                                   10\n",
      "Pathogenic                                       9\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Total count: 156\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Path to your Excel file\n",
    "file_path = '/project/ealexov/compbio/shamrat/250519_energy/00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "\n",
    "# 01 Load Excel file and print sheet names\n",
    "excel_file = pd.ExcelFile(file_path)\n",
    "print(excel_file.sheet_names)\n",
    "\n",
    "# 02 Load the specific sheet\n",
    "sheet_name = 'clinvar_1kgp_hector_gaf_final'\n",
    "df = pd.read_excel(file_path, sheet_name=sheet_name)\n",
    "\n",
    "# Print column names\n",
    "print(df.columns.tolist())\n",
    "\n",
    "# 03 Get value counts for 'Germline classification'\n",
    "value_counts = df['Germline classification'].value_counts(dropna=False)\n",
    "print(value_counts)\n",
    "\n",
    "# Print total count (including NaN if any)\n",
    "total_count = df['Germline classification'].shape[0]\n",
    "print(f\"\\nTotal count: {total_count}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b3869a0d-6667-4a13-8696-6880f02f1720",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Benign</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Conflicting classifications of pathogenicity</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Likely benign</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Germline classification  count\n",
       "0                                        Benign      4\n",
       "1                          Benign/Likely benign     12\n",
       "2  Conflicting classifications of pathogenicity     10\n",
       "3                                 Likely benign      1\n",
       "4                             Likely pathogenic     22\n",
       "5                                    Pathogenic      9\n",
       "6                  Pathogenic/Likely pathogenic     24\n",
       "7                        Uncertain significance     30"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Total count (positions 1-302): 112\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Path to your Excel file\n",
    "file_path = '/project/ealexov/compbio/shamrat/250519_energy/00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "sheet_name = 'clinvar_1kgp_hector_gaf_final'\n",
    "\n",
    "# Load sheet\n",
    "df = pd.read_excel(file_path, sheet_name=sheet_name)\n",
    "\n",
    "# Filter for position range 1-302\n",
    "df_1_302 = df[(df['position'] >= 1) & (df['position'] <= 302)]\n",
    "\n",
    "# Value counts for 'Germline classification', sorted alphabetically by index\n",
    "value_counts = df_1_302['Germline classification'].value_counts(dropna=False).sort_index()\n",
    "\n",
    "# Convert to DataFrame for nice display\n",
    "value_counts_df = value_counts.rename_axis('Germline classification').reset_index(name='count')\n",
    "\n",
    "display(value_counts_df)\n",
    "\n",
    "# Display total count as well\n",
    "print(f\"\\nTotal count (positions 1-302): {df_1_302.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "91523fd9-4170-45ad-8520-b9e3ca9d0c0e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Benign</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Conflicting classifications of pathogenicity</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Likely benign</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Germline classification  count\n",
       "0                                        Benign     20\n",
       "1                          Benign/Likely benign     15\n",
       "2  Conflicting classifications of pathogenicity     13\n",
       "3                                 Likely benign     10\n",
       "4                             Likely pathogenic     22\n",
       "5                                    Pathogenic      9\n",
       "6                  Pathogenic/Likely pathogenic     24\n",
       "7                        Uncertain significance     43"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Total count (positions 1-895): 156\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Path to your Excel file\n",
    "file_path = '/project/ealexov/compbio/shamrat/250519_energy/00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "sheet_name = 'clinvar_1kgp_hector_gaf_final'\n",
    "\n",
    "# Load sheet\n",
    "df = pd.read_excel(file_path, sheet_name=sheet_name)\n",
    "\n",
    "# Filter for position range 1-895\n",
    "df_1_895 = df[(df['position'] >= 1) & (df['position'] <= 895)]\n",
    "\n",
    "# Value counts for 'Germline classification', sorted alphabetically by index\n",
    "value_counts = df_1_895['Germline classification'].value_counts(dropna=False).sort_index()\n",
    "\n",
    "# Convert to DataFrame for nice display\n",
    "value_counts_df = value_counts.rename_axis('Germline classification').reset_index(name='count')\n",
    "\n",
    "display(value_counts_df)\n",
    "\n",
    "# Display total count as well\n",
    "print(f\"\\nTotal count (positions 1-895): {df_1_895.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7b2de88d-3fb2-4ccc-8c76-71c904724254",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>count_1_302</th>\n",
       "      <th>count_1_895</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Benign</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Conflicting classifications of pathogenicity</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Likely benign</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>30</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Germline classification  count_1_302  count_1_895\n",
       "0                                        Benign            4           20\n",
       "1                          Benign/Likely benign           12           15\n",
       "2  Conflicting classifications of pathogenicity           10           13\n",
       "3                                 Likely benign            1           10\n",
       "4                             Likely pathogenic           22           22\n",
       "5                                    Pathogenic            9            9\n",
       "6                  Pathogenic/Likely pathogenic           24           24\n",
       "7                        Uncertain significance           30           43"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total count (positions 1-302): 112\n",
      "Total count (positions 1-895): 156\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# Path to your Excel file\n",
    "file_path = '/project/ealexov/compbio/shamrat/250519_energy/00_data/clinvar_kgp_hector_unique_summary.xlsx'\n",
    "sheet_name = 'clinvar_1kgp_hector_gaf_final'\n",
    "\n",
    "# Load sheet\n",
    "df = pd.read_excel(file_path, sheet_name=sheet_name)\n",
    "\n",
    "# Filter for position ranges\n",
    "df_1_302 = df[(df['position'] >= 1) & (df['position'] <= 302)]\n",
    "df_1_895 = df[(df['position'] >= 1) & (df['position'] <= 895)]\n",
    "\n",
    "# Get value counts, sorted alphabetically\n",
    "vc_1_302 = df_1_302['Germline classification'].value_counts(dropna=False).sort_index()\n",
    "vc_1_895 = df_1_895['Germline classification'].value_counts(dropna=False).sort_index()\n",
    "\n",
    "# Combine in a single DataFrame\n",
    "counts_df = pd.DataFrame({\n",
    "    'Germline classification': vc_1_302.index,\n",
    "    'count_1_302': vc_1_302.values,\n",
    "}).set_index('Germline classification')\n",
    "\n",
    "counts_df['count_1_895'] = vc_1_895\n",
    "\n",
    "# Reset index for display\n",
    "counts_df = counts_df.reset_index()\n",
    "\n",
    "display(counts_df)\n",
    "\n",
    "# Optionally, print total counts for each range\n",
    "print(f\"Total count (positions 1-302): {df_1_302.shape[0]}\")\n",
    "print(f\"Total count (positions 1-895): {df_1_895.shape[0]}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
