{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 250401_01_docking_data_processing "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1.00 Preview input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    SL UniProt                   Gene  pSite Consensus_motif_(R-P-X-[S/T]-[A/G/P/S])                                                             Protein                                                                                                                                                                                                                                Protein function in relation to CDKL5                                    Model system (e.g. animal, cell line, etc.)   Full length or kinase domain           In vitro or in vivo                                                                                                                                                                                                                                                                                              Method of identification        Citation\n",
      "0    1  O76039                  CDKL5   Y171                                     TEY                                      Cyclin-dependent kinase-like 5                                                                                                                                                                                                Mediates phosphorylation of MECP2 (PubMed:15917271, PubMed:16935860).                                                                            NaN                            NaN                           NaN                                                                                                                                                                                                                                                                                                                   NaN             NaN\n",
      "1    2  Q92974                ARHGEF2   S122                           TIRERPSsAIYPS                            Rho guanine nucleotide exchange factor 2                                                                                                                                  Mediates interaction between microtubules and actin cytoskeleton. \\ Activates RHO-GTPases. \\ Regulating dendritic spine morphology.                                              Mouse brain lysates, HEK293 cells  Full length and Kinase Domain  In vitro and In vivo (mouse)                                                                                                                                                                                                         Chemical-genetic approach using an ATP-analogue specific CDKL5 and identified using LC-MS/MS and western blot            [69]\n",
      "2    3  P49418                  AMPH1   S293                           PAPARPRsPSQTR                                                        Amphiphysin1                                                                                                                                                                    Involved in neural transmission and development. \\ Associated with clathrin-mediated endocytosis.                          Mouse brain lysates, Recombinant protein from E. coli                  Kinase domain                      In vitro                                                                                                                                                                          Brain lysates incubated with CDKL5 Kinase domain with radiolabelled ATP. Mouse Amph1 purified from E. coli directly phosphorylated by CDKL5.         [84,85]\n",
      "3    4  Q9UPN4                 CEP131    S35                           PVSRRPGsAATTK                                      Centrosomal protein of 131 kDa                                                                                                                                              Centrosomal protein involved in the formation and function of primary cilia. \\ Maintains centriolar satellite integrity                                                                   HEK293, U2OS                    Full length                      In vitro                                                                                                                                                                               Direct phosphorylation of synthetic peptide with wild-type or kinase dead CDKL5. Co-expression with CDKL5 saw increased phosphorylation            [12]\n",
      "4    5  Q8TDM6                   DLG5  S1115                           QKRRRPKsAPSFR                                               Disks large homolog 5                                                                                                                                              Maintains cell polarity.' \\ Signaling to the microtubule-based cytoskeleton. \\ Maintenance of epithelial cell structure                                                                   HEK293, U2OS                    Full length                      In Vitro                                                                                                                                                                               Direct phosphorylation of synthetic peptide with wild-type or kinase dead CDKL5. Co-expression with CDKL5 saw increased phosphorylation            [12]\n",
      "5    6  Q14241                   ELOA   S311                           EENRRPPsGDNAR                                                           Elongin A                                                                                                                                                                                          Transcriptional elongation factor. \\ Involved in DNA double-stranded repair                                                                   HEK293, U2OS                    Full length                      In vitro                                                                                                                                                              CDKL5 overexpressed in nucleus. Targets identified via phosphoproteomic. Direct phosphorylation of synthetic peptide with wild-type or kinase dead CDKL5            [86]\n",
      "6    7  Q96L91                  EP400   S729                           SPVNRPSsATNKA                                           EE1A-binding protein p400                                                                                                                                                             Involved in transcriptional activation of select genes by acetylation of nucleosome histones H4 and H2A.                                                                   HEK293, U2OS                    Full length                      In vitro                                                                                                                                                              CDKL5 overexpressed in nucleus. Targets identified via phosphoproteomic. Direct phosphorylation of synthetic peptide with wild-type or kinase dead CDKL5            [86]\n",
      "7    8  Q66K74                  MAP1S   S871                           KAPARPSsASATP                                   Microtubule-associated protein 1S                                                                 Binds to both microtubules and actin, potentially cross-linking and stabilizing the two proteins.\\ Bridges microtubules to mitochondria. \\Might link NMDA receptor subunit NR3A to the cytoskeleton.                                              Mousse brain lysate, HEK293, U2OS  Full length and kinase domain          In vivo and in vitro                                                                Chemical-genetic approach using an ATP-analogue specific CDKL5 and identified using LC-MS/MS and western blot. Direct phosphorylation of synthetic peptide with wild-type or kinase dead CDKL5. Co-expression with CDKL5 saw increased phosphorylation         [12,69]\n",
      "8    9  Q66K74                  MAP1S   S900                           DRASRPLsARSEP                                   Microtubule-associated protein 1S                                                                 Binds to both microtubules and actin, potentially cross-linking and stabilizing the two proteins.\\ Bridges microtubules to mitochondria. \\Might link NMDA receptor subunit NR3A to the cytoskeleton.                                              Mousse brain lysate, HEK293, U2OS  Full length and kinase domain          In vivo and in vitro                                                                Chemical-genetic approach using an ATP-analogue specific CDKL5 and identified using LC-MS/MS and western blot. Direct phosphorylation of synthetic peptide with wild-type or kinase dead CDKL5. Co-expression with CDKL5 saw increased phosphorylation         [12,69]\n",
      "9   10  Q15555             EB2_MAPRE2   S222                           STPSRPSsAKRAS                Microtubule-associated protein RP/EB family member 2                                                                                                                                                                                     Microtubule end-binding protein which regulates microtubule dynamic instability.                                              Mouse brain lysates, HEK293 cells  Full length and Kinase Domain  In vitro and In vivo (mouse)                                                                                                                                                                                                         Chemical-genetic approach using an ATP-analogue specific CDKL5 and identified using LC-MS/MS and western blot            [69]\n",
      "10  11  Q8TAP9                  TTDN1    S40                           GGGPRPPsPRDGY                                    TTD non-photosensitive 1 protein                                                                                                                                              TTDN1 is mutated in a form of tricothiodystrophy (TTD), typically caused by failure in transcription–coupled DNA repair                                                                   HEK293, U2OS                    Full length                      In vitro                                                                                                                                                                                                                                      CDKL5 overexpressed in nucleus. Targets identified via phosphoproteomic methods.            [86]\n",
      "11  12  P26358                  DNMT1    NaN                                     NaN                                             DNA methyltransferase 1                                      Maintains patterns of methylated cytosine residues. It should be noted that the phosphorylation of Dnmt1(1–290) by CDKL5 was significantly stimulated in the presence of DNA. MeCP2 is also a well-known DNMT1-binding protein.                 Mouse brain extract, HEK293T, Recombinant protein from E. coli  Full length and Kinase domain                      In vitro                                                                                                                                 Nuclear co-localization studies in HEK293T cells. GST pulldown assays from mouse brain extract and HEK239T cells. Mouse DNMT1 purified from E. coli directly phosphorylated by CDKL5.           [114]\n",
      "12  13  P56524                  HDAC4   S632                         RPLSRAQsSPASAtF                                               Histone deacetylase 4                    An enzyme that catalyzes histone de-acetylation, repressing transcription. Key role in brain development and neuronal survival. In neurons, primarily in cytoplasm but upon CDKL5 loss, translocates to the nucleus reducing histone acetylation.                                                      KO mice, HEK293T, SH-SY5Y                    Full length          In vitro and In vivo                                                                                                                           Phosphoprotein profiling of SH-SY5S cells by the Phospho Explorer antibody microarray. Confirmed targets by co-immunoprecipitation and direct phosphorylation studies in vitro and in vivo.           [115]\n",
      "13  14  Q9HCJ2  NGL-1_KIAA1580_LRRC4C   S631                         PLLIRMNsKDNVQET                                                   Netrin-G ligand-1                                                                                                                                   Transmembrane protein highly expressed in the brain which interacts with a specific axon guidance molecule. Axon guidance molecule  Mouse brain extract, HEK293T, COS-7, Primary fibroblast, iPSC-derived neurons                    Full length          In vitro and In vivo                                                                                                 Co-immunoprecipitation in vitro and in vivo, co-localization studies in neurons and western blotting. Phosphorylation confirmed in overexpression studies and direct phosphorylation of peptide incubated with CDKL5.            [54]\n",
      "14  15  P84022                  SMAD3    NaN                                     NaN                           Mothers against decapentaplegic homolog 3                                                                                                                                            Functions within the TGFβ pathway, regulating transcription by transducing signals from the cell membrane to the nucleus.                    ouse brain extract, HEK293T, SH-SY5Y, primary mouse neurons                    Full length          In vitro and In vivo  Phosphoprotein profiling of cortex extracts from CDKL5 wt and KO mouse cells by the Phospho Explorer antibody microarray. Co-immunoprecipitation assays in established cell lines. Expression analysis in CDKL5 KO mouse neurons and direct phosphorylation confirmed with purified recombinant protein from HEK293T            [52]\n",
      "15  16  P48436                   SOX9   S199                         ATEQTHIsPNAIFKA                                          Transcription factor SOX-9  Transcription factor of the SOX family with critical roles in cell fate determination. Sox9 suppresses cell death during development, adult tissue homeostasis, and oncogenesis. Phosphorylation-dependent suppression of pro-survival transcription regulator Sox9                                                     Mouse renal tissue, HEK293                    Full length          In vitro and In vivo                                                                                              Identified using siRNA Kinomic screening and direct phosphorylation confirmed with recombinant human protein and in mouse renal tissue extract. CDKL5 physically interacts with SOX9 as shown by co-immunoprecipitation.           [116]\n",
      "16  17  P46940                 IQGAP1    NaN                                     NaN                     IQ Motif Containing GTPase Activating Protein 1                                                                                                                                                            IQGAP — Fundamental regulator of cell migration and polarity by regulating cytoskeletal effector proteins                                             Primary mouse neurons, HeLa, COS-7                    Full length                      In vitro                                                                                 Interaction by yeast-two hybrid screening and immunoprecipitation from mouse brain lysates. No direct phosphorylation shown, however, IQGAP localization in CDLK5 KO cells could be rescued with WT CDLK5 but not Kinase dead mutant.            [59]\n",
      "17  18  P51608                  MeCP2    NaN                                     NaN                                        Methyl-CpG binding protein 2                                                                                                                                                      Binds to methylated DNA and represses transcription. Rett syndrome gene. Can activate and repress transcription                                                                        HEK293T                    Full Length                      In vitro                                                                                                                                      GFP pulldown assay in HEK293T. Purified recombinant protein incubated with CDKL5 showed phosphorylation. Follow up in vitro kinase assays contradict MeCP2 as a CDKL5 substrate.         [41,42]\n",
      "18  19  P78352             PSD95_DLG4    NaN                                     NaN                                     Postsynaptic density protein 95                                                                                                                                                                                                                    PSD95 is critical for dendritic spine development                                                     Rat brain extract, HEK293T                    Full length                      In vitro                                                                                                                                                                   GST pulldown with recombinant PSD95 and CDKL5 in HEK293T. Affinity purification of endogenous PSD95 from rat brain extract using column bound CDKL5            [57]\n",
      "19  20  A0MZ66            SHTN1_SHOT1    NaN                                     NaN                                                            Shootin1                                                                                                                                                                                                                     Promotes neuron polarization and axon outgrowth.                                     Mouse brain extract, primary mouse neurons                    Full length                      In vitro                                                                                 Yeast two-hybrid screening identified SHTN1. Confirmed with Co-immunoprecipitation of endogenous SHTN1 and CDKL5 in brain lysates. No direct phosphorylation shown however, SHTN1 phosphorylation levels reduced in CDKL5 KO neurons.      Bergen2018\n",
      "20  21  P78347                  GTF2I   S674                           QSPKRPRsPGSNS                                   General transcription factor II-I                                                                                                                                                                                                                                                                  NaN                                                                            NaN                            NaN                           NaN                                                                                                                                                                                                                                                                                                                   NaN      Massey2024\n",
      "21  22  Q8TAP8                PPP1R35    S52                           SLSPRPDsPQPRH                         Protein phosphatase 1 regulatory subunit 35                                                                                                                                                                                                                                                                  NaN                                                                            NaN                            NaN                           NaN                                                                                                                                                                                                                                                                                                                   NaN      Massey2024\n",
      "22  23  Q86YP4                GATAD2A   S100                           KSERRPPsPDVIV                                 Transcriptional repressor p66-alpha                                                                                                                                                                                                                                                                  NaN                                                                            NaN                            NaN                           NaN                                                                                                                                                                                                                                                                                                                   NaN      Massey2024\n",
      "23  24  Q9P2Y4                 ZNF219   S114                           HQPERPRsPAARL                                             Zinc finger protein 219                                                                                                                                                                                                                                                                  NaN                                                                            NaN                            NaN                           NaN                                                                                                                                                                                                                                                                                                                   NaN      Massey2024\n",
      "24  25  Q15878         CACNA1E_Cav2.3    S14                            AVVARPGsGDGD  Voltage-dependent R-type calcium channel subunit alpha-1E (Cav2.3)                                                                                                                                                                                                                                                                  NaN                                                                            NaN                            NaN                           NaN                                                                                                                                                                                                                                                                                                                   NaN  PMID: 38081835\n",
      "25  26  Q13501             SQSTM1_p62   S272                            RSRLTPVsPESS                                                 Sequestosome-1(p62)                                                                                                                                                                                                                                                                  NaN                                                                            NaN                            NaN                           NaN                                                                                                                                                                                                                                                                                                                   NaN  PMID: 34605059\n",
      "26  27  Q13501             SQSTM1_p62   S272                            GGKRSRLtPVSP                                                 Sequestosome-1(p62)                                                                                                                                                                                                                                                                  NaN                                                                            NaN                            NaN                           NaN                                                                                                                                                                                                                                                                                                                   NaN  PMID: 34605059\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Define the file path\n",
    "file_path = '00_cdkl5_targets.xlsx'\n",
    "\n",
    "# Load the first sheet of the Excel file\n",
    "df = pd.read_excel(file_path, sheet_name=0)\n",
    "\n",
    "# Display the entire DataFrame\n",
    "print(df.to_string())  # Converts the entire DataFrame to a string for full display"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 01 CDKL5 Target Motif Extraction Script\n",
    "\n",
    "This Python script processes an Excel file containing CDKL5 target information, retrieves protein sequences from UniProt, and extracts consensus motifs or sequences around phosphorylation sites (pSites).\n",
    "\n",
    "## Purpose\n",
    "\n",
    "The script aims to:\n",
    "\n",
    "1.  Retrieve protein sequences from UniProt based on UniProt IDs.\n",
    "2.  Identify consensus motifs matching the pattern `RP.[ST][AGPS]` within the sequences.\n",
    "3.  If no consensus motif is found, extract a 5-amino acid sequence around the provided pSite.\n",
    "4.  Record the extracted motifs, their ranges, and the extraction method used.\n",
    "5.  Output the results to a new Excel file.\n",
    "\n",
    "## Input Excel File Format\n",
    "\n",
    "The input Excel file should contain the following columns:\n",
    "\n",
    "* **UniProt:** UniProt accession ID (e.g., P56524).\n",
    "* **pSite:** Phosphorylation site (e.g., S632).\n",
    "* **Other columns:** Any additional information related to the targets.\n",
    "\n",
    "## Output Excel File Format\n",
    "\n",
    "The output Excel file will contain all the original columns from the input file, plus the following:\n",
    "\n",
    "* **Consensus Motifs Found:** The extracted consensus motif(s) or the sequence around the pSite.\n",
    "* **Consensus Motif Range:** The range of amino acid positions for the extracted motif.\n",
    "* **Extraction Method:** Indicates how the motif was extracted (Consensus Motif, pSite, None, pSite Extraction Failed, Sequence Retrieval Failed).\n",
    "\n",
    "## Script Details\n",
    "\n",
    "* **`get_uniprot_sequence(uniprot_id)`:** Retrieves protein sequences from UniProt.\n",
    "* **`find_consensus_motifs(sequence)`:** Finds consensus motifs using the regular expression `RP.[ST][AGPS]`.\n",
    "* **`get_motif_range(motif)`:** Extracts the range of amino acid positions for a motif.\n",
    "* **`extract_motif_around_psite(sequence, psite_aa, psite_pos_excel, motif_length=5)`:** Extracts a sequence around the pSite.\n",
    "    * `psite_aa`: the amino acid, e.g., 'S'\n",
    "    * `psite_pos_excel`: the position from the excel file, e.g. '632'\n",
    "    * `motif_length`: the length of the extracted sequence (default: 5).\n",
    "\n",
    "## Error Handling\n",
    "\n",
    "* The script handles cases where UniProt sequences cannot be retrieved.\n",
    "* It provides informative error messages for pSite extraction failures.\n",
    "* It identifies whether the extracted motif was found using the consensus motif or the pSite.\n",
    "\n",
    "## Example Output\n",
    "\n",
    "| UniProt | pSite | Consensus Motifs Found | Consensus Motif Range | Extraction Method |\n",
    "| ------- | ----- | ---------------------- | --------------------- | ----------------- |\n",
    "| P56524  | S632  | ['RPLSRAQ']           | 631-637               | pSite             |\n",
    "| Q92974  | S122  | ['RPSSA']              | 119-123               | Consensus Motif   |\n",
    "| P84022  |       | []                     |                       | None              |\n",
    "| O76039  | Y171  | []                     |                       | pSite Extraction Failed: pSite amino acid not found in sequence |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 27/27 [00:25<00:00,  1.06it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Analysis complete. Updated file saved as 'updated_consensus_motifs_psite_position.xlsx'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import re\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "def get_uniprot_sequence(uniprot_id):\n",
    "    \"\"\"Retrieves the protein sequence from UniProt.\"\"\"\n",
    "    url = f\"https://www.uniprot.org/uniprot/{uniprot_id}.fasta\"\n",
    "    try:\n",
    "        response = requests.get(url)\n",
    "        response.raise_for_status()\n",
    "        fasta_data = response.text\n",
    "        sequence = \"\".join(fasta_data.splitlines()[1:])\n",
    "        return sequence\n",
    "    except requests.exceptions.RequestException as e:\n",
    "        return None\n",
    "\n",
    "def find_consensus_motifs(sequence):\n",
    "    \"\"\"Finds consensus motifs in the protein sequence.\"\"\"\n",
    "    motif_pattern = re.compile(r\"RP.[ST][AGPS]\")\n",
    "    return [(match.start() + 1, match.group(0)) for match in motif_pattern.finditer(sequence)]\n",
    "\n",
    "def get_motif_range(motif):\n",
    "    \"\"\"Extracts the range from a single motif.\"\"\"\n",
    "    return f\"{motif[0]}-{motif[0] + len(motif[1]) - 1}\"\n",
    "\n",
    "def extract_motif_around_psite(sequence, psite_aa, psite_pos_excel, motif_length=5):\n",
    "    \"\"\"Extracts motif around the pSite with specified length.\"\"\"\n",
    "    if not isinstance(psite_aa, str):\n",
    "        return None, None, \"Invalid pSite amino acid\"\n",
    "\n",
    "    psite_aa_upper = psite_aa.upper()\n",
    "    psite_positions_in_seq = [i + 1 for i, aa in enumerate(sequence.upper()) if aa == psite_aa_upper]\n",
    "\n",
    "    if not psite_positions_in_seq:\n",
    "        return None, None, \"pSite amino acid not found in sequence\"\n",
    "\n",
    "    # Find the closest matching pSite in the sequence to the provided position.\n",
    "    closest_psite_pos = min(psite_positions_in_seq, key=lambda x: abs(x - int(psite_pos_excel)))\n",
    "\n",
    "    start = max(0, closest_psite_pos - (motif_length // 2))\n",
    "    end = min(len(sequence), closest_psite_pos + (motif_length - (motif_length // 2)))\n",
    "    extracted_motif = sequence[start - 1:end - 1] #index correction\n",
    "    extracted_range = f\"{start}-{end}\"\n",
    "    return extracted_motif, extracted_range, None\n",
    "\n",
    "# Load data from Excel\n",
    "excel_path = \"00_cdkl5_targets.xlsx\"\n",
    "df = pd.read_excel(excel_path)\n",
    "df.columns = df.columns.str.strip()\n",
    "\n",
    "results = []\n",
    "\n",
    "# Process each entry\n",
    "for i, row in tqdm(df.iterrows(), total=len(df)):\n",
    "    uniprot_id = row['UniProt']\n",
    "    psite = row['pSite']\n",
    "    sequence = get_uniprot_sequence(uniprot_id)\n",
    "\n",
    "    if sequence:\n",
    "        motifs = find_consensus_motifs(sequence)\n",
    "        if motifs:\n",
    "            for motif in motifs:\n",
    "                result_row = row.to_dict()\n",
    "                result_row['Consensus Motifs Found'] = str([motif])\n",
    "                result_row['Consensus Motif Range'] = get_motif_range(motif)\n",
    "                result_row['Extraction Method'] = 'Consensus Motif'\n",
    "                results.append(result_row)\n",
    "        elif pd.notna(psite) and psite != \"\":\n",
    "            psite_aa = psite[0]  # Get the amino acid (e.g., 'S')\n",
    "            psite_pos_excel = psite[1:] # get position from excel.\n",
    "            extracted_motif, extracted_range, error = extract_motif_around_psite(sequence, psite_aa, psite_pos_excel)\n",
    "            if extracted_motif:\n",
    "                result_row = row.to_dict()\n",
    "                result_row['Consensus Motifs Found'] = str([extracted_motif])\n",
    "                result_row['Consensus Motif Range'] = extracted_range\n",
    "                result_row['Extraction Method'] = 'pSite'\n",
    "                results.append(result_row)\n",
    "            else:\n",
    "                result_row = row.to_dict()\n",
    "                result_row['Consensus Motifs Found'] = '[]'\n",
    "                result_row['Consensus Motif Range'] = ''\n",
    "                result_row['Extraction Method'] = f'pSite Extraction Failed: {error}'\n",
    "                results.append(result_row)\n",
    "        else:\n",
    "            result_row = row.to_dict()\n",
    "            result_row['Consensus Motifs Found'] = '[]'\n",
    "            result_row['Consensus Motif Range'] = ''\n",
    "            result_row['Extraction Method'] = 'None'\n",
    "            results.append(result_row)\n",
    "    else:\n",
    "        result_row = row.to_dict()\n",
    "        result_row['Consensus Motifs Found'] = ''\n",
    "        result_row['Consensus Motif Range'] = ''\n",
    "        result_row['Extraction Method'] = 'Sequence Retrieval Failed'\n",
    "        results.append(result_row)\n",
    "\n",
    "df_results = pd.DataFrame(results)\n",
    "df_results.to_excel(\"01_updated_consensus_motifs_psite_position.xlsx\", index=False)\n",
    "print(\"✅ Analysis complete. Updated file saved as 'updated_consensus_motifs_psite_position.xlsx'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 02 Get PDB using uniprot\n",
    "get available protein structure into excel both from protein data bank and alphafold corresponding to each uniprot ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated file saved as 02_updated_consensus_motifs_with_pdb.xlsx\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "# Define file paths\n",
    "input_file = \"01_updated_consensus_motifs_psite_position.xlsx\"  # Input file\n",
    "output_file = \"02_updated_consensus_motifs_with_pdb.xlsx\"  # Updated output file\n",
    "\n",
    "# Load the Excel sheet\n",
    "df = pd.read_excel(input_file, sheet_name=0)\n",
    "\n",
    "# Check if required columns exist\n",
    "required_columns = [\"UniProt\", \"Gene\", \"Protein\", \"pSite\", \"Consensus_motif_(R-P-X-[S/T]-[A/G/P/S])\"]\n",
    "for col in required_columns:\n",
    "    if col not in df.columns:\n",
    "        raise ValueError(f\"Missing required column: {col}\")\n",
    "\n",
    "# Function to fetch PDB structural data from UniProt\n",
    "def get_uniprot_structures(uniprot_id):\n",
    "    url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.json\"\n",
    "    response = requests.get(url)\n",
    "    \n",
    "    structures = []\n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        if \"uniProtKBCrossReferences\" in data:\n",
    "            for entry in data[\"uniProtKBCrossReferences\"]:\n",
    "                if entry[\"database\"] == \"PDB\":\n",
    "                    pdb_id = entry[\"id\"]\n",
    "                    method = next((prop[\"value\"] for prop in entry.get(\"properties\", []) if prop[\"key\"] == \"Method\"), \"N/A\")\n",
    "                    resolution = next((prop[\"value\"] for prop in entry.get(\"properties\", []) if prop[\"key\"] == \"Resolution\"), \"N/A\")\n",
    "                    chains_positions = next((prop[\"value\"] for prop in entry.get(\"properties\", []) if prop[\"key\"] == \"Chains\"), \"N/A\")\n",
    "\n",
    "                    chain_list, position_list = [], []\n",
    "                    if chains_positions != \"N/A\":\n",
    "                        for part in chains_positions.split(\", \"):\n",
    "                            if \"=\" in part:\n",
    "                                chain, position = part.split(\"=\")\n",
    "                                chain_list.append(chain.strip())\n",
    "                                position_list.append(position.strip())\n",
    "                            else:\n",
    "                                chain_list.append(part.strip())\n",
    "                                position_list.append(\"N/A\")\n",
    "\n",
    "                    structures.append({\n",
    "                        \"PDB_ID\": pdb_id,\n",
    "                        \"Method\": method,\n",
    "                        \"Resolution\": resolution,\n",
    "                        \"Chain\": \", \".join(chain_list),\n",
    "                        \"Positions\": \", \".join(position_list),\n",
    "                        \"Source\": \"UniProt-PDB\"\n",
    "                    })\n",
    "    \n",
    "    return structures\n",
    "\n",
    "# Function to add AlphaFold structure\n",
    "def get_alphafold_structure(uniprot_id):\n",
    "    alphafold_id = f\"AF-{uniprot_id}-F1\"\n",
    "    \n",
    "    return [{\n",
    "        \"PDB_ID\": alphafold_id,\n",
    "        \"Method\": \"AlphaFold Prediction\",\n",
    "        \"Resolution\": \"N/A\",\n",
    "        \"Chain\": \"N/A\",\n",
    "        \"Positions\": \"N/A\",\n",
    "        \"Source\": \"AlphaFold DB\"\n",
    "    }]\n",
    "\n",
    "# Collect all structures and map them to input data\n",
    "pdb_data = []\n",
    "for _, row in df.iterrows():\n",
    "    uniprot_id = row[\"UniProt\"]\n",
    "\n",
    "    # Fetch UniProt PDB structures\n",
    "    pdb_results = get_uniprot_structures(uniprot_id)\n",
    "\n",
    "    # Always add AlphaFold prediction\n",
    "    pdb_results.extend(get_alphafold_structure(uniprot_id))\n",
    "\n",
    "    for pdb_entry in pdb_results:\n",
    "        pdb_entry.update({\n",
    "            \"UniProt\": uniprot_id\n",
    "        })\n",
    "        pdb_data.append(pdb_entry)\n",
    "\n",
    "# Convert to DataFrame\n",
    "pdb_df = pd.DataFrame(pdb_data)\n",
    "\n",
    "# Merge original data with PDB structures\n",
    "df_updated = df.merge(pdb_df, on=\"UniProt\", how=\"left\")\n",
    "\n",
    "# Save the updated DataFrame to Excel\n",
    "df_updated.to_excel(output_file, index=False)\n",
    "\n",
    "print(f\"Updated file saved as {output_file}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 03 Alphafold entries only\n",
    "- Upon investigation most PDB structures missing pSite\n",
    "- Thus focus on alphafold entries only\n",
    "- get updated excel sheet with alphafold entreis only\n",
    "- Downlaod alphafold structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated AlphaFold entries saved to 03_updated_alphafold_entries.xlsx\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "import os\n",
    "\n",
    "# Define file paths\n",
    "input_file = \"01_updated_consensus_motifs_psite_position.xlsx\"  # Input Excel file\n",
    "output_file = \"03_updated_alphafold_entries.xlsx\"  # Output Excel file\n",
    "structure_folder = \"03_alphafold_structures/\"  # Folder to save AlphaFold PDB files\n",
    "\n",
    "# Ensure the structure folder exists\n",
    "os.makedirs(structure_folder, exist_ok=True)\n",
    "\n",
    "# Load the Excel sheet\n",
    "df = pd.read_excel(input_file, sheet_name=0)\n",
    "\n",
    "# Function to generate AlphaFold PDB ID and URL\n",
    "def get_alphafold_pdb_info(uniprot_id):\n",
    "    pdb_id = f\"AF-{uniprot_id}-F1\"\n",
    "    pdb_url = f\"https://alphafold.ebi.ac.uk/files/{pdb_id}-model_v4.pdb\"\n",
    "    return pdb_id, pdb_url\n",
    "\n",
    "# Function to download AlphaFold PDB structure\n",
    "def download_alphafold_pdb(uniprot_id):\n",
    "    pdb_id, pdb_url = get_alphafold_pdb_info(uniprot_id)\n",
    "    output_path = os.path.join(structure_folder, f\"{pdb_id}.pdb\")\n",
    "    \n",
    "    response = requests.get(pdb_url)\n",
    "    if response.status_code == 200:\n",
    "        with open(output_path, \"wb\") as f:\n",
    "            f.write(response.content)\n",
    "        return output_path  # Return file path if successful\n",
    "    else:\n",
    "        return \"Download failed\"  # Indicate failure\n",
    "\n",
    "# Process each UniProt ID\n",
    "df[\"AlphaFold_PDB_ID\"], df[\"AlphaFold_PDB_URL\"] = zip(*df[\"UniProt\"].apply(get_alphafold_pdb_info))\n",
    "df[\"AlphaFold_PDB_File_Path\"] = df[\"UniProt\"].apply(download_alphafold_pdb)\n",
    "\n",
    "# Save the updated DataFrame to an Excel file\n",
    "df.to_excel(output_file, index=False)\n",
    "\n",
    "print(f\"Updated AlphaFold entries saved to {output_file}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
